// @oracle/oraclejet-preact/cjs/utils/UNSAFE_interpolations/boxalignment.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require("identity-obj-proxy");
var keys = require('../../keys-65201aa4.js');
var utils_UNSAFE_size = require('../UNSAFE_size.js');
var utils_UNSAFE_arrayUtils = require('../UNSAFE_arrayUtils.js');
require('../../_curry1-6e9511bf.js');
require('../../_has-aa203696.js');
require('../../UNSAFE_Theme.js');
require('../../utils-b9f57550.js');
require('../../Common/themes/themeContract.css.js');

// boxalignment are styles that help with alignment for Grid/Flex styles, like justifyContent,
// align, and gap.
const alignStyles = {
  baseline: "_1m3e7gi",
  center: "_1fef7u5",
  end: "_145hvzw",
  start: "_189tazq",
  inherit: "_1le3nti",
  initial: "_1wi46qp",
  stretch: "_1lildin"
};
const justifyStyles = {
  center: "_1ua877h",
  end: "zialjf",
  start: "tm4314",
  inherit: "_1xhxxpj",
  initial: "_67b5nf",
  around: "_1nnxib8",
  between: "ty75dt",
  evenly: "c9mvrk"
};
const boxAlignmentInterpolations = {
  align: ({
    align
  }) => align === undefined ? {} : {
    class: alignStyles[align]
  },
  justify: ({
    justify
  }) => justify === undefined ? {} : {
    class: justifyStyles[justify]
  },
  // See https://developer.mozilla.org/en-US/docs/Web/CSS/gap
  gap: ({
    gap
  }) => {
    if (gap === undefined) {
      return {};
    } else {
      const [rowSize, columnSize = rowSize] = utils_UNSAFE_arrayUtils.coerceArray(gap);
      return {
        gap: `${utils_UNSAFE_size.sizeToCSS(rowSize)} ${utils_UNSAFE_size.sizeToCSS(columnSize)}`
      };
    }
  }
};
// These consts are exported so they can be used in stories to show the allowed options
// e.g., See Flex.stories.tsx:   align: {control: 'select',options: aligns}
const aligns = keys.keys_1(alignStyles);
const justifies = keys.keys_1(justifyStyles);

exports.aligns = aligns;
exports.boxAlignmentInterpolations = boxAlignmentInterpolations;
exports.justifies = justifies;
