// @oracle/oraclejet-preact/es/Chip-58b02e10.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx } from 'preact/jsx-runtime';
import { useActionable } from './hooks/UNSAFE_useActionable.js';
import { useCallback } from 'preact/hooks';
import { useTabbableMode } from './hooks/UNSAFE_useTabbableMode.js';
import { multiVariantStyles } from './UNSAFE_Chip/themes/ChipStyles.css.js';
import { useInteractionStyle } from './hooks/UNSAFE_useInteractionStyle.js';
import { mergeProps } from './utils/UNSAFE_mergeProps.js';

function Chip({ isSelected, isDisabled, 'aria-label': accessibleLabel, children, onToggle }) {
    const handleAction = useCallback(() => {
        onToggle?.({ previousValue: isSelected, value: !isSelected });
    }, [isSelected, onToggle]);
    const { actionableProps } = useActionable(handleAction);
    const { interactionProps, applyActiveStyle, applyHoverStyle, applyPseudoHoverStyle } = useInteractionStyle();
    const mergedProps = mergeProps(actionableProps, interactionProps);
    const classes = multiVariantStyles({
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        hover: applyHoverStyle && !isDisabled && !isSelected ? 'isHover' : 'notHover',
        pseudoHover: applyPseudoHoverStyle && !isDisabled && !isSelected ? 'isPseudoHover' : 'notPseudoHover',
        active: isSelected || applyActiveStyle ? 'isActive' : 'notActive'
    });
    const { isTabbable } = useTabbableMode();
    return isDisabled ? (jsx("div", { class: classes, children: children })) : (jsx("div", { ...mergedProps, class: classes, tabIndex: isTabbable ? 0 : -1, "aria-label": accessibleLabel, role: "switch", "aria-checked": isSelected ? 'true' : 'false', children: children }));
}

export { Chip as C };
