// @oracle/oraclejet-preact/es/CompactHelpSource-8d377623.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsxs, Fragment, jsx } from 'preact/jsx-runtime';
import { useRef, useCallback, useState } from 'preact/hooks';
import { useTabbableMode } from './hooks/UNSAFE_useTabbableMode.js';
import { useTranslationBundle } from './hooks/UNSAFE_useTranslationBundle.js';
import './utils/UNSAFE_classNames.js';
import { HelpIcon as SvgHelp } from './PRIVATE_ThemedIcons/HelpIcon.js';
import { Popup } from './UNSAFE_Popup.js';
import './hooks/UNSAFE_useFocusWithin.js';
import { u as useFocusWithin } from './useFocusWithin-f96a205e.js';
import { helpIconBase, helpIconLabelEdgeStart, helpContentStyles } from './UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js';
import { useComponentTheme } from './hooks/UNSAFE_useComponentTheme.js';
import { LabelRedwoodTheme } from './UNSAFE_Label/themes/redwood/LabelTheme.js';
import { c as classNames } from './classNames-f414daad.js';

function usePopupFocusWithin() {
    const ref = useRef(null);
    const { focusProps, isFocused } = useFocusWithin();
    const onTransitionEnd = useCallback((isAfterOpen) => {
        if (isAfterOpen) {
            if (focusProps.onfocusin) {
                ref.current?.addEventListener('focusin', focusProps.onfocusin, false);
            }
            if (focusProps.onfocusout) {
                ref.current?.addEventListener('focusout', focusProps.onfocusout, false);
            }
        }
        else {
            if (focusProps.onfocusin) {
                ref.current?.removeEventListener('focusin', focusProps.onfocusin, false);
            }
            if (focusProps.onfocusout) {
                ref.current?.removeEventListener('focusout', focusProps.onfocusout, false);
            }
        }
    }, [focusProps.onfocusin, focusProps.onfocusout, ref]);
    return {
        isFocused,
        popupProps: {
            onTransitionEnd,
            ref
        }
    };
}

function CompactHelpSource({ children, id, labelEdge, source }) {
    const [isOver, setOver] = useState(false);
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const userAssistance_learnMoreStr = translations.userAssistance_learnMore();
    const { isTabbable, tabbableModeProps } = useTabbableMode();
    const flip = { mainAxis: true, crossAxis: false };
    // TODO: Implement and use a preact Link component instead of using an anchor tag
    const anchorRef = useRef(null);
    const text = children ?? userAssistance_learnMoreStr;
    // helpIconBase references 'Label' vars, so we need to get the baseTheme for 'Label' and include it in the class list
    // so that we pick up the definitions for the labelVars we are using.
    const { baseTheme: labelBaseTheme } = useComponentTheme(LabelRedwoodTheme);
    const styles = classNames([
        labelBaseTheme,
        helpIconBase,
        labelEdge === 'start' && helpIconLabelEdgeStart
    ]);
    const { isFocused: isPopupFocused, popupProps } = usePopupFocusWithin();
    const isOpen = isOver === true || isPopupFocused === true;
    return (jsxs(Fragment, { children: [jsx("a", { target: "_blank", ...(source && { href: source }), "aria-label": text, class: styles, id: id, ref: anchorRef, role: "link", onMouseEnter: () => setOver(true), onMouseLeave: () => setOver(false), onFocus: () => setOver(true), onBlur: () => setOver(false), ...(!isTabbable ? tabbableModeProps : { tabIndex: 0 }), children: jsx(SvgHelp, {}) }), jsx(Popup, { ...popupProps, placement: "end-top-corner", isOpen: isOpen, anchorRef: anchorRef, flipOptions: flip, children: jsx("div", { class: helpContentStyles, children: text }) })] }));
}

export { CompactHelpSource as C, usePopupFocusWithin as u };
