// @oracle/oraclejet-preact/es/Flex-46adb0e4.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx } from 'preact/jsx-runtime';
import { dimensionInterpolations } from './utils/UNSAFE_interpolations/dimensions.js';
import { mergeInterpolations } from './utils/UNSAFE_mergeInterpolations.js';

import { boxAlignmentInterpolations } from './utils/UNSAFE_interpolations/boxalignment.js';
import './utils/UNSAFE_interpolations/flexbox.js';
import './utils/UNSAFE_interpolations/flexitem.js';
import './utils/UNSAFE_classNames.js';
import { f as flexboxInterpolations } from './flexbox-645317db.js';
import { f as flexitemInterpolations } from './flexitem-89d8a00a.js';
import { c as classNames } from './classNames-f414daad.js';

const baseStyles = "_1yl1w5m";
// exported for testing
// TODO: This will be removed when converting the Flex commponent to the new theming,
// and there the marginZeroStyles class name will be exported in the FlexStyles.css.ts file
// and can be used for testing.
const marginZeroStyles = "c6r810";
const interpolations = [...Object.values(dimensionInterpolations), ...Object.values(flexboxInterpolations), ...Object.values(flexitemInterpolations), ...Object.values(boxAlignmentInterpolations)];
const styleInterpolations = mergeInterpolations(interpolations);
const Flex = ({
  children,
  hasZeroMargins = true,
  ...props
}) => {
  const flexComponentClasses = classNames([baseStyles, hasZeroMargins && marginZeroStyles]);
  const {
    class: cls,
    ...styles
  } = styleInterpolations(props);
  return jsx("div", {
    class: `${flexComponentClasses} ${cls}`,
    style: styles,
    children: children
  });
};

export { Flex as F, marginZeroStyles as m };
