// @oracle/oraclejet-preact/es/HighlightText-1591802b.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx } from 'preact/jsx-runtime';
import { Fragment } from 'preact';


const highlightStyles = "_1k6j1aj";
/**
 * Renders a text string with highlighting applied to the given text to match.
 */
function HighlightText({
  children = '',
  matchText = ''
}) {
  return highlighter(children, matchText);
}
const HIGHLIGHT_TOKEN = '__@@__';
const HIGHLIGHT_PATTERN = `${HIGHLIGHT_TOKEN}$&${HIGHLIGHT_TOKEN}`;
function highlighter(unhighlightedText, matchText) {
  if (matchText.length > 0 && unhighlightedText.length > 0) {
    const escapedMatchText = escapeRegExp(matchText);
    const highlightedText = unhighlightedText.replace(new RegExp(escapedMatchText, 'gi'), HIGHLIGHT_PATTERN);
    const tokens = highlightedText.split(HIGHLIGHT_TOKEN);
    const nodes = tokens.map((current, index) => index % 2 == 0 ? current : jsx("span", {
      class: highlightStyles,
      children: current
    }));
    return jsx(Fragment, {
      children: nodes
    });
  }
  return jsx(Fragment, {
    children: unhighlightedText
  });
}
function escapeRegExp(str) {
  // copied from:
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping
  return str.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
}

export { HighlightText as H };
