// @oracle/oraclejet-preact/es/listViewSelectionTests-90878733.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx } from 'preact/jsx-runtime';
import { render, fireEvent, waitFor } from '@testing-library/preact';
import { expect } from 'chai';
import { R as RootEnvironmentProvider } from './index-f26154f2.js';
import { matchTranslationBundle } from './utils/UNSAFE_matchTranslationBundle.js';
import { CollectionGestureContext } from './hooks/PRIVATE_useCollectionGestureContext.js';
import { p as populateData } from './listViewTestFixtures-f603dff6.js';
import { a as getItems, g as getRoot, b as getContents, m as mouseItemSelection, c as mouseCheckboxSelection } from './testHelpers-1b2a7fbd.js';

function __variableDynamicImportRuntime0__$1(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }
const testRender = (TestComponent) => {
    it('initial render', async function () {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata }));
        const items = await getItems(list);
        expect(items.length).eq(25);
    });
    // skip this test in jest, only run it in karma mocha
    if (typeof jest == typeof undefined) {
        it('initial render with sparkle', async () => {
            const data = populateData(100);
            let dataMetadata = data.map((data) => {
                return { data: data, metadata: { key: data.id } };
            });
            // add suggestions
            dataMetadata[3].metadata.suggestion = { enable: true };
            dataMetadata[7].metadata.suggestion = { enable: true };
            dataMetadata[9].metadata.suggestion = { enable: true };
            const suggestions = dataMetadata.filter((d) => d.metadata.suggestion);
            dataMetadata = dataMetadata.filter((d) => !d.metadata.suggestion);
            dataMetadata.unshift(...suggestions);
            const list = render(jsx(TestComponent, { data: dataMetadata }));
            const grid = await getRoot(list);
            const sparkle = grid.querySelector('.oj-collection-sparkle')
                ?.firstElementChild;
            expect(sparkle !== undefined).eq(true);
            const items = await getItems(list);
            expect(items?.length).eq(25);
            expect(items[0].dataset['ojSuggestion']).eq('true');
            expect(items[1].dataset['ojSuggestion']).eq('true');
            expect(items[2].dataset['ojSuggestion']).eq('true');
            const sparkleHeight = items[0].offsetHeight + items[1].offsetHeight + items[2].offsetHeight;
            expect(sparkleHeight).eq(sparkle.offsetHeight);
        });
    }
};
const testAria = (TestComponent) => {
    it('aria structure and properties (select-none)', async function () {
        const data = populateData(10);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata }));
        const grid = await getRoot(list);
        expect(grid.getAttribute('aria-rowcount')).eq('10');
        expect(grid.getAttribute('aria-multiselectable')).eq(null);
        const items = await getItems(list);
        expect(items.length).eq(10);
        expect(items[0].getAttribute('aria-rowindex')).eq('1');
        const cells = await getContents(list);
        expect(cells.length).eq(10);
        expect(cells[0].getAttribute('aria-colindex')).eq('1');
        expect(cells[0].getAttribute('aria-selected')).eq(null);
    });
    it('aria structure and properties (single-select)', async function () {
        const data = populateData(10);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata, selectionMode: "single" }));
        const grid = await getRoot(list);
        expect(grid.getAttribute('aria-rowcount')).eq('10');
        expect(grid.getAttribute('aria-multiselectable')).eq('false');
        const items = await getItems(list);
        expect(items.length).eq(10);
        expect(items[0].getAttribute('aria-rowindex')).eq('1');
        const cells = await getContents(list);
        expect(cells.length).eq(10);
        expect(cells[0].getAttribute('aria-colindex')).eq('1');
        expect(cells[0].getAttribute('aria-selected')).eq('false');
    });
    it('aria structure and properties (multi-select)', async function () {
        const data = populateData(10);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime0__$1(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const grid = await getRoot(list);
        expect(grid.getAttribute('aria-rowcount')).eq('10');
        expect(grid.getAttribute('aria-multiselectable')).eq('true');
        const items = await getItems(list);
        expect(items.length).eq(10);
        expect(items[0].getAttribute('aria-rowindex')).eq('1');
        const cells = await getContents(list);
        expect(cells.length).eq(10);
        expect(cells[0].getAttribute('aria-colindex')).eq('1');
        expect(cells[0].getAttribute('aria-selected')).eq('false');
    });
};
const testKeys = (TestComponent) => {
    it('supports currentKey logic with numeric key', async () => {
        const data = populateData(100, true);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata }));
        const grid = await getRoot(list);
        expect(grid.hasAttribute('aria-activedescendant')).eq(false);
        const items = await getContents(list);
        fireEvent.click(items[0]);
        expect(grid.getAttribute('aria-activedescendant')).eq(items[0].id);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[1].id);
        fireEvent.keyDown(grid, { key: 'ArrowUp' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[0].id);
    });
    it('supports currentKey logic with string key', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata }));
        const grid = await getRoot(list);
        expect(grid.hasAttribute('aria-activedescendant')).eq(false);
        const items = await getContents(list);
        fireEvent.click(items[0]);
        expect(grid.getAttribute('aria-activedescendant')).eq(items[0].id);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[1].id);
        fireEvent.keyDown(grid, { key: 'ArrowUp' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[0].id);
    });
    it('checking key type - string key', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata }));
        const items = await getItems(list);
        expect(items[0].dataset['ojKeyType'] === undefined).eq(true);
        expect(items[1].dataset['ojKeyType'] === undefined).eq(true);
        expect(items[2].dataset['ojKeyType'] === undefined).eq(true);
    });
    it('checking key type - numeric key', async () => {
        const data = populateData(3, true);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata }));
        const items = await getItems(list);
        expect(items[0].dataset['ojKeyType']).eq('number');
        expect(items[1].dataset['ojKeyType']).eq('number');
        expect(items[2].dataset['ojKeyType']).eq('number');
    });
};
const testGridlines = (TestComponent) => {
    const isTopGridlineVisible = (item) => {
        const className = item.className;
        return className.indexOf('gridlineTop_visible') > 0;
    };
    const isBottomGridlineVisible = (item) => {
        const className = item.className;
        return className.indexOf('gridlineBottom_visible') > 0;
    };
    const isGridlineHidden = (item) => {
        return !isTopGridlineVisible(item) && !isBottomGridlineVisible(item);
    };
    it('gridlines - gridlines are hidden', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata }));
        const items = await getItems(list);
        expect(isGridlineHidden(items[0])).to.be.true;
        expect(isGridlineHidden(items[1])).to.be.true;
        expect(isGridlineHidden(items[2])).to.be.true;
    });
    it('gridlines - bottom gridlines are visible', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const visible = { item: 'visible', bottom: 'visible' };
        const list = render(jsx(TestComponent, { data: dataMetadata, gridlines: visible }));
        const items = await getItems(list);
        expect(isBottomGridlineVisible(items[0])).to.be.true;
        expect(isBottomGridlineVisible(items[1])).to.be.true;
        expect(isBottomGridlineVisible(items[2])).to.be.true;
    });
    it('gridlines - bottom gridlines are visible except for the last list item', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const visibleExceptLast = { item: 'visible' };
        const list = render(jsx(TestComponent, { data: dataMetadata, gridlines: visibleExceptLast }));
        const items = await getItems(list);
        expect(isBottomGridlineVisible(items[0])).to.be.true;
        expect(isBottomGridlineVisible(items[1])).to.be.true;
        expect(isBottomGridlineVisible(items[2])).to.be.false;
    });
    it('gridlines - bottom gridlines are visible; The top gridlines of the first list item is visible', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const visibleWithFirst = {
            item: 'visible',
            top: 'visible',
            bottom: 'visible'
        };
        const list = render(jsx(TestComponent, { data: dataMetadata, gridlines: visibleWithFirst }));
        const items = await getItems(list);
        expect(isTopGridlineVisible(items[0])).to.be.true;
        expect(isBottomGridlineVisible(items[0])).to.be.true;
        expect(isBottomGridlineVisible(items[1])).to.be.true;
        expect(isBottomGridlineVisible(items[2])).to.be.true;
    });
    it('gridlines - bottom gridlines are visible except for the last list item. The top gridlines of the first list item is visible.', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const visibleWithFirstButNotLast = { item: 'visible', top: 'visible' };
        const list = render(jsx(TestComponent, { data: dataMetadata, gridlines: visibleWithFirstButNotLast }));
        const items = await getItems(list);
        expect(isTopGridlineVisible(items[0])).to.be.true;
        expect(isBottomGridlineVisible(items[0])).to.be.true;
        expect(isBottomGridlineVisible(items[1])).to.be.true;
        expect(isBottomGridlineVisible(items[2])).to.be.false;
    });
};
const testItemAction = (TestComponent) => {
    const isMetadataEqual = (metadata1, metadata2) => {
        return metadata1.key === metadata2.key;
    };
    const isFocusHighlight = (item) => {
        const className = item.className;
        return (className.indexOf('focusHighlight_isFocusHighlight') > 0 ||
            className.indexOf('MultiVariantStyles_compound') > 0);
    };
    it('ItemAction - test click on item', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        let count = 0;
        const onItemAction = (detail) => {
            if (count === 0) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[1].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[1].metadata), 'Context metadata is correct').eq(true);
            }
            else if (count === 1) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[1].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[1].metadata), 'Context metadata is correct').eq(true);
            }
            else if (count === 2) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[2].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[2].metadata), 'Context metadata is correct').eq(true);
            }
            count++;
        };
        const list = render(jsx(TestComponent, { data: dataMetadata, onItemAction: onItemAction }));
        const items = await getItems(list, 'row');
        fireEvent.click(items[1]);
        fireEvent.click(items[1]);
        fireEvent.click(items[2]);
        expect(count).eq(3);
    });
    it('ItemAction - test press enter on item', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        let count = 0;
        const onItemAction = (detail) => {
            if (count === 1) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[1].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[1].metadata), 'Context metadata is correct').eq(true);
            }
            else if (count === 2) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[1].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[1].metadata), 'Context metadata is correct').eq(true);
            }
            else if (count === 3) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[2].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[2].metadata), 'Context metadata is correct').eq(true);
            }
            count++;
        };
        const list = render(jsx(TestComponent, { data: dataMetadata, onItemAction: onItemAction }));
        const grid = await getRoot(list);
        const items = await getItems(list, 'row');
        fireEvent.click(items[0]);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        fireEvent.keyDown(items[1], { key: 'Enter' });
        fireEvent.keyDown(items[1], { key: 'Enter' });
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        fireEvent.keyDown(items[2], { key: 'Enter' });
        expect(count).eq(4);
    });
    it('ItemAction - test press spacebar on item (Context value is embedded)', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        let count = 0;
        const onItemAction = (detail) => {
            if (count === 1) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[1].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[1].metadata), 'Context metadata is correct').eq(true);
            }
            else if (count === 2) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[1].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[1].metadata), 'Context metadata is correct').eq(true);
            }
            else if (count === 3) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[2].data);
                expect(isMetadataEqual(detail.context.metadata, dataMetadata[2].metadata), 'Context metadata is correct').eq(true);
            }
            count++;
        };
        const list = render(jsx(CollectionGestureContext.Provider, { value: 'embedded', children: jsx(TestComponent, { data: dataMetadata, onItemAction: onItemAction }) }));
        const grid = await getRoot(list);
        const items = await getItems(list, 'row');
        fireEvent.click(items[0]);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        fireEvent.keyDown(items[1], { key: ' ' });
        fireEvent.keyDown(items[1], { key: ' ' });
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        fireEvent.keyDown(items[2], { key: ' ' });
        expect(count).eq(4);
    });
    it('ItemAction - test press spacebar on item (Context value is default)', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        let count = 0;
        const onItemAction = (detail) => {
            if (count === 1) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[1].data);
                expect(detail.context.metadata, 'Context metadata is correct').eq(dataMetadata[1].metadata);
            }
            else if (count === 2) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[1].data);
                expect(detail.context.metadata, 'Context metadata is correct').eq(dataMetadata[1].metadata);
            }
            else if (count === 3) {
                expect(detail.context.data, 'Context data is correct').eq(dataMetadata[2].data);
                expect(detail.context.metadata, 'Context metadata is correct').eq(dataMetadata[2].metadata);
            }
            count++;
        };
        const list = render(jsx(CollectionGestureContext.Provider, { value: 'default', children: jsx(TestComponent, { data: dataMetadata, onItemAction: onItemAction }) }));
        const grid = await getRoot(list);
        const items = await getItems(list, 'row');
        fireEvent.click(items[0]);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        fireEvent.keyDown(items[1], { key: ' ' });
        fireEvent.keyDown(items[1], { key: ' ' });
        expect(count).eq(1);
    });
    it('currentItemVariant - highlight', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata, initialKey: dataMetadata[1].metadata.key, currentItemVariant: "highlight" }));
        const items = await getItems(list);
        expect(isFocusHighlight(items[0])).to.be.false;
        fireEvent.click(items[0]);
        expect(isFocusHighlight(items[0])).to.be.true;
    });
    it('currentItemVariant - none', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata, initialKey: dataMetadata[1].metadata.key, currentItemVariant: "none" }));
        const items = await getItems(list);
        expect(isFocusHighlight(items[0])).to.be.false;
        fireEvent.click(items[0]);
        expect(isFocusHighlight(items[0])).to.be.false;
    });
};
const testNavigation = (TestComponent) => {
    it('arrow keys', async () => {
        const data = populateData(3);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata }));
        const grid = await getRoot(list);
        expect(grid.hasAttribute('aria-activedescendant')).eq(false);
        const items = await getContents(list);
        fireEvent.click(items[0]);
        expect(grid.getAttribute('aria-activedescendant')).eq(items[0].id);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[1].id);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[2].id);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[2].id);
        fireEvent.keyDown(grid, { key: 'ArrowUp' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[1].id);
        fireEvent.keyDown(grid, { key: 'ArrowUp' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[0].id);
        fireEvent.keyDown(grid, { key: 'ArrowUp' });
        expect(grid.getAttribute('aria-activedescendant')).eq(items[0].id);
    });
};

function __variableDynamicImportRuntime11__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime10__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime9__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime8__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime7__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime6__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime5__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime4__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime3__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime2__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime1__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }

function __variableDynamicImportRuntime0__(path) {
  switch (path) {
    case '../../../resources/nls/ar/bundle.ts': return import('./resources/nls/ar/bundle.js');
    case '../../../resources/nls/ar-XB/bundle.ts': return import('./resources/nls/ar-XB/bundle.js');
    case '../../../resources/nls/bg/bundle.ts': return import('./resources/nls/bg/bundle.js');
    case '../../../resources/nls/bs/bundle.ts': return import('./resources/nls/bs/bundle.js');
    case '../../../resources/nls/bs-Cyrl/bundle.ts': return import('./resources/nls/bs-Cyrl/bundle.js');
    case '../../../resources/nls/cs/bundle.ts': return import('./resources/nls/cs/bundle.js');
    case '../../../resources/nls/da/bundle.ts': return import('./resources/nls/da/bundle.js');
    case '../../../resources/nls/de/bundle.ts': return import('./resources/nls/de/bundle.js');
    case '../../../resources/nls/el/bundle.ts': return import('./resources/nls/el/bundle.js');
    case '../../../resources/nls/en/bundle.ts': return import('./resources/nls/en/bundle.js');
    case '../../../resources/nls/en-XA/bundle.ts': return import('./resources/nls/en-XA/bundle.js');
    case '../../../resources/nls/en-XC/bundle.ts': return import('./resources/nls/en-XC/bundle.js');
    case '../../../resources/nls/es/bundle.ts': return import('./resources/nls/es/bundle.js');
    case '../../../resources/nls/et/bundle.ts': return import('./resources/nls/et/bundle.js');
    case '../../../resources/nls/fi/bundle.ts': return import('./resources/nls/fi/bundle.js');
    case '../../../resources/nls/fr/bundle.ts': return import('./resources/nls/fr/bundle.js');
    case '../../../resources/nls/fr-CA/bundle.ts': return import('./resources/nls/fr-CA/bundle.js');
    case '../../../resources/nls/he/bundle.ts': return import('./resources/nls/he/bundle.js');
    case '../../../resources/nls/hr/bundle.ts': return import('./resources/nls/hr/bundle.js');
    case '../../../resources/nls/hu/bundle.ts': return import('./resources/nls/hu/bundle.js');
    case '../../../resources/nls/is/bundle.ts': return import('./resources/nls/is/bundle.js');
    case '../../../resources/nls/it/bundle.ts': return import('./resources/nls/it/bundle.js');
    case '../../../resources/nls/ja/bundle.ts': return import('./resources/nls/ja/bundle.js');
    case '../../../resources/nls/ko/bundle.ts': return import('./resources/nls/ko/bundle.js');
    case '../../../resources/nls/lt/bundle.ts': return import('./resources/nls/lt/bundle.js');
    case '../../../resources/nls/lv/bundle.ts': return import('./resources/nls/lv/bundle.js');
    case '../../../resources/nls/ms/bundle.ts': return import('./resources/nls/ms/bundle.js');
    case '../../../resources/nls/nl/bundle.ts': return import('./resources/nls/nl/bundle.js');
    case '../../../resources/nls/no/bundle.ts': return import('./resources/nls/no/bundle.js');
    case '../../../resources/nls/pl/bundle.ts': return import('./resources/nls/pl/bundle.js');
    case '../../../resources/nls/pt/bundle.ts': return import('./resources/nls/pt/bundle.js');
    case '../../../resources/nls/pt-PT/bundle.ts': return import('./resources/nls/pt-PT/bundle.js');
    case '../../../resources/nls/ro/bundle.ts': return import('./resources/nls/ro/bundle.js');
    case '../../../resources/nls/ru/bundle.ts': return import('./resources/nls/ru/bundle.js');
    case '../../../resources/nls/sk/bundle.ts': return import('./resources/nls/sk/bundle.js');
    case '../../../resources/nls/sl/bundle.ts': return import('./resources/nls/sl/bundle.js');
    case '../../../resources/nls/sr/bundle.ts': return import('./resources/nls/sr/bundle.js');
    case '../../../resources/nls/sr-Latn/bundle.ts': return import('./resources/nls/sr-Latn/bundle.js');
    case '../../../resources/nls/sv/bundle.ts': return import('./resources/nls/sv/bundle.js');
    case '../../../resources/nls/th/bundle.ts': return import('./resources/nls/th/bundle.js');
    case '../../../resources/nls/tr/bundle.ts': return import('./resources/nls/tr/bundle.js');
    case '../../../resources/nls/uk/bundle.ts': return import('./resources/nls/uk/bundle.js');
    case '../../../resources/nls/vi/bundle.ts': return import('./resources/nls/vi/bundle.js');
    case '../../../resources/nls/zh-Hans/bundle.ts': return import('./resources/nls/zh-Hans/bundle.js');
    case '../../../resources/nls/zh-Hant/bundle.ts': return import('./resources/nls/zh-Hant/bundle.js');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }
const testSelection = (TestComponent) => {
    it('select single - numeric key', async () => {
        const data = populateData(10, true);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const list = render(jsx(TestComponent, { data: dataMetadata, selectionMode: "single", onClick: () => {
                expect(false, 'Click event should not propagate').to.be.true;
            } }));
        const items = await getContents(list);
        mouseItemSelection(items[3]);
        await waitFor(() => {
            expect(items[3].getAttribute('aria-selected')).eq('true');
        });
        mouseItemSelection(items[5]);
        await waitFor(() => {
            expect(items[5].getAttribute('aria-selected')).eq('true');
        });
        mouseItemSelection(items[7]);
        await waitFor(() => {
            expect(items[7].getAttribute('aria-selected')).eq('true');
        });
    });
    it('select multiple items using shift click', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime0__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const items = await getContents(list);
        mouseItemSelection(items[2]);
        mouseItemSelection(items[5], { shiftKey: true });
        // console.log('here');
        // console.log(list.debug());
        await waitFor(() => {
            expect(items[1].getAttribute('aria-selected')).eq('false');
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[3].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('true');
            expect(items[5].getAttribute('aria-selected')).eq('true');
            expect(items[6].getAttribute('aria-selected')).eq('false');
        });
    }, 3000);
    it('select multiple items using shift space bar', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime1__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const items = await getContents(list);
        fireEvent.keyUp(items[2], { key: ' ' });
        fireEvent.keyUp(items[5], { key: ' ', shiftKey: true });
        expect(items[1].getAttribute('aria-selected')).eq('false');
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('true');
        expect(items[4].getAttribute('aria-selected')).eq('true');
        expect(items[5].getAttribute('aria-selected')).eq('true');
        expect(items[6].getAttribute('aria-selected')).eq('false');
    });
    it('selects multiple items when clicking on Selector checkboxes', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const selected = { all: false, keys: new Set([data[2].id]) };
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime2__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple", selectedKeys: selected }) }));
        const items = await getContents(list);
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[4].getAttribute('aria-selected')).eq('false');
        mouseCheckboxSelection(items[4]);
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('true');
        });
    });
    it('correctly modifies range selection when using SPACE to select', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime3__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const items = await getContents(list);
        fireEvent.keyUp(items[2], { key: ' ' });
        fireEvent.keyUp(items[5], { key: ' ', shiftKey: true });
        expect(items[1].getAttribute('aria-selected')).eq('false');
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('true');
        expect(items[4].getAttribute('aria-selected')).eq('true');
        expect(items[5].getAttribute('aria-selected')).eq('true');
        expect(items[6].getAttribute('aria-selected')).eq('false');
        // keyboard select to "toggle" one item off
        fireEvent.keyUp(items[3], { key: ' ' });
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('false');
        expect(items[4].getAttribute('aria-selected')).eq('true');
        // keyboard select again to "replace" with single selection only
        fireEvent.keyUp(items[3], { key: ' ' });
        expect(items[2].getAttribute('aria-selected')).eq('false');
        expect(items[3].getAttribute('aria-selected')).eq('true');
        expect(items[4].getAttribute('aria-selected')).eq('false');
    });
    it('selects single items using space keypress', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime4__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const items = await getContents(list);
        fireEvent.keyUp(items[2], { key: ' ' });
        expect(items[1].getAttribute('aria-selected')).eq('false');
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('false');
        fireEvent.keyUp(items[3], { key: ' ' });
        expect(items[1].getAttribute('aria-selected')).eq('false');
        expect(items[2].getAttribute('aria-selected')).eq('false');
        expect(items[3].getAttribute('aria-selected')).eq('true');
    }, 3000);
    it('selects single items using enter keypress (Context value is embedded)', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime5__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(CollectionGestureContext.Provider, { value: 'embedded', children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }) }));
        const items = await getContents(list);
        fireEvent.keyUp(items[2], { key: 'Enter' });
        expect(items[1].getAttribute('aria-selected')).eq('false');
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('false');
        fireEvent.keyUp(items[3], { key: 'Enter' });
        expect(items[1].getAttribute('aria-selected')).eq('false');
        expect(items[2].getAttribute('aria-selected')).eq('false');
        expect(items[3].getAttribute('aria-selected')).eq('true');
    });
    it('selects single items using enter keypress (Context value is default)', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime6__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(CollectionGestureContext.Provider, { value: 'default', children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }) }));
        const items = await getContents(list);
        fireEvent.keyUp(items[2], { key: 'Enter' });
        expect(items[1].getAttribute('aria-selected')).eq('false');
        expect(items[2].getAttribute('aria-selected')).eq('false');
        expect(items[3].getAttribute('aria-selected')).eq('false');
    });
    it('select multiple items then click on a selected items', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime7__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const items = await getContents(list);
        mouseItemSelection(items[2]);
        mouseItemSelection(items[5], { shiftKey: true });
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[3].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('true');
            expect(items[5].getAttribute('aria-selected')).eq('true');
        });
        mouseItemSelection(items[3]);
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('false');
            expect(items[3].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('false');
            expect(items[5].getAttribute('aria-selected')).eq('false');
        });
    });
    it('select multiple items then shift click on a selected items', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime8__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const items = await getContents(list);
        mouseItemSelection(items[2]);
        mouseItemSelection(items[5], { shiftKey: true });
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[3].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('true');
            expect(items[5].getAttribute('aria-selected')).eq('true');
        });
        mouseItemSelection(items[3], { shiftKey: true });
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[3].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('false');
            expect(items[5].getAttribute('aria-selected')).eq('false');
        });
    });
    it('select multiple items then metaKey click on a selected items, to select and unselect', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime9__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const items = await getContents(list);
        mouseItemSelection(items[2]);
        mouseItemSelection(items[5], { shiftKey: true });
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[3].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('true');
            expect(items[5].getAttribute('aria-selected')).eq('true');
        });
        // first click to deselect
        mouseItemSelection(items[3], { metaKey: true });
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[3].getAttribute('aria-selected')).eq('false');
            expect(items[4].getAttribute('aria-selected')).eq('true');
            expect(items[5].getAttribute('aria-selected')).eq('true');
        });
        // second click to reselect
        mouseItemSelection(items[3], { metaKey: true });
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[3].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('true');
            expect(items[5].getAttribute('aria-selected')).eq('true');
        });
        // click outside of range should be additive too
        mouseItemSelection(items[7], { metaKey: true });
        await waitFor(() => {
            expect(items[2].getAttribute('aria-selected')).eq('true');
            expect(items[3].getAttribute('aria-selected')).eq('true');
            expect(items[4].getAttribute('aria-selected')).eq('true');
            expect(items[5].getAttribute('aria-selected')).eq('true');
            expect(items[6].getAttribute('aria-selected')).eq('false');
            expect(items[7].getAttribute('aria-selected')).eq('true');
        });
    }, 3000);
    it('supports multi-select (and deselect) via keyboard arrow keys', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime10__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const grid = await getRoot(list);
        const items = await getContents(list);
        // select items 1-3
        fireEvent.click(items[0]);
        fireEvent.keyDown(grid, { key: 'ArrowDown' });
        fireEvent.keyDown(grid, { key: 'ArrowDown', shiftKey: true });
        fireEvent.keyDown(grid, { key: 'ArrowDown', shiftKey: true });
        expect(items[0].getAttribute('aria-selected')).eq('false');
        expect(items[1].getAttribute('aria-selected')).eq('true');
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('true');
        // deselect 2 and 3 via up-arrow
        fireEvent.keyDown(grid, { key: 'ArrowUp', shiftKey: true });
        fireEvent.keyDown(grid, { key: 'ArrowUp', shiftKey: true });
        expect(items[0].getAttribute('aria-selected')).eq('false');
        expect(items[1].getAttribute('aria-selected')).eq('true');
        expect(items[2].getAttribute('aria-selected')).eq('false');
        expect(items[3].getAttribute('aria-selected')).eq('false');
        // continue up past initial anchor, to select items 0 and 1
        fireEvent.keyDown(grid, { key: 'ArrowUp', shiftKey: true });
        fireEvent.keyDown(grid, { key: 'ArrowUp', shiftKey: true });
        expect(items[0].getAttribute('aria-selected')).eq('true');
        expect(items[1].getAttribute('aria-selected')).eq('true');
        expect(items[2].getAttribute('aria-selected')).eq('false');
        expect(items[3].getAttribute('aria-selected')).eq('false');
    });
    it('supports extending mouse initiated range select via arrow keys', async () => {
        const data = populateData(100);
        const dataMetadata = data.map((data) => {
            return { data: data, metadata: { key: data.id } };
        });
        const locale = matchTranslationBundle(['en'], new Set(['en-US', 'en']));
        const { default: translations } = await __variableDynamicImportRuntime11__(`../../../resources/nls/${locale}/bundle.ts`);
        const env = {
            translations: { '@oracle/oraclejet-preact': translations }
        };
        const list = render(jsx(RootEnvironmentProvider, { environment: env, children: jsx(TestComponent, { data: dataMetadata, selectionMode: "multiple" }) }));
        const grid = await getRoot(list);
        const items = await getContents(list);
        // select items 2-5 via click
        mouseItemSelection(items[2]);
        mouseItemSelection(items[5], { shiftKey: true });
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('true');
        expect(items[4].getAttribute('aria-selected')).eq('true');
        expect(items[5].getAttribute('aria-selected')).eq('true');
        expect(items[6].getAttribute('aria-selected')).eq('false');
        expect(items[7].getAttribute('aria-selected')).eq('false');
        // now extend by two more rows
        fireEvent.keyDown(grid, { key: 'ArrowDown', shiftKey: true });
        fireEvent.keyDown(grid, { key: 'ArrowDown', shiftKey: true });
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('true');
        expect(items[4].getAttribute('aria-selected')).eq('true');
        expect(items[5].getAttribute('aria-selected')).eq('true');
        expect(items[6].getAttribute('aria-selected')).eq('true');
        expect(items[7].getAttribute('aria-selected')).eq('true');
        // now deselect those via up-arrow
        fireEvent.keyDown(grid, { key: 'ArrowUp', shiftKey: true });
        fireEvent.keyDown(grid, { key: 'ArrowUp', shiftKey: true });
        expect(items[2].getAttribute('aria-selected')).eq('true');
        expect(items[3].getAttribute('aria-selected')).eq('true');
        expect(items[4].getAttribute('aria-selected')).eq('true');
        expect(items[5].getAttribute('aria-selected')).eq('true');
        expect(items[6].getAttribute('aria-selected')).eq('false');
        expect(items[7].getAttribute('aria-selected')).eq('false');
    });
};

export { testAria as a, testKeys as b, testGridlines as c, testItemAction as d, testNavigation as e, testSelection as f, testRender as t };
