// @oracle/oraclejet-preact/es/menuTestsUtils-9c391e65.js
/* @oracle/oraclejet-preact: 15.1.2 */
function getIcon(menuItem, selector) {
    return menuItem.querySelector(selector);
}
function getAllMenuItemsByRole(content, role, hidden = true) {
    return content.getAllByRole(role, { hidden });
}
function getMenuItemByRole(content, role, hidden = true) {
    return content.getByRole(role, { hidden });
}
function getLabel(content, label) {
    return content.getByText(label);
}
function setObserverConfig() {
    if (!window.ResizeObserver) {
        // ResizeObserver isn't available in test environment
        const MockResizeObserver = jest.fn();
        MockResizeObserver.mockReturnValue({
            observe: () => null,
            unobserve: () => null,
            disconnect: () => null
        });
        Object.defineProperty(global, 'ResizeObserver', {
            writable: true,
            configurable: true,
            value: MockResizeObserver
        });
    }
}

export { getAllMenuItemsByRole as a, getLabel as b, getIcon as c, getMenuItemByRole as g, setObserverConfig as s };
