// @oracle/oraclejet-preact/es/Menu-31a689f5.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx, jsxs, Fragment } from 'preact/jsx-runtime';
import { forwardRef, useCallback as useCallback$1 } from 'preact/compat';
import './UNSAFE_Floating.js';
import { Layer } from './UNSAFE_Layer.js';
import { useMemo, useCallback, useRef, useState, useEffect } from 'preact/hooks';
import { mergeProps } from './utils/UNSAFE_mergeProps.js';
import './utils/UNSAFE_classNames.js';
import { useUser } from './hooks/UNSAFE_useUser.js';
import { u as useMenuContext, k as keyElemExtractor, g as getPrevNext, a as getNVisible, b as getLastVisible, c as getKeyboardCloseProps, d as getCloseSubmenuEvent, e as getMenuContainerFocusMethods, f as getMenuContainerMouseLeave, h as getMenuPointerDown, i as MenuContainerContext, j as getIsMobile, l as MenuContext } from './MenuItem-151ce959.js';
import { menuStyles } from './UNSAFE_Menu/themes/MenuStyles.css.js';
import { c as classNames } from './classNames-f414daad.js';
import { dropdownMenuStyles } from './UNSAFE_Menu/themes/DropdownMenuStyles.css.js';
import { F as Floating } from './Floating-2d909840.js';
import { u as useModal } from './useModal-e580c081.js';
import { sheetMenuStyles } from './UNSAFE_Menu/themes/SheetMenuStyles.css.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Hook for handling current key update due to user interaction including keyboard navigation.
 *
 * @param getPrev function to get the previous key based on the current key
 * @param getNext function to get the next key based on the current key
 * @param getFirstVisible function to get the key of the first available item
 * @param getLastVisible function to get the key of the last available item
 * @param onChange function to invoke if the current key has changed
 * @returns
 */
function useMenuCurrentKey(getPrev, getNext, getFirstVisible, getLastVisible, onChange) {
    const keyDownKeyFunctions = useMemo(() => {
        return {
            ArrowDown: getNext,
            ArrowUp: getPrev,
            Home: getFirstVisible,
            End: getLastVisible
        };
    }, [getNext, getPrev, getFirstVisible, getLastVisible]);
    const onKeyDown = useCallback((event) => {
        if (onChange) {
            // TODO: support horizontal layout on icon menu (left/right arrow key)
            const keyFunction = keyDownKeyFunctions[event.key];
            if (keyFunction && !event.shiftKey) {
                const key = keyFunction();
                if (key) {
                    onChange({ key, reason: 'keyboard' });
                }
                event.preventDefault();
            }
        }
    }, [onChange, keyDownKeyFunctions]);
    const currentKeyProps = onChange == null ? {} : { onKeyDown };
    return { currentKeyProps };
}

/*Component where general behavior is handled. Here we controled focus behavior, focus styling, submenu
opening behavior and current active item behavior. */
const MenuContainer = ({ children, 'aria-label': ariaLabel, currentFocus = 'menu', display, closeSubmenu, menuId, menuLevel }) => {
    const { onClose } = useMenuContext();
    //Important to know direction so closeSubmenu is handled properly if needed.
    const { direction } = useUser();
    const menuContainerRef = useRef(null);
    /*We share same behavior as legacy menu on mouseLeave that is:
    1. If we mouse leave a menuItem but we still on menuContainer and we are not hovering any other menuItem, we keep
    the focus on the menuItem and hovering style.
    2. If we mouseLeave but next target is a submenu we open a submenu.
    3. If we mouseLeave the container and next target is not a submenu, we reset current key and set the focus on the
    menuContainer.If we mouseLeaved a submenuItem, submenu is closed.
    General information used by menuItems:
    -wasMouseMoved: Used to don't allow mouseEnter callbacks being triggered if mouse was not moved already.
    -lastMouseLeaveValidatedRelatedTarget: We use bubbling nature of events to first check/validate the related target in onMouseLeave events of
      menuItems. For menuItems we check if the related target is the menuContainer itself or a submenu. We validate there
      since at this level we don't have access to any submenu. Everytime mouseEnter triggers at menuContainer we have to compare if both
      relatedTargets are the same(menuContainer and last menuItem one), if not, this means mouse is now outside of the container.
    */
    const menuContainerGeneralInformation = useRef({
        wasMouseMoved: false,
        lastMouseLeaveMenuItemdRelatedTarget: null
    });
    //Used so focus is handled just one time using currentFocus. Used combined with menuLevel due to main menu can't
    // change its focus via currentFocus prop.
    const initialFocusDone = useRef(false);
    //Key of the active element on this menu container
    const [currentKey, setCurrentKey] = useState();
    //State that represents if focused outline styles needs
    const [showFocusRing, setShowFocusRing] = useState(false);
    //Based on this we would include or not hovering styling. Same as legacy
    const [isContainerFocused, setIsContainerFocused] = useState(false);
    //Funtion triggered on blur. This is called since if focus is not anymore on the container we would need to do
    //some validation
    const blurFunction = () => {
        if (!menuContainerRef.current?.contains(document.activeElement)) {
            setShowFocusRing(false);
            setIsContainerFocused(false);
        }
    };
    const focusFunction = () => {
        setIsContainerFocused(true);
    };
    const mouseLeaveFunction = (e) => {
        //We have to validate if new hover target was already validated correctly, if not this means mouse is now outside container
        if (!e.relatedTarget.isEqualNode(menuContainerGeneralInformation.current.lastMouseLeaveMenuItemdRelatedTarget)) {
            //We reset current key when mouse is outside the container or any submenu. This is how legacy menu works
            changeKey({
                key: undefined,
                reason: 'mouseEnter'
            });
            //We focus menuContainer so keyboard navigation can be still used
            menuContainerRef.current?.focus();
        }
    };
    //Function used to change the currentKey
    const changeKey = useCallback((detail) => {
        setCurrentKey(detail.key);
        if (detail.reason === 'keyboard') {
            setShowFocusRing(true);
        }
        else {
            setShowFocusRing(false);
        }
    }, []);
    //This is used since if mouse overlaps space when menu is rendered, menuItem could trigger menuEnter,
    // which is not desirable.Once mouse was moved inside container, we could trigger mouseEnter events normally.
    const selfDestructiveMouseMoveEvent = useCallback(() => {
        const handler = (e) => {
            const keyElem = keyElemExtractor(e.target);
            const key = keyElem?.key;
            const elem = keyElem?.elem;
            //We will trigger this event until a menuItem is found. Usually this don't take to much to happen
            if (key && elem) {
                menuContainerGeneralInformation.current.wasMouseMoved = true;
                const evt = new MouseEvent('mouseenter');
                elem.dispatchEvent(evt);
                menuContainerRef.current?.removeEventListener('mousemove', handler);
            }
        };
        menuContainerRef.current?.addEventListener('mousemove', handler);
    }, []);
    const { currentKeyProps } = useMenuCurrentKey(getPrevNext(menuContainerRef.current, true, currentKey), getPrevNext(menuContainerRef.current, false, currentKey), () => getNVisible(menuContainerRef.current, 0), () => getLastVisible(menuContainerRef.current), changeKey);
    useEffect(() => {
        if (menuLevel === 'main' && initialFocusDone.current)
            return;
        initialFocusDone.current = true;
        const menuContainer = menuContainerRef.current;
        let timer;
        /*TODO: There are some cases where is not safe to focus since
        visibility initially is hidden at the Floating component at the initial render. Need a way to solve this in a
        less dirty way (we are now using timeouts). More details on the descripton of JET-57353 */
        if (currentFocus !== 'none' && menuContainer) {
            if (currentFocus === 'menu') {
                changeKey({ key: undefined, reason: 'click' });
                const focusMenuContainer = () => {
                    menuContainer.focus();
                };
                if (display === 'sheet') {
                    focusMenuContainer();
                }
                else {
                    // At this point floating has visibility set as hidden, so also menuContainer. Settimeout added so menuContainer
                    // is visible at that point and can be focused.
                    timer = setTimeout(() => {
                        focusMenuContainer();
                    }, FOCUS_SHIFT_TIMEOUT);
                }
            }
            else if (currentFocus === 'firstItem') {
                const firstItem = getNVisible(menuContainerRef.current, 0);
                if (firstItem) {
                    changeKey({ key: firstItem, reason: 'keyboard' });
                }
            }
            else if (currentFocus === 'secondItem') {
                //This is just used in sheet menu, so no need to set a timeout
                const item = getNVisible(menuContainerRef.current, 1);
                if (item) {
                    changeKey({ key: item, reason: 'keyboard' });
                }
            }
        }
        return () => {
            //We have to clean timers, because comp can be unmounted after finishing timeouts
            clearTimeout(timer);
        };
    }, [changeKey, currentFocus, menuLevel, display]);
    useEffect(() => {
        selfDestructiveMouseMoveEvent();
    }, [selfDestructiveMouseMoveEvent]);
    return (jsx("div", { ...(menuId && { id: menuId }), ref: menuContainerRef, tabIndex: !currentKey && (menuLevel === 'main' || display === 'sheet') ? 0 : -1, role: "menu", "aria-label": ariaLabel, class: classNames([
            menuStyles.menuContainerStyle,
            menuStyles[`${display}MenuContainerStyle`]
        ]), ...mergeProps(currentKeyProps, getKeyboardCloseProps(onClose, menuLevel), getCloseSubmenuEvent(direction, closeSubmenu), getMenuContainerFocusMethods(blurFunction, focusFunction), getMenuContainerMouseLeave(mouseLeaveFunction), getMenuPointerDown()), children: jsx(MenuContainerContext.Provider, { value: {
                /*We use this ref on dropdownSubmenu since unhover only close the menu if menuContainer ref does not include
                the new element that receives the hovering */
                menuContainerRef,
                currentKey,
                showFocusRing,
                changeKey,
                menuContainerGeneralInformation,
                isContainerFocused
            }, children: children }) }));
};
//Temporal constant, will probably be removed with JET-57353
const FOCUS_SHIFT_TIMEOUT = 0;

/*Component used to render a dropdown menu. For now this is just rendered when device is not a phone. */
const DropdownMenu = forwardRef(({ anchorRef, defaultPlacement, offsetValue, closeSubmenu, children, 'aria-label': ariaLabel, currentFocus, menuId, menuLevel, flipOptions }, ref) => {
    const { onClose } = useMenuContext();
    const useOutsideClickHandler = useCallback$1((e) => {
        if (menuLevel === 'sub')
            return;
        // Outside click is not considered when menuItem was clicked
        const menuItemKey = keyElemExtractor(e.target)?.key;
        if (!menuItemKey) {
            onClose?.({ reason: 'outsideClick', target: e.target });
        }
    }, [onClose, menuLevel]);
    // TODO - Reuse a dropdown component instead of creating floating by ourselves.
    return (jsx(Layer, { logicalParentRef: anchorRef, children: jsx(Floating, { class: dropdownMenuStyles.base, onClickOutside: useOutsideClickHandler, anchorRef: anchorRef, flipOptions: flipOptions, shiftOptions: {
                mainAxis: false,
                crossAxis: false
            }, sizeOptions: { isMaxHeightAdjusted: true, isMaxWidthAdjusted: true }, 
            //TODO: JET-57950. Need to see how we set 24px boundary.
            //TODO: Boundaries need to match with the dropdowns used in other comps.  Also figure out how to make it themable.
            boundaryOptions: { padding: 0 }, placement: defaultPlacement, offsetValue: offsetValue, children: jsx("div", { style: {
                    minWidth: `${anchorRef.current && anchorRef.current.offsetWidth > MENU_MIN_WIDTH
                        ? anchorRef.current.offsetWidth
                        : MENU_MIN_WIDTH}px`
                }, ref: ref, children: jsx(MenuContainer, { menuLevel: menuLevel, closeSubmenu: closeSubmenu, "aria-label": ariaLabel, currentFocus: currentFocus, menuId: menuId, display: "dropdown", children: children }) }) }) }));
});
const MENU_MIN_WIDTH = 200;

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/*This "private" hook will live until we end up stating a final decision on useSwipe hook.
For that hook to work, element that use it has to set touch-action:none;, since pointer events
are canceled once a touch action is going to be handled by the browser. This works for most
of the components that don't scroll. But if the component scroll, there is a need for the browser to
handle scroll which is imposible with a hook just using pointer events.
*/
function useMenuSwipe(onSwipe, { threshold = 10, maximumTime = 300, tolerance = 50, isDisabled = false } = {
    threshold: 10,
    maximumTime: 300,
    tolerance: 50,
    isDisabled: false
}) {
    const ref = useRef({
        pointerDownId: 0,
        startTime: null,
        prevpageX: 0,
        prevpageY: 0,
        elementScrollTop: 0
    });
    const clearSwipe = useCallback(() => {
        ref.current = {
            pointerDownId: null,
            startTime: 0,
            prevpageX: 0,
            prevpageY: 0,
            elementScrollTop: 0
        };
    }, []);
    const onTouchStart = useCallback((e) => {
        const touchobj = e.changedTouches[0];
        if (!ref.current.pointerDownId) {
            ref.current = {
                pointerDownId: touchobj.identifier,
                startTime: e.timeStamp,
                prevpageX: touchobj.pageX,
                prevpageY: touchobj.pageY,
                elementScrollTop: e.currentTarget.scrollTop
            };
        }
        else {
            clearSwipe();
        }
    }, [clearSwipe]);
    const onTouchEnd = useCallback((ev) => {
        const touchobj = ev.changedTouches[0];
        if (touchobj.identifier === ref.current.pointerDownId) {
            if (ref.current.startTime && ev.timeStamp - ref.current.startTime < maximumTime) {
                if (touchobj.pageY - ref.current.prevpageY > threshold &&
                    Math.abs(touchobj.pageX - ref.current.prevpageX) <= tolerance &&
                    ref.current.elementScrollTop === 0) {
                    onSwipe?.({ direction: 'down' });
                }
            }
        }
        // state would be reset to its original state whether swipe is completed or cancelled
        clearSwipe();
    }, [threshold, onSwipe, clearSwipe, maximumTime, tolerance]);
    const swipeProps = isDisabled
        ? {}
        : {
            /** event handlers to spread on the target element. */
            onTouchStart,
            onTouchEnd,
            ontouchcancel: clearSwipe
        };
    return {
        swipeProps
    };
}

const ModalWrapper = ({ children }) => (jsx("div", { class: sheetMenuStyles.modalWrapper, children: children }));
const ModalContainer = forwardRef(({ children }, ref) => (jsx("div", { class: sheetMenuStyles.modalContainer, ref: ref, children: children })));
//ModalBackdrop on modal folder not setting the style on oj-c enviroment
//Creating it here so we don't see that issue
const ModalBackdrop = ({ backdropVisible = false }) => (jsx("div", { class: classNames([
        sheetMenuStyles.modalBackdrop,
        backdropVisible && sheetMenuStyles.modalBackdropVisible
    ]) }));
const SheetMenu = ({ children, 'aria-label': ariaLabel, currentFocus, anchorRef, closeSubmenu, menuLevel }) => {
    const { onClose } = useMenuContext();
    useEffect(() => {
        // We set the overflow body as hidden when mounted and revert to original overflow value when is
        // unmounted. This is done to don't allow backgound being scrolled
        // This will be handled by modal/popup when ready
        const initialOverflow = document.body.style.overflow;
        document.body.style.overflow = 'hidden';
        return () => {
            document.body.style.overflow = initialOverflow;
        };
    }, []);
    const dismissHandler = useCallback(() => {
        onClose?.({ reason: 'dismissed' });
    }, [onClose]);
    const { backdropRef } = useModal({ isOpen: true, onBackdropClick: dismissHandler });
    //TODO: Use useSwipe when it works on components that could scroll.
    const { swipeProps } = useMenuSwipe(dismissHandler);
    // TODO: For now we are mimicking current modal but using a "different position setting" (using absolute positioning instead of flex).
    // Have to use modal/popup component directly instead doing this composition when that component could satisfy our requirments.
    return (jsx(Layer, { logicalParentRef: anchorRef, children: jsxs(ModalWrapper, { children: [jsx(ModalBackdrop, { backdropVisible: menuLevel === 'main' }), jsx(ModalContainer, { ref: backdropRef, children: jsx("div", { class: sheetMenuStyles.sheetMenuContainerWrapper, ...swipeProps, children: jsx(MenuContainer, { ...{ ariaLabel, currentFocus, display: 'sheet', closeSubmenu, menuLevel }, children: children }) }) })] }) }));
};

const isMobile = getIsMobile();
/* Component that is exposed to final user to create a menu. Important to notice that this component is just used by
the final user to express a main menu. It's not reused by us for submenus (we use SheetMenu and DropdownMenu directly there) */
const Menu = ({ onClose, isOpen, initialFocus, ...menuProps }) => {
    //On close is consumed by the whole menu structure including the main menu and submenu ones.
    return (jsx(MenuContext.Provider, { value: { onClose, isMobile }, children: !isOpen ? null : (jsx(Fragment, { children: isMobile ? (jsx(SheetMenu, { ...{
                    ...menuProps,
                    currentFocus: initialFocus,
                    menuLevel: 'main'
                } })) : (jsx(DropdownMenu, { ...{
                    ...menuProps,
                    currentFocus: initialFocus,
                    menuLevel: 'main',
                    defaultPlacement: 'bottom-start',
                    offsetValue: {
                        mainAxis: 4,
                        crossAxis: 0
                    },
                    flipOptions: {
                        crossAxis: true,
                        mainAxis: true,
                        fallbackPlacements: [
                            'bottom-end',
                            'top-start',
                            'top-end',
                            'end-top',
                            'start-top',
                            'end-bottom',
                            'start-bottom'
                        ]
                    }
                } })) })) }));
};

export { DropdownMenu as D, Menu as M, SheetMenu as S };
