// @oracle/oraclejet-preact/es/PRIVATE_CheckboxRadioField.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsxs, Fragment, jsx } from 'preact/jsx-runtime';
import { forwardRef, useRef, useImperativeHandle } from 'preact/compat';
import { useCollectionFocusRing } from './hooks/PRIVATE_useCollectionFocusRing.js';
import './UNSAFE_Flex.js';
import './UNSAFE_Label.js';
import { InlineUserAssistanceContainer, InlineHelp, InlineUserAssistance } from './UNSAFE_UserAssistance.js';
import { useMessageSeverity } from './hooks/PRIVATE_useMessageSeverity.js';
import { useAccessibleContext } from './hooks/UNSAFE_useAccessibleContext.js';
import './utils/UNSAFE_stringUtils.js';
import { m as merge } from './stringUtils-5f2d2e94.js';
import './UNSAFE_Theme.js';
import './CheckboxRadioFieldStyles.styles.css';
import { LabelValueLayout } from './UNSAFE_LabelValueLayout.js';
import { C as CheckboxRadioContext } from './CheckboxRadioContext-0aebb1d4.js';
import { F as Flex } from './Flex-46adb0e4.js';
import { x as xUnits } from './utils-9a0cb8dc.js';
import { L as Label } from './Label-642674b7.js';
import 'preact/hooks';
import './hooks/UNSAFE_useComponentTheme.js';
import './utils/UNSAFE_logger.js';
import './_curry1-7629ff3f.js';
import './_curry2-820ed778.js';
import './InlineHelpSource-6e0b687e.js';
import './utils/UNSAFE_classNames.js';
import './classNames-f414daad.js';
import './hooks/UNSAFE_useTabbableMode.js';
import 'preact';
import './hooks/UNSAFE_useTranslationBundle.js';
import './index-f26154f2.js';
import './Common.js';
import './Common/themes.js';
import './Common/themes/redwood/theme.js';
import './Common/themes/themeContract.css.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js';
import './UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js';
import './UserAssistanceStyles.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-3311d9ef.js';
import './UNSAFE_ComponentMessage.js';
import './ComponentMessage-a41ebbdc.js';
import './PRIVATE_Message.js';
import './utils/PRIVATE_timer.js';
import './PRIVATE_Message/themes/MessageStyles.css.js';
import './MessageStyles.styles.css';
import './MessageCloseButton-b3e6b568.js';
import './UNSAFE_IconButton.js';
import './IconButton-d6cc2f62.js';
import './UNSAFE_BaseButton.js';
import './BaseButton-06d74a34.js';
import './hooks/UNSAFE_usePress.js';
import './hooks/UNSAFE_useHover.js';
import './hooks/UNSAFE_useToggle.js';
import './hooks/UNSAFE_useActive.js';

import './utils/UNSAFE_interpolations/dimensions.js';
import './utils/UNSAFE_arrayUtils.js';
import './utils/UNSAFE_size.js';
import './utils/UNSAFE_mergeInterpolations.js';
import './_curry3-bb17e901.js';
import './_has-057b6f90.js';
import './utils/PRIVATE_clientHints.js';
import './clientHints-26cce6ec.js';
import './utils/UNSAFE_mergeProps.js';
import './UNSAFE_ButtonLayout.js';
import './ButtonLayout-8ea4b106.js';
import './UNSAFE_Text.js';
import './Text-0dad47bb.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './hooks/UNSAFE_useTooltip.js';
import './hooks/UNSAFE_useId.js';
import './UNSAFE_Floating.js';
import './Floating-2d909840.js';
import './index-326d7a5f.js';
import './hooks/UNSAFE_useUser.js';
import './utils/PRIVATE_floatingUtils.js';
import './utils/PRIVATE_refUtils.js';
import './hooks/UNSAFE_useOutsideClick.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-33365337.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './UNSAFE_Layer.js';
import './useThemeInterpolations-48ee20a7.js';
import './hooks/UNSAFE_useColorScheme.js';
import './hooks/UNSAFE_useScale.js';
import './utils/UNSAFE_interpolations/theme.js';
import './hooks/UNSAFE_useFocus.js';
import './hooks/UNSAFE_useTouch.js';
import './hooks/UNSAFE_useAnimation.js';
import './useAnimation-50c5f5fb.js';
import './hooks/UNSAFE_useThemeInterpolations.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './PRIVATE_ThemedIcons/CloseIcon.js';
import './UNSAFE_Icon.js';
import './Icon-e04c1503.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './IconStyle.styles.css';
import './MessageDetail-558fff70.js';
import './MessageFormattingUtils-5a95e8e9.js';
import './utils/UNSAFE_getLocale.js';
import './Message.types-06b5fdd0.js';
import './MessageStartIcon-a9a1a453.js';
import './PRIVATE_ThemedIcons/MessageConfirmationIcon.js';
import './PRIVATE_ThemedIcons/MessageErrorIcon.js';
import './PRIVATE_ThemedIcons/MessageInfoIcon.js';
import './PRIVATE_ThemedIcons/MessageWarningIcon.js';
import './MessageSummary-29c1f7ef.js';
import './MessageTimestamp-d46daaa0.js';
import './MessageUtils-7e2e288a.js';
import './utils/UNSAFE_soundUtils.js';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerTheme.js';
import './UNSAFE_MessageBanner/themes/MessageBannerStyles.css.js';
import './MessageBannerStyles.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.css.js';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js';
import './MessageBannerVariants.styles.css';
import './MessagesManager-6e35eea0.js';
import './PRIVATE_TransitionGroup.js';
import './hooks/UNSAFE_useMessagesContext.js';
import './UNSAFE_HiddenAccessible.js';
import './HiddenAccessible-6ca9b782.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js';
import './UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js';
import './ComponentMessageStyles.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js';
import './ComponentMessageContainer-94ad5e1d.js';
import './hooks/UNSAFE_useFormContext.js';
import './hooks/UNSAFE_useFormFieldContext.js';
import './hooks/UNSAFE_useInputGroupContext.js';
import './UNSAFE_Popup.js';
import './utils/PRIVATE_tabbableUtils.js';
import './_arity-cb2c60f3.js';
import './UNSAFE_FocusTrap.js';
import './FocusTrap-9dc31155.js';
import './hooks/PRIVATE_useAnimationStatus.js';
import './UNSAFE_Popup/themes/PopupContract.css.js';
import './UNSAFE_Popup/themes/redwood/PopupTheme.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.styles.css';
import './UNSAFE_Popup/themes/redwood/PopupVariants.css.js';
import './PopupStyles.styles.css';
import './UNSAFE_Divider.js';
import './Divider-709d2810.js';
import './UNSAFE_Divider/themes/DividerStyles.css.js';
import './DividerStyles.styles.css';
import './CompactHelpSource-8d377623.js';
import './PRIVATE_ThemedIcons/HelpIcon.js';
import './hooks/UNSAFE_useFocusWithin.js';
import './useFocusWithin-f96a205e.js';
import './UNSAFE_Label/themes/redwood/LabelTheme.js';
import './UNSAFE_Label/themes/LabelStyles.css.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import './LabelValueLayoutStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldTheme.js';
import './UNSAFE_TextField/themes/TextFieldStyles.css.js';
import './SkeletonStyles.styles.css';
import './TextFieldLoadingStyles.styles.css';
import './TextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js';
import './TextFieldVariants.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioTheme.js';
import './UNSAFE_Radio/themes/RadioStyles.css.js';
import './RadioStyles.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioBaseTheme.css.js';
import './UNSAFE_Radio/themes/redwood/RadioBaseTheme.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioVariants.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutTheme.js';
import './UNSAFE_TextField/themes/FormLayoutStyles.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.styles.css';
import './utils/UNSAFE_interpolations/boxalignment.js';
import './keys-f766c992.js';
import './utils/UNSAFE_interpolations/flexbox.js';
import './flexbox-645317db.js';
import './utils/UNSAFE_interpolations/flexitem.js';
import './flexitem-89d8a00a.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function useCheckboxRadioField({ ariaDescribedBy, baseId, hasValue, hasTopUserAssistance, hasBottomUserAssistance, isReadonly, isRequired = false, label, labelEdge, messages }) {
    const optionsGroupId = `${baseId}-group`;
    const labelId = `${baseId}-label`;
    const topUaId = hasTopUserAssistance ? `${baseId}-top-ua` : undefined;
    const bottomUaId = hasBottomUserAssistance ? `${baseId}-bottom-ua` : undefined;
    const { UNSAFE_ariaLabelledBy } = useAccessibleContext();
    const ariaLabelledBy = label === '' && labelEdge === 'none' ? UNSAFE_ariaLabelledBy : labelId;
    const highestMessageSeverity = useMessageSeverity(messages);
    const hasErrorMessage = highestMessageSeverity === 'error';
    return {
        groupProps: {
            'aria-describedby': merge([topUaId, bottomUaId, ariaDescribedBy]),
            'aria-labelledby': ariaLabelledBy,
            'aria-required': isRequired,
            ...(isReadonly && { readonly: true }),
            ...(hasErrorMessage && { 'aria-invalid': true }),
            id: optionsGroupId
            // TODO: ariaInvalid when one of message is error type?
        },
        labelProps: {
            forId: optionsGroupId,
            id: labelId
        },
        // TOP UA is assistive text and help link/text
        ...(hasTopUserAssistance && {
            topUserAssistanceProps: {
                id: topUaId,
                // Use reflow variant so that top UA doesn't have min height
                variant: 'reflow'
            }
        }),
        // Bottom UA is required indicator or messages
        ...(hasBottomUserAssistance && {
            bottomUserAssistanceProps: {
                id: bottomUaId,
                isRequiredShown: !!isRequired && !hasValue
            }
        })
    };
}

var topUAContainerStyles = {top:'CheckboxRadioFieldStyles_topUAContainerStyles_top__j1xru60',start:'CheckboxRadioFieldStyles_topUAContainerStyles_start__j1xru61'};

const CheckboxRadioField = forwardRef(({ 'aria-describedby': ariaDescribedBy, assistiveText, baseId, direction = 'column', hasValue, helpSourceLink, helpSourceText, isDisabled, isReadonly, isRequired, label, labelEdge, labelStartWidth = '33%', parentComponentVariant, messages, userAssistanceDensity, role, children }, ref = null) => {
    const hasTopUserAssistance = !!assistiveText || (!!helpSourceLink && !!helpSourceText);
    const hasBottomUserAssistance = !isDisabled && !isReadonly;
    const groupRef = useRef(null);
    const { groupProps, labelProps, topUserAssistanceProps, bottomUserAssistanceProps } = useCheckboxRadioField({
        ariaDescribedBy,
        baseId,
        isRequired,
        isReadonly,
        label,
        labelEdge,
        messages,
        hasValue,
        hasTopUserAssistance,
        hasBottomUserAssistance
    });
    useImperativeHandle(ref, () => ({
        focus: () => {
            if (!isDisabled) {
                const inputChildren = groupRef.current?.querySelectorAll('input') ?? [];
                const selectedChild = Array.from(inputChildren).find((child) => child.checked);
                if (selectedChild) {
                    selectedChild.focus();
                }
                else {
                    inputChildren[0].focus();
                }
            }
        },
        blur: () => {
            if (groupRef.current?.contains(document.activeElement)) {
                document.activeElement.blur();
            }
        }
    }));
    const [isFocusRingShown, focusRingProps] = useCollectionFocusRing((elem) => {
        return !!groupRef.current?.contains(elem);
    }, ['ArrowUp', 'ArrowRight', 'ArrowDown', 'ArrowLeft']);
    const content = (jsxs(Fragment, { children: [hasTopUserAssistance && (jsx("div", { id: `${baseId}-top-ua-container`, class: labelEdge === 'start' ? topUAContainerStyles.start : topUAContainerStyles.top, children: jsx(InlineUserAssistanceContainer, { ...topUserAssistanceProps, children: jsx(InlineHelp, { assistiveText: assistiveText, sourceLink: helpSourceLink, sourceText: helpSourceText }) }) })), jsx(Flex, { direction: direction, ...(direction === 'row' && { gap: xUnits(8) }), children: children }), hasBottomUserAssistance && !isReadonly && (jsx(InlineUserAssistance, { ...bottomUserAssistanceProps, requiredAlignment: "start", messages: messages, userAssistanceDensity: userAssistanceDensity === 'reflow' ? userAssistanceDensity : 'efficient', id: `${baseId}-bottom-ua-container` }))] }));
    // TODO: spacing under label when labelEdge === 'top' and hasTopUA.
    return (jsxs("div", { role: role, ref: groupRef, ...groupProps, ...focusRingProps, children: [labelEdge === 'none' && content, labelEdge !== 'none' && (jsx(LabelValueLayout, { parentComponentVariant: parentComponentVariant, hasTopUserAssistance: hasTopUserAssistance, label: jsx(Label, { ...labelProps, variant: labelEdge, userAssistanceDensity: "efficient", parentComponentVariant: parentComponentVariant, children: label }), labelEdge: labelEdge === 'inside' ? 'top' : labelEdge, labelStartWidth: labelStartWidth, children: jsx(CheckboxRadioContext.Provider, { value: { isFocusRingShown }, children: content }) }))] }));
});

export { CheckboxRadioField };
