// @oracle/oraclejet-preact/es/UNSAFE_MeterBar.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx, jsxs, Fragment } from 'preact/jsx-runtime';
import { sizeToCSS } from './utils/UNSAFE_size.js';
import './utils/UNSAFE_classNames.js';
import { mergeProps } from './utils/UNSAFE_mergeProps.js';
import { useRef, useEffect } from 'preact/hooks';
import { useUser } from './hooks/UNSAFE_useUser.js';
import { usePointerEvents, useKeyboardEvents, useMeterDatatip } from './hooks/PRIVATE_useDvtMeterEvents.js';
import { validateRange, getMeterAriaProps, getTrackAndIndicatorColor } from './utils/PRIVATE_meterUtils.js';
import { BarTrack, BarValue } from './PRIVATE_Meter.js';
import { getDvtThresholdColor } from './utils/UNSAFE_dvtCommonUtils.js';
import { styles } from './UNSAFE_MeterBar/themes/MeterBarStyles.css.js';
import { c as classNames } from './classNames-f414daad.js';
import { useTabbableMode } from './hooks/UNSAFE_useTabbableMode.js';
import { useComponentTheme } from './hooks/UNSAFE_useComponentTheme.js';
import { MeterBarRedwoodTheme } from './UNSAFE_MeterBar/themes/redwood/MeterBarTheme.js';
import './UNSAFE_Theme.js';
import './utils-9a0cb8dc.js';
import './Common/themes/themeContract.css.js';
import './index-f26154f2.js';
import 'preact';
import './Common.js';
import './Common/themes.js';
import './Common/themes/redwood/theme.js';
import 'preact/compat';
import './hooks/UNSAFE_useTooltip.js';
import './hooks/UNSAFE_useId.js';
import './UNSAFE_Floating.js';
import './Floating-2d909840.js';
import './index-326d7a5f.js';
import './utils/PRIVATE_floatingUtils.js';
import './utils/PRIVATE_refUtils.js';
import './hooks/UNSAFE_useOutsideClick.js';
import './utils/UNSAFE_arrayUtils.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-3311d9ef.js';
import './vanilla-extract-dynamic.esm-33365337.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './utils/UNSAFE_logger.js';
import './_curry1-7629ff3f.js';
import './_curry2-820ed778.js';
import './UNSAFE_Layer.js';
import './useThemeInterpolations-48ee20a7.js';
import './hooks/UNSAFE_useColorScheme.js';
import './hooks/UNSAFE_useScale.js';
import './utils/UNSAFE_interpolations/theme.js';
import './utils/UNSAFE_mergeInterpolations.js';
import './_curry3-bb17e901.js';
import './_has-057b6f90.js';
import './hooks/UNSAFE_useHover.js';
import './hooks/UNSAFE_useToggle.js';
import './hooks/UNSAFE_useFocus.js';
import './hooks/UNSAFE_useTouch.js';
import './hooks/UNSAFE_useAnimation.js';
import './useAnimation-50c5f5fb.js';
import './hooks/UNSAFE_useThemeInterpolations.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './MeterBarStyles.styles.css';
import './UNSAFE_MeterBar/themes/redwood/MeterBarBaseTheme.css.js';
import './UNSAFE_MeterBar/themes/redwood/MeterBarBaseTheme.styles.css';
import './UNSAFE_MeterBar/themes/redwood/MeterBarVariants.css.js';
import './MeterBarVariants.styles.css';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Returns the metric value of the clicked point.
 * @param event The PointerEvent
 * @param dimensionsRef Ref to stored dimensions of the meter bar.
 * @param min The minimum value of the meter bar.
 * @param max The maximum value of the meter bar.
 * @param step The step value of the meter bar.
 * @param orientation The orientation of the meter bar.
 * @param isRtl Whether the reading mode is 'rtl'
 * @returns The metric value of the clicked point.
 */
function getValue(event, dimensionsRef, min, max, step, orientation, isRtl) {
    const dims = dimensionsRef.current;
    if (!dims || dims.width === 0) {
        return undefined;
    }
    let val;
    if (orientation === 'horizontal') {
        const width = event.offsetX;
        val = min + ((max - min) * (isRtl ? dims.width - width : width)) / dims.width;
    }
    else {
        const height = dims.height - event.offsetY;
        val = min + ((max - min) * height) / dims.height;
    }
    val = Math.round(val / step) * step;
    return Math.max(min, Math.min(max, val));
}
/**
 * Returns the true dimension of the meter bar after first render. All dimensions are 0 for initial render.
 * @param element The meter bar track HTML element
 * @returns The dimension of the meter bar.
 */
function getDimensions(element) {
    if (!element) {
        return {
            width: 0,
            height: 0
        };
    }
    const trackDims = element.getBoundingClientRect();
    return {
        width: trackDims.width,
        height: trackDims.height
    };
}
/**
 * Returns object with 'start' and 'end' keys. Value for each key will be true if at least one ref line with correposnding position is present in meter bar.
 * @param lines The array of reference lines.
 * @returns Returns object with start and end keys.
 */
function getRefLinesInfo(lines) {
    let refLineInfo = { start: false, end: false };
    if (lines) {
        refLineInfo = lines.reduce((acc, cur) => {
            acc.start = acc.start || cur.position === 'start';
            acc.end = acc.end || cur.position === 'end' || cur.position == null;
            return acc;
        }, refLineInfo);
    }
    return refLineInfo;
}

function ReferenceLineBar({ orientation, percentage, color, position }) {
    const isHoriz = orientation === 'horizontal';
    const { direction } = useUser();
    const isRtl = direction === 'rtl';
    const isStartPos = position === 'start';
    return (jsx("div", { class: classNames([
            styles.referenceLinesBaseStyle,
            isHoriz ? styles.referenceLinesHorizontalStyle : styles.referenceLinesVerticalStyle,
            isStartPos
                ? isHoriz
                    ? styles.referenceLinesHoriontalStartStyle
                    : isRtl
                        ? styles.referenceLinesVerticalStartRtlStyle
                        : styles.referenceLinesVerticalStartLtrStyle
                : ''
        ]), style: {
            left: isHoriz && !isRtl ? percentage : undefined,
            right: isHoriz && isRtl ? percentage : undefined,
            bottom: isHoriz ? undefined : percentage,
            backgroundColor: getDvtThresholdColor(color)
        } }));
}

function ThresholdBar({ color, length, orientation, ...otherProps }) {
    const { direction } = useUser();
    const isHoriz = orientation === 'horizontal';
    const isRtl = direction === 'rtl';
    return (jsx("div", { class: classNames([
            styles.thresholdBarBaseStyle,
            isHoriz
                ? isRtl
                    ? styles.thresholdBarHorizontalBarTrackSectionRtlStyle
                    : styles.thresholdBarHorizontalBarTrackSectionLtrStyle
                : styles.thresholdBarVerticalBarTrackSectionStyle,
            otherProps.class
        ]), style: {
            width: isHoriz ? length : '100%',
            height: isHoriz ? '100%' : length,
            backgroundColor: getDvtThresholdColor(color)
        } }));
}

/**
 * A Meter bar displays information graphically in a rectangular bar, highlighting a specific metric value's progress in relation to its min, max, or thresholds.
 */
function MeterBar({ max = 100, min = 0, value = 0, step = 1, length = '100%', size, orientation = 'horizontal', isTrackRendered = true, thresholdDisplay = 'all', indicatorSize = 1, ...props }) {
    validateRange(min, max, value, step);
    const currentInputRef = useRef();
    const isHoriz = orientation === 'horizontal';
    const barTrackRef = useRef(null);
    const dimensionsRef = useRef(getDimensions(null));
    const { direction } = useUser();
    const isRtl = direction === 'rtl';
    useEffect(() => {
        if (props.onCommit || props.onInput) {
            dimensionsRef.current = getDimensions(barTrackRef.current);
        }
    }, [size, length, orientation, props.onCommit, props.onInput]);
    const getValueFromEvent = (event) => {
        // only evaluate value from bar track
        if (event.target == barTrackRef.current) {
            return getValue(event, dimensionsRef, min, max, step, orientation, isRtl);
        }
        return;
    };
    const pointerEventsProps = usePointerEvents(value, getValueFromEvent, barTrackRef, currentInputRef, props.onCommit, props.onInput);
    const keyboardEventsProps = useKeyboardEvents(value, min, max, step, currentInputRef, props.onCommit, props.onInput);
    const { datatipContent, datatipProps } = useMeterDatatip(value, props.datatip, props['aria-describedby']);
    const ariaLabelProps = getMeterAriaProps(value, min, max, props['aria-label'], props['aria-labelledby'], props.thresholds);
    const mergedProps = mergeProps(pointerEventsProps, keyboardEventsProps, datatipProps, ariaLabelProps);
    const validIndicatorSize = Math.max(0, Math.min(1, indicatorSize));
    const valueBarLength = `${((value - min) * 100) / (max - min)}%`;
    const valueBarSize = `${validIndicatorSize * 100}%`;
    const { trackColor, indicatorColor } = getTrackAndIndicatorColor(value, thresholdDisplay, props.trackColor, props.indicatorColor, props.thresholds);
    const sizeStyle = getRefLinesInfo(props.referenceLines);
    const isInteractive = props.onCommit || props.onInput;
    const { isTabbable } = useTabbableMode();
    const { classes, styles: { interactiveStyle, barValueBaseStyle, barValueVerticalStyle, barValueHorizontalStyle, barTrackBaseStyle, barTrackHorizontalSizeStyle, barTrackVerticalSizeStyle, barTrackMarginTopStyle, barTrackMarginStartStyle, barTrackMarginBottomStyle, barTrackMarginEndStyle, trackSectionHorizontalFirstStyle, trackSectionHorizontalLastStyle, trackSectionVerticalFirstStyle, trackSectionVerticalLastStyle, trackSectionRoundedCornersStyle } } = useComponentTheme(MeterBarRedwoodTheme, {
        size,
        orientation
    });
    return (jsxs(Fragment, { children: [jsx("div", { tabIndex: !isTabbable ? -1 : 0, ...mergedProps, style: {
                    width: isHoriz ? sizeToCSS(length) : undefined,
                    height: isHoriz ? undefined : sizeToCSS(length)
                }, class: classNames([classes]), children: jsxs(BarTrack, { ref: barTrackRef, orientation: orientation, color: !isTrackRendered || (thresholdDisplay === 'all' && props.thresholds)
                        ? 'transparent'
                        : trackColor, class: classNames([
                        barTrackBaseStyle,
                        isInteractive ? interactiveStyle : '',
                        isHoriz ? barTrackHorizontalSizeStyle : barTrackVerticalSizeStyle,
                        ...getMarginClass(sizeStyle, isHoriz, barTrackMarginTopStyle, barTrackMarginStartStyle, barTrackMarginBottomStyle, barTrackMarginEndStyle)
                    ]), children: [thresholdDisplay === 'all' &&
                            props.thresholds
                                ?.filter((threshold) => threshold.max > min && threshold.max <= max)
                                .sort((a, b) => a.max - b.max)
                                .map((threshold, index, thresholds) => {
                                const diff = index === 0 ? threshold.max - min : threshold.max - thresholds[index - 1].max;
                                const trackSectionLastStyle = isHoriz
                                    ? trackSectionHorizontalLastStyle
                                    : trackSectionVerticalLastStyle;
                                let trackSectionStyleClass = index === 0
                                    ? isHoriz
                                        ? trackSectionHorizontalFirstStyle
                                        : trackSectionVerticalFirstStyle
                                    : '';
                                if (threshold.max === max)
                                    trackSectionStyleClass =
                                        index === 0
                                            ? trackSectionRoundedCornersStyle // Single threshold from min to max
                                            : trackSectionLastStyle;
                                return (jsxs(Fragment, { children: [jsx(ThresholdBar, { class: trackSectionStyleClass, orientation: orientation, length: `${(diff * 100) / (max - min)}%`, color: threshold.color }), isTrackRendered && index === thresholds.length - 1 && threshold.max < max ? (jsx(ThresholdBar, { class: trackSectionLastStyle, color: trackColor, orientation: orientation, length: `${((max - thresholds[index].max) * 100) / (max - min)}%` })) : ('')] }));
                            }), jsx(BarValue, { length: valueBarLength, size: valueBarSize, orientation: orientation, class: classNames([
                                barValueBaseStyle,
                                isHoriz ? barValueHorizontalStyle : barValueVerticalStyle
                            ]), color: indicatorColor }), props.referenceLines?.map((item) => (jsx(ReferenceLineBar, { color: item.color, orientation: orientation, percentage: `${((item.value - min) * 100) / (max - min)}%`, position: item.position })))] }) }), datatipContent] }));
}
const getMarginClass = function (info, isHoriz, barTrackMarginTopStyle, barTrackMarginStartStyle, barTrackMarginBottomStyle, barTrackMarginEndStyle) {
    const className = [];
    if (info.start) {
        className.push(isHoriz ? barTrackMarginTopStyle : barTrackMarginStartStyle);
    }
    if (info.end) {
        className.push(isHoriz ? barTrackMarginBottomStyle : barTrackMarginEndStyle);
    }
    return className;
};

export { MeterBar };
