// @oracle/oraclejet-preact/es/UNSAFE_Theme.js
/* @oracle/oraclejet-preact: 15.1.2 */
export { f as fallbackVar, b as max, m as min, r as rgb, a as rgba, x as xUnits } from './utils-9a0cb8dc.js';
import './Common/themes/themeContract.css.js';

/**
 * The global theme's structure is defined here, where the theme associates the list of components
 * with their respective theme types. The theme is passed via the Environment context
 * and utilized by components using the `useComponentTheme` hook.
 */
// Constants for colorscheme classnames
// These are needed to support applications using legacy code
const DARK_CLASS = 'oj-c-colorscheme-dark';
const LIGHT_CLASS = 'oj-c-colorscheme-light';
const INVERT_CLASS = 'oj-color-invert'; // legacy class name for compatibility
const COLORSCHEME_DEPENDENT_CLASS = 'oj-c-colorscheme-dependent'; // The dependent class name is used to force re-evaluation
// of CSS variables that reference colorScheme token vars
// Constants for scale classnames
const SCALE_SM_CLASS = 'oj-scale-sm'; // legacy class name for compatibility
// We need to specify :root.oj-scale-sm in the selector due to how legacy switches scale
const SCALE_SM_SELECTORS = `.${SCALE_SM_CLASS}, :root.${SCALE_SM_CLASS}`;
const SCALE_MD_CLASS = 'oj-scale-md'; // legacy class name for compatibility
// We need to specify :root.oj-scale-md in the selector due to how legacy switches scale
const SCALE_MD_SELECTORS = `.${SCALE_MD_CLASS}, :root.${SCALE_MD_CLASS}`;
const SCALE_LG_CLASS = 'oj-scale-lg'; // legacy class name for compatibility
const SCALE_LG_SELECTORS = `.${SCALE_LG_CLASS}`; // 'lg' scale only has a single selector.
const SCALE_DEFAULT_SELECTORS = ':root';
const SCALE_DEPENDENT_CLASS = 'oj-scale-dependent'; // The dependent class name is used to force re-evaluation of CSS variables
// For legacy compatibility, we need to add :root.${SCALE_SM_CLASS} and :root${SACEL_MD_CLASS} so that the
// :root vars that refer to scale vars will re-evaluate in legacy
const SCALE_DEPENDENT_SELECTORS = `.${SCALE_DEPENDENT_CLASS}, :root.${SCALE_SM_CLASS}, :root.${SCALE_MD_CLASS}`;

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

export { COLORSCHEME_DEPENDENT_CLASS, DARK_CLASS, INVERT_CLASS, LIGHT_CLASS, SCALE_DEFAULT_SELECTORS, SCALE_DEPENDENT_CLASS, SCALE_DEPENDENT_SELECTORS, SCALE_LG_CLASS, SCALE_LG_SELECTORS, SCALE_MD_CLASS, SCALE_MD_SELECTORS, SCALE_SM_CLASS, SCALE_SM_SELECTORS };
