// @oracle/oraclejet-preact/es/UNSAFE_TextField.js
/* @oracle/oraclejet-preact: 15.1.2 */
export { R as ReadonlyTextFieldInput } from './ReadonlyTextFieldInput-cc60e2b7.js';
import { jsxs, jsx, Fragment } from 'preact/jsx-runtime';
import { useFormContext } from './hooks/UNSAFE_useFormContext.js';
import { LabelValueLayout } from './UNSAFE_LabelValueLayout.js';
import './utils/UNSAFE_classNames.js';
import { useFormFieldContext } from './hooks/UNSAFE_useFormFieldContext.js';
import { useComponentTheme } from './hooks/UNSAFE_useComponentTheme.js';
import { useInputGroupContext } from './hooks/UNSAFE_useInputGroupContext.js';
import { useUser } from './hooks/UNSAFE_useUser.js';
import { TextFieldRedwoodTheme } from './UNSAFE_TextField/themes/redwood/TextFieldTheme.js';
import { LabelRedwoodTheme } from './UNSAFE_Label/themes/redwood/LabelTheme.js';
import { FormLayoutRedwoodTheme } from './UNSAFE_TextField/themes/redwood/FormLayoutTheme.js';
import { ReadonlyTextFieldRedwoodTheme } from './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldTheme.js';
import { c as classNames } from './classNames-f414daad.js';
import './UNSAFE_Flex.js';
import { F as Flex } from './Flex-46adb0e4.js';
import { SkeletonRedwoodTheme } from './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
export { T as TextFieldInput, g as getInputId } from './TextFieldInput-f1ae0ae9.js';
import { Fragment as Fragment$1 } from 'preact';
import { useDebounce } from './hooks/UNSAFE_useDebounce.js';
import { useTranslationBundle } from './hooks/UNSAFE_useTranslationBundle.js';
import { LiveRegion } from './UNSAFE_LiveRegion.js';
import './hooks/UNSAFE_useAccessibleContext.js';
import 'preact/hooks';
import './hooks/UNSAFE_useTabbableMode.js';
import './utils/UNSAFE_interpolations/text.js';

import './index-0882a837.js';
import './utils/UNSAFE_mergeInterpolations.js';
import './_curry1-7629ff3f.js';
import './_curry3-bb17e901.js';
import './_curry2-820ed778.js';
import './_has-057b6f90.js';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldInputTheme.js';
import './UNSAFE_TextField/themes/ReadonlyTextFieldInputStyles.css.js';
import './ReadonlyTextFieldInputStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldInputVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-3311d9ef.js';
import './utils/UNSAFE_size.js';
import './UNSAFE_Theme.js';
import './utils-9a0cb8dc.js';
import './Common/themes/themeContract.css.js';
import './LabelValueLayoutStyles.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioTheme.js';
import './UNSAFE_Radio/themes/RadioStyles.css.js';
import './RadioStyles.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioBaseTheme.css.js';
import './UNSAFE_Radio/themes/redwood/RadioBaseTheme.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioVariants.css.js';
import './utils/UNSAFE_interpolations/dimensions.js';
import './utils/UNSAFE_arrayUtils.js';
import './utils/UNSAFE_interpolations/boxalignment.js';
import './keys-f766c992.js';
import './utils/UNSAFE_interpolations/flexbox.js';
import './flexbox-645317db.js';
import './utils/UNSAFE_interpolations/flexitem.js';
import './flexitem-89d8a00a.js';
import './utils/UNSAFE_logger.js';
import './UNSAFE_TextField/themes/TextFieldStyles.css.js';
import './SkeletonStyles.styles.css';
import './TextFieldLoadingStyles.styles.css';
import './TextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js';
import './TextFieldVariants.styles.css';
import './UNSAFE_Label/themes/LabelStyles.css.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import './UNSAFE_TextField/themes/FormLayoutStyles.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.styles.css';
import './index-f26154f2.js';
import './Common.js';
import './Common/themes.js';
import './Common/themes/redwood/theme.js';
import 'preact/compat';
import './UNSAFE_TextField/themes/ReadonlyTextFieldStyles.css.js';
import './ReadonlyTextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldVariants.css.js';
import './ReadonlyTextFieldVariants.styles.css';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './hooks/UNSAFE_useTextFieldInputHandlers.js';
import './utils/PRIVATE_clientHints.js';
import './clientHints-26cce6ec.js';
import './utils/UNSAFE_mergeProps.js';
import './UNSAFE_TextField/themes/redwood/TextFieldInputTheme.js';
import './UNSAFE_TextField/themes/TextFieldInputStyles.css.js';
import './TextFieldInputStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldInputVariants.css.js';

const StartTopLabelReadonlyTextField = ({ outerClassNames, innerClassNames, label, labelEdge, labelStartWidth, compactUserAssistance, inlineUserAssistance, children, ...props }) => {
    return (jsxs("div", { ...props, class: outerClassNames, children: [compactUserAssistance, jsx(LabelValueLayout, { label: label, labelEdge: labelEdge, labelStartWidth: labelStartWidth, children: jsx("div", { class: innerClassNames, children: children }) }), inlineUserAssistance] }));
};
const InsideLabelReadonlyTextField = ({ outerClassNames, innerClassNames, compactUserAssistance, inlineUserAssistance, children, ...props }) => {
    return (jsxs("div", { ...props, class: outerClassNames, children: [compactUserAssistance, jsx("div", { class: innerClassNames, children: children }), inlineUserAssistance] }));
};
const ReadonlyTextField = ({ label, labelEdge, children, variant, ...props }) => {
    const { isFormLayout, isReadonly: isReadonlyForm } = useFormContext();
    const { isLoading } = useFormFieldContext();
    const hasInsideLabel = label !== undefined && labelEdge === 'inside';
    const isInsideNonReadonlyForm = isFormLayout && !isReadonlyForm;
    const { direction } = useUser();
    const isLtr = direction === 'ltr';
    const inputGroupContext = useInputGroupContext();
    const hasGroupContext = inputGroupContext !== null;
    // depending on browser direction, apply the appropriate styling.
    // when 'rtl' the flex row direction is reversed.
    const inputGroupPosition = inputGroupContext
        ? inputGroupContext.last
            ? isLtr
                ? 'groupRight'
                : 'groupLeft'
            : inputGroupContext.index === 0
                ? isLtr
                    ? 'groupLeft'
                    : 'groupRight'
                : 'groupMiddle'
        : undefined;
    const { classes, styles } = useComponentTheme(TextFieldRedwoodTheme, {
        readonly: 'isReadonly'
    });
    const { baseTheme: labelBaseTheme } = useComponentTheme(LabelRedwoodTheme);
    const { baseTheme: formLayoutTheme } = useComponentTheme(FormLayoutRedwoodTheme);
    const { classes: readonlyClassNames } = useComponentTheme(ReadonlyTextFieldRedwoodTheme, {
        insideNonReadonlyForm: isInsideNonReadonlyForm
            ? 'isInsideNonReadonlyForm'
            : 'notInsideNonReadonlyForm',
        insideLabel: hasInsideLabel ? 'hasInsideLabel' : 'noInsideLabel',
        textarea: variant === 'textarea' ? 'isTextArea' : 'notTextArea',
        loading: isLoading ? 'isLoading' : 'notLoading',
        withinGroup: hasGroupContext ? 'isWithinGroup' : undefined,
        inputGroupPosition
    });
    const { formControlBase } = styles;
    const outerClassNames = classNames([classes, formControlBase, formLayoutTheme, labelBaseTheme]);
    if (label !== undefined && (labelEdge === 'start' || labelEdge === 'top')) {
        return (jsx(StartTopLabelReadonlyTextField, { outerClassNames: outerClassNames, innerClassNames: readonlyClassNames, label: label, labelEdge: labelEdge, ...props, children: children }));
    }
    return (jsxs(InsideLabelReadonlyTextField, { outerClassNames: outerClassNames, innerClassNames: readonlyClassNames, ...props, children: [label, children] }));
};

// Renders the oj-text-field-container dom which includes the
// start, middle (where the inside label and inputElem goes),
// and end pieces of the form component.
// This does not include user assistance because that is rendered outside the
// oj-text-field-container; that is rendered in the TextField component.
const TextFieldContent = ({ contentVariant = 'input', insideLabel, mainContent, startContent, statusVariant = 'none', styleVariant = 'default', endContent, resize, rootRef, hasZeroStartMargin }) => {
    const { isDisabled, isFocused, isLoading } = useFormFieldContext();
    const { direction } = useUser();
    const isLtr = direction === 'ltr';
    const hasInsideLabel = insideLabel !== undefined;
    const inputGroupContext = useInputGroupContext();
    const hasGroupContext = inputGroupContext !== null;
    // depending on browser direction, apply the appropriate styling.
    // when 'rtl' the flex row direction is reversed.
    const inputGroupPosition = inputGroupContext
        ? inputGroupContext.last
            ? isLtr
                ? 'groupRight'
                : 'groupLeft'
            : inputGroupContext.index === 0
                ? isLtr
                    ? 'groupLeft'
                    : 'groupRight'
                : 'groupMiddle'
        : undefined;
    const { variantClasses, styles: { textFieldContainerBase } } = useComponentTheme(TextFieldRedwoodTheme, {
        statusVariant,
        styleVariant,
        contentVariant,
        insideLabel: hasInsideLabel ? 'hasInsideLabel' : 'noInsideLabel',
        focused: isFocused ? 'isFocused' : 'notFocused',
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        loading: isLoading ? 'isLoading' : 'notLoading',
        resize: resize ?? 'none',
        withinGroup: hasGroupContext ? 'isWithinGroup' : undefined,
        inputGroupPosition
    });
    const { variantClasses: startContentStyles } = useComponentTheme(TextFieldRedwoodTheme, {
        startContent: hasInsideLabel ? 'hasInsideLabel' : 'noInsideLabel',
        startContentMargin: hasZeroStartMargin ? 'noStartMargin' : 'hasStartMargin',
        startContentDisabled: isDisabled ? 'isDisabled' : 'notDisabled'
    });
    const { variantClasses: middleStyles } = useComponentTheme(TextFieldRedwoodTheme, { middleContent: contentVariant });
    const { variantClasses: endContentStyles } = useComponentTheme(TextFieldRedwoodTheme, { endContent: hasInsideLabel ? 'hasInsideLabel' : 'noInsideLabel' });
    return (jsxs("div", { role: "presentation", class: classNames([variantClasses, textFieldContainerBase]), ref: rootRef, children: [startContent && (jsx("span", { class: startContentStyles, children: jsx(Flex, { justify: "center", align: "center", hasZeroMargins: true, children: startContent }) })), jsxs("div", { class: middleStyles, children: [insideLabel, mainContent] }), endContent && (jsx("span", { class: endContentStyles, children: jsx(Flex, { justify: "center", align: "center", hasZeroMargins: true, children: endContent }) }))] }));
};

const StyledTextField = ({ hasInsideLabel = false, variant = 'default', ...props }) => {
    // because of the mismatch of height for inputs with inside labels and no labels,
    // we'll have to set everything to the same min-height while within an input group context.
    // Here we'll look to see if it's present and apply the style as needed
    const groupContext = useInputGroupContext();
    const { isReadonly } = useFormFieldContext();
    const { baseTheme: formLayoutTheme } = useComponentTheme(FormLayoutRedwoodTheme);
    const { baseTheme: labelBaseTheme } = useComponentTheme(LabelRedwoodTheme);
    const { classes: textFieldClasses, styles } = useComponentTheme(TextFieldRedwoodTheme, {
        readonly: isReadonly ? 'isReadonly' : 'notReadonly',
        formControlInsideLabel: hasInsideLabel ? 'hasInsideLabel' : 'noInsideLabel',
        withinGroup: groupContext !== null ? 'isWithinGroup' : undefined
    });
    const { formControlBase, textFieldBase, textFieldEmbedded } = styles;
    // because TextField components reference Skeleton scoped vars, we need to bring in the Skeleton base theme
    // to pick up the scoped var definitions.
    const { baseTheme: skeletonBaseTheme } = useComponentTheme(SkeletonRedwoodTheme);
    const classes = classNames([
        skeletonBaseTheme,
        formControlBase,
        textFieldBase,
        variant === 'embedded' && textFieldEmbedded,
        textFieldClasses,
        formLayoutTheme,
        labelBaseTheme
    ]);
    return jsx("div", { ...props, class: classes });
};

// TODO: Do we need named props for start/endContent, or could they just be children that get
// wrapped in a flex layout?  (from review on 3/8/22)
// TODO: Instead of passing the inputElem, label, and userAssistance as named props here, should we
// have a separate layout component that lays them out, which could just be passed as a child
// to the TextField?  (from review on 3/8/22)
const TextField = ({ contentVariant, id, endContent, mainContent, startContent, compactUserAssistance, inlineUserAssistance, label, labelEdge, labelStartWidth, mainFieldRef, resize, statusVariant, styleVariant, hasZeroStartMargin, onFocus, onBlur, onKeyDown, onMouseDown, onMouseEnter, onMouseLeave }) => {
    // TODO: What are the implications for interoperability between new preact-based form components
    // and the existing JET oj-form-layout?  The preact form components create their own start/top
    // labels, while oj-form-layout expects to create those itself and wire them to the existing form
    // components through attributes like labelled-by.  (from review on 3/8/22)
    // TODO: How should a parent form component render component-specific content within TextField,
    // for example an aria live region?  Instead of a children prop on TextField, could the parent
    // component pass comp-specific content through other props, like inputElem?   The parent could
    // have its own component that composes inputElem with additional DOM.  (from review on 3/8/22)
    const isStartTop = label !== undefined && (labelEdge === 'start' || labelEdge === 'top');
    const textFieldContent = (jsx(TextFieldContent, { contentVariant: contentVariant, endContent: endContent, mainContent: mainContent, resize: resize, rootRef: mainFieldRef, startContent: startContent, statusVariant: statusVariant, styleVariant: styleVariant, hasZeroStartMargin: hasZeroStartMargin, ...(!isStartTop ? { insideLabel: label } : {}) }));
    const innerContent = (jsxs(Fragment, { children: [compactUserAssistance, textFieldContent, inlineUserAssistance] }));
    const styledTextFieldChildren = isStartTop ? (jsx(LabelValueLayout, { label: label, labelEdge: labelEdge, labelStartWidth: labelStartWidth, children: innerContent })) : (innerContent);
    return (jsx(StyledTextField, { id: id, hasInsideLabel: label !== undefined && labelEdge === 'inside', onfocusin: onFocus, onfocusout: onBlur, onKeyDown: onKeyDown, onMouseDown: onMouseDown, onMouseEnter: onMouseEnter, onMouseLeave: onMouseLeave, variant: styleVariant, children: styledTextFieldChildren }));
};

const MAX_LENGTH_UPDATE_DELAY = 500;
function MaxLengthLiveRegion({ isMaxLengthExceeded, maxLength, valueLength = 0 }) {
    // Comment copied from InputBase.js in Big JET:
    // Only update the aria-live div when the user has paused for more than
    // 500 milliseconds. That way, we avoid queued up aria-live messages which
    // would be annoying and not helpful. The 500ms was agreed upon in the
    // accessibility review meeting.
    const debouncedRemainingChars = useDebounce(maxLength - valueLength, MAX_LENGTH_UPDATE_DELAY);
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const maxLengthExceededTranslatedString = translations.formControl_maxLengthExceeded({
        MAX_LENGTH: `${maxLength}`
    });
    const remainingCharsTranslatedString = translations.formControl_maxLengthRemaining({
        CHARACTER_COUNT: `${debouncedRemainingChars}`
    });
    return (jsxs(Fragment$1, { children: [jsx(LiveRegion, { children: remainingCharsTranslatedString }), isMaxLengthExceeded && (jsx(LiveRegion, { type: "assertive", children: maxLengthExceededTranslatedString }))] }));
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

export { MaxLengthLiveRegion, ReadonlyTextField, TextField };
