// @oracle/oraclejet-preact/es/UNSAFE_Train.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { useCallback, useContext, useState } from 'preact/hooks';
import { createContext, toChildArray } from 'preact';
import { useComponentTheme } from './hooks/UNSAFE_useComponentTheme.js';
import { TrainRedwoodTheme } from './UNSAFE_Train/themes/redwood/TrainTheme.js';
import { usePress } from './hooks/UNSAFE_usePress.js';
import { useInteractionStyle } from './hooks/UNSAFE_useInteractionStyle.js';
import './utils/UNSAFE_classNames.js';
import { mergeProps } from './utils/UNSAFE_mergeProps.js';
import { c as classNames } from './classNames-f414daad.js';
import { MessageConfirmationIcon as SvgSuccessS } from './PRIVATE_ThemedIcons/MessageConfirmationIcon.js';
import { MessageErrorIcon as SvgErrorS } from './PRIVATE_ThemedIcons/MessageErrorIcon.js';
import { MessageInfoIcon as SvgInformationS } from './PRIVATE_ThemedIcons/MessageInfoIcon.js';
import { MessageWarningIcon as SvgWarningS } from './PRIVATE_ThemedIcons/MessageWarningIcon.js';
import { useActive } from './hooks/UNSAFE_useActive.js';
import { useTranslationBundle } from './hooks/UNSAFE_useTranslationBundle.js';
import './UNSAFE_HiddenAccessible.js';
import { H as HiddenAccessible } from './HiddenAccessible-6ca9b782.js';
import './utils/UNSAFE_logger.js';
import './_curry1-7629ff3f.js';
import './_curry2-820ed778.js';
import './UNSAFE_Train/themes/TrainStyles.css.js';
import './TrainStyles.styles.css';
import './UNSAFE_Train/themes/redwood/TrainBaseTheme.css.js';
import './UNSAFE_Train/themes/redwood/TrainBaseTheme.styles.css';
import './UNSAFE_Train/themes/redwood/TrainVariants.css.js';
import './TrainVariants.styles2.css';
import './vanilla-extract-recipes-createRuntimeFn.esm-3311d9ef.js';
import './utils/PRIVATE_clientHints.js';
import './clientHints-26cce6ec.js';
import './hooks/UNSAFE_useHover.js';
import './hooks/UNSAFE_useToggle.js';
import './UNSAFE_Icon.js';
import './Icon-e04c1503.js';
import './utils/UNSAFE_size.js';
import './UNSAFE_Theme.js';
import './utils-9a0cb8dc.js';
import './Common/themes/themeContract.css.js';
import './hooks/UNSAFE_useTooltip.js';
import './hooks/UNSAFE_useId.js';
import './UNSAFE_Floating.js';
import './Floating-2d909840.js';
import 'preact/compat';
import './index-326d7a5f.js';
import './hooks/UNSAFE_useUser.js';
import './index-f26154f2.js';
import './Common.js';
import './Common/themes.js';
import './Common/themes/redwood/theme.js';
import './utils/PRIVATE_floatingUtils.js';
import './utils/PRIVATE_refUtils.js';
import './hooks/UNSAFE_useOutsideClick.js';
import './utils/UNSAFE_arrayUtils.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-33365337.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './UNSAFE_Layer.js';
import './useThemeInterpolations-48ee20a7.js';
import './hooks/UNSAFE_useColorScheme.js';
import './hooks/UNSAFE_useScale.js';
import './utils/UNSAFE_interpolations/theme.js';
import './utils/UNSAFE_mergeInterpolations.js';
import './_curry3-bb17e901.js';
import './_has-057b6f90.js';
import './hooks/UNSAFE_useFocus.js';
import './hooks/UNSAFE_useTouch.js';
import './hooks/UNSAFE_useAnimation.js';
import './useAnimation-50c5f5fb.js';
import './hooks/UNSAFE_useThemeInterpolations.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './IconStyle.styles.css';


const Context = createContext(null);

/**
The JET Train component serves as a visual navigator enabling users to traverse through different 'steps'. Each 'step' is represented by the Step sub-component which can manifest its own state - 'visited', 'unvisited', or 'disabled', and is capable of communicating various message types including 'error', 'confirmation', 'warning', or 'info'.
 */
function Train({ onSelect, selectedStep, children }) {
    const { classes, styles } = useComponentTheme(TrainRedwoodTheme);
    const selectedStepIndex = children.findIndex((step) => step.props.id === selectedStep);
    const onStepSelect = useCallback((event, stepId) => {
        onSelect?.({
            event: event,
            toStep: stepId
        });
    }, [onSelect]);
    const trainContext = {
        onSelect: onStepSelect,
        selectedStep: selectedStep,
        selectedStepIndex: selectedStepIndex,
        numberOfSteps: children.length,
        index: 0
    };
    const childrenArray = toChildArray(children).filter(Boolean);
    return (jsx("div", { class: classes, children: childrenArray.map((step, index) => (jsx(Context.Provider, { value: { ...trainContext, index }, children: jsx("div", { class: styles.baseListStyle, children: step }) }))) }));
}

const Connector = ({ isConnected }) => {
    const { styles, variantClasses } = useComponentTheme(TrainRedwoodTheme, {
        connectorConnected: isConnected ? 'isConnected' : 'notConnected'
    });
    return jsx("span", { class: classNames([styles.connectorBaseStyle, variantClasses]) });
};

const messageTypeIcons = {
    confirmation: SvgSuccessS,
    error: SvgErrorS,
    fatal: SvgErrorS,
    info: SvgInformationS,
    warning: SvgWarningS
};
/**
 * UserChosenIcon Component for rendering the severity based icon in Message
 */
function UserChosenIcon({ messageType, isDisabled = false }) {
    const IconComponent = messageTypeIcons[messageType];
    const { styles, variantClasses } = useComponentTheme(TrainRedwoodTheme, {
        messageType: messageType,
        stepIconDisabled: isDisabled ? 'isDisabled' : 'notDisabled'
    });
    return (jsx("div", { class: classNames([styles.messageTypeIconBaseStyle, variantClasses]), role: "presentation", children: jsx(IconComponent, {}) }));
}

/**
 * Step icon
 */
const StepIcon = ({ isDisabled, isVisited, messageType }) => {
    const { index, selectedStepIndex } = useContext(Context);
    const isCurrent = selectedStepIndex === index;
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle } = useInteractionStyle(isDisabled || isCurrent);
    const { isActive, activeProps } = useActive({ isDisabled: isDisabled || isCurrent });
    console.log('isActive', isActive);
    const mergedProps = mergeProps(interactionProps, activeProps);
    const { styles, variantClasses } = useComponentTheme(TrainRedwoodTheme, {
        stepIconDisabled: isDisabled ? 'isDisabled' : 'notDisabled',
        stepIconCurrent: isCurrent ? 'isCurrent' : 'notCurrent',
        stepIconVisited: isVisited ? 'isVisited' : 'notVisited',
        stepIconNeedsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        stepIconPseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        stepIconActive: isActive ? 'isActive' : 'notActive'
    });
    const iconStyles = () => {
        if (messageType) {
            return jsx(UserChosenIcon, { messageType: messageType, isDisabled: isDisabled });
        }
        else {
            return (jsx("span", { ...mergedProps, class: classNames([styles.stepIconBaseStyle, variantClasses]), children: index + 1 }));
        }
    };
    return jsx("span", { class: styles.stepIconContainerStyle, children: iconStyles() });
};

/**
 * Step label component
 */
const StepLabel = ({ label, isDisabled, isVisited, messageType }) => {
    const { numberOfSteps, index, selectedStepIndex } = useContext(Context);
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const isCurrent = selectedStepIndex === index;
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle } = useInteractionStyle(isDisabled || isCurrent);
    const currentStepStatus = isCurrent
        ? translations.train_current()
        : isVisited
            ? translations.train_visited()
            : translations.train_not_visited();
    const currentStep = index + 1;
    const statusTextForCurrentStep = messageType
        ? `${translations.train_status({
            currentStep: currentStep.toString(),
            numberOfSteps: numberOfSteps.toString()
        })}. ${currentStepStatus}, ${translations.train_message_type()}: ${messageType}`
        : `${translations.train_status({
            currentStep: currentStep.toString(),
            numberOfSteps: numberOfSteps.toString()
        })}. ${currentStepStatus}`;
    const { styles, variantClasses } = useComponentTheme(TrainRedwoodTheme, {
        stepLabelDisabled: isDisabled ? 'isDisabled' : 'notDisabled',
        stepLabelCurrent: isCurrent ? 'isCurrent' : 'notCurrent',
        stepLabelVisited: isVisited ? 'isVisited' : 'notVisited',
        stepLabelNeedsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        stepLabelPseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover'
    });
    const handleClick = (event) => {
        event.preventDefault();
    };
    return (jsx("span", { class: styles.stepLabelWrapperStyle, children: jsxs("a", { ...interactionProps, href: "#", class: classNames([styles.stepLabelBaseStyle, variantClasses]), onClick: handleClick, children: [label, jsx(HiddenAccessible, { children: statusTextForCurrentStep })] }) }));
};

/**
 * Step component
 */
const Step = ({ id, isDisabled, isVisited, messageType, label }) => {
    const { onSelect, /* selectedStep ,*/ numberOfSteps, selectedStepIndex, index } = useContext(Context) || // get the context object or...
        (() => {
            // if there's no context object, throw an error
            throw new Error('Step component must be rendered within a Train component.');
        })();
    const isCurrent = selectedStepIndex === index;
    const [isActive, setIsActive] = useState(false);
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle } = useInteractionStyle(isDisabled || isCurrent);
    const isConnectorConnected = selectedStepIndex > index;
    const onSelectHandler = useCallback((event) => {
        if (!isDisabled && !isCurrent) {
            onSelect(event, id);
        }
    }, [isDisabled, onSelect, id]);
    const { pressProps } = usePress(onSelectHandler, {});
    const isLast = index !== numberOfSteps - 1;
    const onMouseUpHandler = useCallback(() => {
        setIsActive(false);
    }, [isActive]);
    const onMouseDownHandler = useCallback(() => {
        setIsActive(true);
    }, [isActive]);
    const { styles, variantClasses } = useComponentTheme(TrainRedwoodTheme, {
        stepDisabled: isDisabled ? 'isDisabled' : 'notDisabled',
        needsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        stepActive: isActive && !isCurrent ? 'isActive' : 'notActive'
    });
    const mergedProps = mergeProps(pressProps, interactionProps, {
        onMouseUp: onMouseUpHandler,
        onMouseDown: onMouseDownHandler
    });
    return (jsxs("span", { class: styles.stepContainerStyle, children: [isLast && jsx(Connector, { isConnected: isConnectorConnected }), jsxs("span", { ...mergedProps, class: classNames([styles.stepBaseStyle, variantClasses]), children: [jsx(StepIcon, { isVisited: isVisited, isDisabled: isDisabled, messageType: messageType }), jsx(StepLabel, { label: label, isVisited: isVisited, isDisabled: isDisabled, messageType: messageType })] })] }));
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

export { Step, Train };
