// @oracle/oraclejet-preact/es/UNSAFE_ExpandableList.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { Fragment } from 'preact';
import { useCallback, useState, useMemo } from 'preact/hooks';
import { memo } from 'preact/compat';
import "./UNSAFE_ExpandableList.css";
import './utils/UNSAFE_classNames.js';
import { useId } from './hooks/UNSAFE_useId.js';
import { useTranslationBundle } from './hooks/UNSAFE_useTranslationBundle.js';
import { containsKey, isSameKey } from './utils/UNSAFE_keys.js';
import './UNSAFE_Button.js';
import './UNSAFE_HiddenAccessible.js';
import './PRIVATE_ThemedIcons/ExpandIcon.js';
import { CollapseIcon } from './PRIVATE_ThemedIcons/CollapseIcon.js';
import { GROUP_SELECTOR, excludeGroup, ListGroupHeader, GroupLoadingIndicator } from './PRIVATE_List.js';
import { LiveRegion } from './UNSAFE_LiveRegion.js';
import { useUser } from './hooks/UNSAFE_useUser.js';
import { keyExtractor } from './utils/PRIVATE_collectionUtils.js';
import { L as List } from './List-1441b11d.js';
import { H as HiddenAccessible } from './HiddenAccessible-6ca9b782.js';
import { c as classNames } from './classNames-f414daad.js';
import { B as Button } from './Button-66829d09.js';
import { S as SvgChevronDown } from './index-8b1d59ff.js';
import './index-f26154f2.js';
import './Common.js';
import './Common/themes.js';
import './Common/themes/redwood/theme.js';
import './Common/themes/themeContract.css.js';
import './UNSAFE_BaseButton.js';
import './BaseButton-06d74a34.js';
import './hooks/UNSAFE_usePress.js';
import './hooks/UNSAFE_useHover.js';
import './hooks/UNSAFE_useToggle.js';
import './hooks/UNSAFE_useActive.js';
import './utils/UNSAFE_interpolations/dimensions.js';
import './utils/UNSAFE_arrayUtils.js';
import './utils/UNSAFE_size.js';
import './UNSAFE_Theme.js';
import './utils-9a0cb8dc.js';
import './_curry1-7629ff3f.js';
import './utils/UNSAFE_mergeInterpolations.js';
import './_curry3-bb17e901.js';
import './_curry2-820ed778.js';
import './_has-057b6f90.js';
import './utils/PRIVATE_clientHints.js';
import './clientHints-26cce6ec.js';
import './hooks/UNSAFE_useTabbableMode.js';
import './utils/UNSAFE_mergeProps.js';
import './UNSAFE_ButtonLayout.js';
import './ButtonLayout-8ea4b106.js';
import './UNSAFE_Text.js';
import './Text-0dad47bb.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './vanilla-extract-recipes-createRuntimeFn.esm-3311d9ef.js';
import './UNSAFE_Icon.js';
import './Icon-e04c1503.js';
import './hooks/UNSAFE_useTooltip.js';
import './UNSAFE_Floating.js';
import './Floating-2d909840.js';
import './index-326d7a5f.js';
import './utils/PRIVATE_floatingUtils.js';
import './utils/PRIVATE_refUtils.js';
import './hooks/UNSAFE_useOutsideClick.js';
import './hooks/UNSAFE_useComponentTheme.js';
import './utils/UNSAFE_logger.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-33365337.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './UNSAFE_Layer.js';
import './useThemeInterpolations-48ee20a7.js';
import './hooks/UNSAFE_useColorScheme.js';
import './hooks/UNSAFE_useScale.js';
import './utils/UNSAFE_interpolations/theme.js';
import './hooks/UNSAFE_useFocus.js';
import './hooks/UNSAFE_useTouch.js';
import './hooks/UNSAFE_useAnimation.js';
import './useAnimation-50c5f5fb.js';
import './hooks/UNSAFE_useThemeInterpolations.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './IconStyle.styles.css';
import './hoc/PRIVATE_withDirectionIcon.js';
import './hooks/UNSAFE_useViewportIntersect.js';
import './hooks/UNSAFE_useInteractionStyle.js';
import './utils/UNSAFE_interpolations/flexitem.js';
import './flexitem-89d8a00a.js';
import './keys-f766c992.js';
import './PRIVATE_List/themes/ListStyles.css.js';
import './ListStyles.styles.css';
import './UNSAFE_GroupedList/themes/redwood/GroupedListTheme.js';
import './UNSAFE_GroupedList/themes/GroupedListStyles.css.js';
import './GroupedListStyles.styles.css';
import './UNSAFE_GroupedList/themes/redwood/GroupedListBaseTheme.css.js';
import './UNSAFE_GroupedList/themes/redwood/GroupedListBaseTheme.styles.css';
import './SkeletonContainer-8f229433.js';
import './UNSAFE_Flex.js';
import './Flex-46adb0e4.js';
import './utils/UNSAFE_interpolations/boxalignment.js';
import './utils/UNSAFE_interpolations/flexbox.js';
import './flexbox-645317db.js';
import './UNSAFE_Skeleton.js';
import './Skeleton-18c022d9.js';
import './utils/UNSAFE_interpolations/borders.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './SkeletonStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './PRIVATE_LoadMoreCollection.js';
import './LoadMoreCollection-213ad68a.js';
import './PRIVATE_Collection.js';
import './Collection-cb68c0ee.js';
import './PRIVATE_VirtualizedCollection.js';
import './VirtualizedCollection-435128ef.js';
import './UNSAFE_FocusTrap.js';
import './FocusTrap-9dc31155.js';
import './utils/PRIVATE_tabbableUtils.js';
import './_arity-cb2c60f3.js';
import './hooks/PRIVATE_useSelection.js';
import './hooks/PRIVATE_useCurrentKey.js';
import './hooks/PRIVATE_useCollectionFocusRing.js';
import './hooks/PRIVATE_useTabbableModeSet.js';
import './hooks/PRIVATE_useItemAction.js';
import './UNSAFE_Selector.js';
import './PRIVATE_ThemedIcons/CheckboxOffIcon.js';
import './PRIVATE_Icons/CheckboxOff.js';
import './PRIVATE_ThemedIcons/CheckboxOnIcon.js';
import './PRIVATE_Icons/CheckboxOn.js';
import './PRIVATE_ThemedIcons/CheckboxMixedIcon.js';
import './PRIVATE_Icons/CheckboxMixed.js';
import './UNSAFE_Selector/themes/SelectorStyles.css.js';
import './SelectorStyles.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioIconTheme.js';
import './UNSAFE_Radio/themes/RadioIconStyles.css.js';
import './RadioIconStyles.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioIconBaseTheme.css.js';
import './UNSAFE_Radio/themes/redwood/RadioIconBaseTheme.styles.css';
import './UNSAFE_Radio/themes/redwood/RadioIconVariants.css.js';
import './RadioIconVariants.styles.css';
import './hooks/PRIVATE_useCollectionGestureContext.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * A hook that handles mouse and keyboard gesture that toggles the expanded state.
 * @param keyExtractor
 * @param expanded
 * @param onToggle
 * @returns
 */
function useExpandCollapse(keyExtractor, expanded, onToggle) {
    const { direction } = useUser();
    const isRtl = direction === 'rtl';
    const onClick = useCallback((event) => {
        handleToggle(event, keyExtractor, onToggle);
    }, [keyExtractor, onToggle]);
    const onKeyUp = useCallback((event) => {
        if (event.key === 'Enter' || event.key === ' ') {
            handleToggle(event, keyExtractor, onToggle);
        }
        else if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
            const isLeftArrow = event.key === 'ArrowLeft' && !isRtl;
            handleArrowKey(isLeftArrow, event.target, expanded, keyExtractor, onToggle);
            // arrow key might cause it to scroll horizontally
            event.preventDefault();
            event.stopPropagation();
        }
    }, [keyExtractor, onToggle]);
    return onToggle ? { onClick, onKeyDown, onKeyUp } : {};
}
// prevent default (propagation) for keyDown to prevent scrolling the container
const onKeyDown = (event) => {
    if (event.key === 'Enter' || event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
        event.preventDefault();
    }
};
const handleArrowKey = (isLeftArrow, target, expanded, keyExtractor, onToggle) => {
    const itemKey = keyExtractor(target);
    if (onToggle && itemKey) {
        const contains = containsKey(expanded, itemKey);
        if ((isLeftArrow && contains) || (!isLeftArrow && !contains)) {
            onToggle({ value: itemKey });
        }
    }
};
const handleToggle = (event, keyExtractor, onToggle) => {
    const itemKey = keyExtractor(event.target);
    if (onToggle && itemKey) {
        onToggle({ value: itemKey });
        if (event.type === 'click') {
            event.stopPropagation();
        }
    }
};

/**
 * An ExpandableList shows a two-level group of items where each group header can be expand or collapse.
 */
function ExpandableList({
  children,
  currentKey,
  data,
  onSelectionChange,
  onToggle,
  ...props
}) {
  const [statusText, setStatusText] = useState('');
  const translations = useTranslationBundle('@oracle/oraclejet-preact');
  const expandCollapseProps = useExpandCollapse(element => {
    if (element.getAttribute('role') === 'treegrid') {
      return currentKey === undefined ? null : currentKey;
    }
    return keyExtractor(element, GROUP_SELECTOR);
  }, data == null ? emptyKeys : data.expanded, onToggle ? useCallback(detail => {
    onToggle(detail);
    if (data) {
      const previouslyExpanded = containsKey(data?.expanded, detail.value);
      setStatusText(previouslyExpanded ? translations.expandableList_groupCollapse() : translations.expandableList_groupExpand());
      // we need to clear the live region text otherwise screenreader
      // won't read again if the content did not change the next time
      setTimeout(() => {
        setStatusText('');
      }, 1000);
    }
  }, [onToggle, data]) : undefined);
  const descId = useId();
  const stableDescId = useMemo(() => {
    // this should never change after mounted
    return descId;
  }, []);
  const createHierarchicalContext = (context, options) => {
    const expanderRenderer = options !== undefined ? () => jsx(Expander, {
      itemKey: context.metadata.key,
      expanded: options.expanded,
      onToggle: options.onToggle
    }) : undefined;
    return {
      ...context,
      parentKey: context.metadata.parentKey,
      leaf: context.metadata.isLeaf ?? true,
      depth: context.metadata.treeDepth ?? 1,
      expander: expanderRenderer
    };
  };
  const handleSelectionChange = useCallback(detail => {
    if (onSelectionChange && data) {
      const keys = excludeGroup(data, detail.value);
      if (isSameKey(keys, detail.value)) {
        onSelectionChange(detail);
      } else if (keys.keys && keys.keys.size > 0) {
        onSelectionChange({
          ...detail,
          value: keys
        });
      }
    }
  }, [onSelectionChange, data]);
  const customItemRenderer = context => {
    const itemKey = context.listItemContext.metadata.key;
    const options = data && onToggle ? {
      expanded: containsKey(data.expanded, itemKey),
      onToggle
    } : undefined;
    const itemContext = createHierarchicalContext(context.listItemContext, options);
    if (itemContext.leaf) {
      return context.defaultListItem(itemContext, {
        itemDepth: itemContext.depth + 1
      });
    } else {
      return jsx(ListGroupHeader, {
        ariaDescribedBy: stableDescId,
        itemKey: itemContext.metadata.key,
        itemIndex: itemContext.index,
        itemDepth: itemContext.depth,
        isFocused: context.isFocused,
        isFocusRingVisible: context.isFocusRingVisible,
        isActive: context.isActive,
        isGridlineVisible: false,
        isExpandable: true,
        expandedKeys: data == null ? emptyKeys : data.expanded,
        onToggle: onToggle,
        children: children(itemContext)
      }, itemContext.metadata.key);
    }
  };
  return jsxs(Fragment, {
    children: [jsx(List, {
      isVirtualized: false,
      currentKey: currentKey,
      data: data,
      role: "treegrid",
      loadingIndicator: jsx(GroupLoadingIndicator, {}),
      onSelectionChange: handleSelectionChange,
      customItemRenderer: customItemRenderer,
      eventHandlers: expandCollapseProps,
      ...props,
      children: context => {
        return children(context);
      }
    }), onToggle && jsxs(Fragment, {
      children: [jsx(LiveRegion, {
        children: statusText
      }), jsx("span", {
        id: stableDescId,
        children: jsx(HiddenAccessible, {
          children: translations.expandableList_expandCollapseInstructionText()
        })
      })]
    })]
  });
}
const emptyKeys = {
  all: false,
  keys: new Set()
};
const Expander = memo(props => {
  const styles = {
    base: "o609kf"
  };
  const classes = classNames([styles.base, 'oj-listview-expander']);
  const actionHandler = () => {
    props.onToggle({
      value: props.itemKey
    });
  };
  return jsx("span", {
    class: classes,
    role: "presentation",
    children: jsx(Button, {
      onAction: actionHandler,
      variant: "ghost",
      startIcon: props.expanded ? jsx(SvgChevronDown, {
        size: "6x"
      }) : jsx(CollapseIcon, {
        size: "6x"
      })
    })
  });
}, (prev, next) => {
  return prev.itemKey === next.itemKey && prev.expanded === next.expanded && prev.onToggle === next.onToggle;
});

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

export { ExpandableList };
