// @oracle/oraclejet-preact/es/UNSAFE_Dialog.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { useState, useEffect } from 'preact/hooks';
import { forwardRef } from 'preact/compat';
import './hooks/UNSAFE_useAnimation.js';
import { useComponentTheme } from './hooks/UNSAFE_useComponentTheme.js';
import { Layer } from './UNSAFE_Layer.js';
import './UNSAFE_Modal.js';
import './UNSAFE_FocusTrap.js';
import { WindowOverlay } from './UNSAFE_WindowOverlay.js';
import { useId } from './hooks/UNSAFE_useId.js';
import { DialogRedwoodTheme } from './UNSAFE_Dialog/themes/redwood/DialogTheme.js';
import { u as useAnimation } from './useAnimation-50c5f5fb.js';
import { F as FocusTrap } from './FocusTrap-9dc31155.js';
import { M as Modal } from './Modal-33f98fb6.js';
import './utils/UNSAFE_logger.js';
import './_curry1-7629ff3f.js';
import './_curry2-820ed778.js';
import './index-f26154f2.js';
import 'preact';
import './Common.js';
import './Common/themes.js';
import './Common/themes/redwood/theme.js';
import './Common/themes/themeContract.css.js';
import './useThemeInterpolations-48ee20a7.js';
import './hooks/UNSAFE_useColorScheme.js';
import './hooks/UNSAFE_useScale.js';
import './utils/UNSAFE_interpolations/theme.js';
import './UNSAFE_Theme.js';
import './utils-9a0cb8dc.js';
import './utils/UNSAFE_mergeInterpolations.js';
import './utils/UNSAFE_classNames.js';
import './classNames-f414daad.js';
import './_curry3-bb17e901.js';
import './_has-057b6f90.js';

import './utils/PRIVATE_tabbableUtils.js';
import './_arity-cb2c60f3.js';
import './vanilla-extract-dynamic.esm-33365337.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './hooks/UNSAFE_useUser.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-3311d9ef.js';
import './UNSAFE_Dialog/themes/DialogStyles.css.js';
import './DialogStyles.styles.css';
import './UNSAFE_Dialog/themes/redwood/DialogBaseTheme.css.js';
import './UNSAFE_Dialog/themes/redwood/DialogBaseTheme.styles.css';
import './UNSAFE_Dialog/themes/redwood/DialogVariants.css.js';
import './useModal-e580c081.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';

const StyledDialogHeader = ({ header, id }) => {
    const { styles } = useComponentTheme(DialogRedwoodTheme);
    const content = typeof header === 'string' ? (jsx("h1", { id: id, className: styles.dialogTitleStyle, children: header })) : (header);
    return (jsx("div", { "aria-labelledby": id, className: styles.dialogHeaderStyle, children: content }));
};
const StyledDialog = forwardRef(({ children, header, footer, labelId, role, ariaLabelledBy, ariaDescribedBy, onKeyDown }, ref) => {
    const { classes, styles } = useComponentTheme(DialogRedwoodTheme);
    return (jsx("div", { ref: ref, className: classes, style: { opacity: 0 }, onKeyDown: onKeyDown, role: role, "aria-labelledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy, children: jsxs("div", { className: styles.dialogContainerStyle, children: [jsx(StyledDialogHeader, { id: labelId, header: header }), jsx("div", { className: styles.dialogContentStyle, children: jsx("div", { className: styles.dialogBodyStyle, children: children }) }), jsx("div", { className: styles.dialogFooterStyle, children: footer })] }) }));
});
// main Dialog component
/**
 * WAI-ARIA-compliant dialog component. A dialog is a floating window that typically contains
 * a title bar and a content area.
 * The dialog is generally placed on top of the rest of the page content. It can be either modal
 * (only the content in the dialog can be interacted with) or non-modal (it's still possible
 * to interact with content outside of the dialog).
 */
const Dialog = ({ autoFocusRef, children, isOpen, header = '', footer, modality = 'modal', placement = 'center', offset = 0, onClose, onTransitionEnd, role = 'dialog', 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy }) => {
    const [state, setState] = useState(isOpen ? 'opening' : 'unmounted');
    const uniqueId = useId();
    const labelId = ariaLabelledBy ? ariaLabelledBy : uniqueId;
    useEffect(() => {
        // Ignore state update on initial render
        if (state === 'unmounted' && !isOpen) {
            return;
        }
        setState(isOpen ? 'opening' : 'closing');
    }, [isOpen, state]);
    const { nodeRef } = useAnimation(state, {
        animationStates,
        isAnimatedOnMount: true,
        onAnimationEnd: ({ animationState }) => {
            if (animationState === 'closing') {
                setState('unmounted');
                onTransitionEnd?.(false);
            }
            else {
                onTransitionEnd?.(true);
            }
        }
    });
    const handleKeyDown = (event) => {
        switch (event.code) {
            case 'Escape': {
                onClose?.({ reason: 'escapeKey' });
                break;
            }
        }
    };
    const renderDialog = () => {
        return (jsx(WindowOverlay, { placement: placement, offset: offset, children: jsx(FocusTrap, { autoFocusRef: autoFocusRef, children: jsx(StyledDialog, { ref: nodeRef, labelId: labelId, header: header, footer: footer, role: role, ariaLabelledBy: labelId, ariaDescribedBy: ariaDescribedBy, onKeyDown: handleKeyDown, children: children }) }) }));
    };
    if (modality === 'modal') {
        return jsx(Modal, { isOpen: state !== 'unmounted', children: renderDialog() });
    }
    else {
        return state !== 'unmounted' ? jsx(Layer, { children: renderDialog() }) : null;
    }
};
const animationStates = {
    opening: {
        from: {
            scaleX: 0,
            scaleY: 0,
            transformOrigin: 'center',
            opacity: 0
        },
        to: {
            scaleX: 1,
            scaleY: 1,
            opacity: 1
        },
        options: {
            duration: 200
        }
    },
    closing: {
        from: {
            scaleX: 1,
            scaleY: 1,
            transformOrigin: 'center',
            opacity: 1
        },
        to: {
            scaleX: 0,
            scaleY: 0,
            opacity: 0
        },
        options: {
            duration: 200
        }
    }
};

export { Dialog };
