// @oracle/oraclejet-preact/cjs/UNSAFE_Radio/themes/RadioStyles.css.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });



var radioBaseStyle = 'RadioStyles_radioBaseStyle__rjrvbk0';
var radioDefaultColor = 'RadioStyles_radioDefaultColor__rjrvbk3';
var radioDisabledColor = 'RadioStyles_radioDisabledColor__rjrvbk4';
var styles = {radioBaseStyle:'RadioStyles_radioBaseStyle__rjrvbk0',valueLabelStyle:'RadioStyles_valueLabelStyle__rjrvbk1',valueTextStyle:'RadioStyles_valueTextStyle__rjrvbk2'};
var valueLabelStyle = 'RadioStyles_valueLabelStyle__rjrvbk1';
var valueTextStyle = 'RadioStyles_valueTextStyle__rjrvbk2';
var variants = {disabled:{isDisabled:'RadioStyles_radioDisabledColor__rjrvbk4',notDisabled:'RadioStyles_radioDefaultColor__rjrvbk3'},focused:{isFocused:'RadioStyles_radioFocused__rjrvbk5'}};

exports.radioBaseStyle = radioBaseStyle;
exports.radioDefaultColor = radioDefaultColor;
exports.radioDisabledColor = radioDisabledColor;
exports.styles = styles;
exports.valueLabelStyle = valueLabelStyle;
exports.valueTextStyle = valueTextStyle;
exports.variants = variants;
