// @oracle/oraclejet-preact/cjs/Common/themes/redwood/theme.base.tokens.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('../../../UNSAFE_Theme.js');
var Common_themes_themeContract_css = require('../themeContract.css.js');
var utils = require('../../../utils-b9f57550.js');

/**
 * These theme tokens are manually updated and are not generated from Figma tokens. These will eventually be generated.
 */
const base = {
    cursor: {
        clickable: 'pointer'
    },
    boxShadow: {
        xs: `rgba(${Common_themes_themeContract_css.colorSchemeVars.boxshadow.shadowcolor}, .12) 0 ${utils.xUnits(0.25)} ${utils.xUnits(1)} 0`,
        sm: `rgba(${Common_themes_themeContract_css.colorSchemeVars.boxshadow.shadowcolor}, .16) 0 ${utils.xUnits(1)} ${utils.xUnits(2)} 0`,
        md: `rgba(${Common_themes_themeContract_css.colorSchemeVars.boxshadow.shadowcolor}, .2) 0 ${utils.xUnits(1.5)} ${utils.xUnits(3)} 0`,
        lg: `rgba(${Common_themes_themeContract_css.colorSchemeVars.boxshadow.shadowcolor}, .24) 0 ${utils.xUnits(2)} ${utils.xUnits(4)} 0`,
        xl: `rgba(${Common_themes_themeContract_css.colorSchemeVars.boxshadow.shadowcolor}, .28) 0 ${utils.xUnits(3)} ${utils.xUnits(5)} 0`
    },
    borderRadius: {
        sm: utils.xUnits(0.5),
        md: utils.xUnits(1),
        lg: utils.xUnits(1.5),
        xl: utils.xUnits(2)
    }
};
const collection = {
    listRowHeight: utils.xUnits(12),
    listCellPaddingVertical: utils.xUnits(3),
    sparkleImg: 'url(./images/AI-Sparkle.gif), url(../../css/redwood/images/AI-Sparkle.gif)'
};
const dvt = {
    label: {
        sm: {
            fontSize: Common_themes_themeContract_css.typographyVars.body.xs.fontSize
        },
        md: {
            fontSize: Common_themes_themeContract_css.typographyVars.body.sm.fontSize
        },
        lg: {
            fontSize: Common_themes_themeContract_css.typographyVars.body.md.fontSize
        }
    }
};

exports.base = base;
exports.collection = collection;
exports.dvt = dvt;
