// @oracle/oraclejet-preact/cjs/ComponentMessage-12850cb4.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var hooks_UNSAFE_useComponentTheme = require('./hooks/UNSAFE_useComponentTheme.js');
var hooks_UNSAFE_useTranslationBundle = require('./hooks/UNSAFE_useTranslationBundle.js');
require('./PRIVATE_Message.js');
require('./UNSAFE_HiddenAccessible.js');
var UNSAFE_ComponentMessage_themes_redwood_ComponentMessageTheme = require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js');
var HiddenAccessible = require('./HiddenAccessible-c4ad96e7.js');
var MessageUtils = require('./MessageUtils-ff85758f.js');
var MessageStartIcon = require('./MessageStartIcon-8a59b0b6.js');
var MessageSummary = require('./MessageSummary-5d0c793a.js');

/**
 * The component that renders an individual message for inline component messaging.
 */
function ComponentMessage({ detail, fieldLabel, severity = 'error' }) {
    const { classes } = hooks_UNSAFE_useComponentTheme.useComponentTheme(UNSAFE_ComponentMessage_themes_redwood_ComponentMessageTheme.ComponentMessageRedwoodTheme, {
        severity: severity ?? 'none'
    });
    const translations = hooks_UNSAFE_useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const severityTranslations = {
        confirmation: translations.confirmation(),
        error: translations.error(),
        info: translations.info(),
        warning: translations.warn()
    };
    return (jsxRuntime.jsxs("div", { class: classes, role: "alert", "aria-atomic": "true", children: [fieldLabel && jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { children: fieldLabel }), MessageUtils.isSeverityIconNeeded(severity) && (jsxRuntime.jsx(MessageStartIcon.MessageStartIcon, { variant: "inline", severity: severity, translations: severityTranslations })), jsxRuntime.jsx(MessageSummary.MessageSummary, { variant: "inline", text: detail })] }));
}

exports.ComponentMessage = ComponentMessage;
