// @oracle/oraclejet-preact/cjs/ComponentMessageContainer-93be9be1.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
require('./PRIVATE_Message.js');
var ComponentMessage = require('./ComponentMessage-12850cb4.js');
var UNSAFE_ComponentMessage_themes_ComponentMessageStyles_css = require('./UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js');
var MessagesManager = require('./MessagesManager-411c2e38.js');

/**
 * Converts the messages data into Item
 *
 * @param messages The messages data passed down as prop
 * @returns The messages data as Item
 */
function generateMessagesData(messages) {
    return messages.map((message, index) => {
        return {
            key: index,
            data: {
                closeAffordance: 'off',
                severity: message.severity || 'error',
                detail: message.detail
            }
        };
    });
}
function ComponentMessageContainer({ fieldLabel, messages = [] }) {
    return (jsxRuntime.jsx("div", { class: UNSAFE_ComponentMessage_themes_ComponentMessageStyles_css.componentMessageContainerBase, children: jsxRuntime.jsx(MessagesManager.MessagesManager, { data: generateMessagesData(messages), children: ({ item }) => (jsxRuntime.jsx(ComponentMessage.ComponentMessage, { detail: item.data.detail, fieldLabel: fieldLabel, severity: item.data.severity }, item.key)) }) }));
}

exports.ComponentMessageContainer = ComponentMessageContainer;
