// @oracle/oraclejet-preact/cjs/Heading-c5069994.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var utils_UNSAFE_arrayUtils = require('./utils/UNSAFE_arrayUtils.js');
var UNSAFE_Heading_themes_HeadingStyles_css = require('./UNSAFE_Heading/themes/HeadingStyles.css.js');

const headingElementTypes = utils_UNSAFE_arrayUtils.stringLiteralArray(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']);
const headingsizeMap = {
    h1: '2xl',
    h2: 'xl',
    h3: 'lg',
    h4: 'md',
    h5: 'sm',
    h6: 'xs'
};
const sizeOptions = ['xs', 'sm', 'md', 'lg', 'xl', '2xl', 'inherit'];
function Heading({ children, as, size, id, variant }) {
    const Component = as;
    //@ts-ignore
    size = size ?? headingsizeMap[Component];
    const classes = UNSAFE_Heading_themes_HeadingStyles_css.multiVariantStyles({ size, variant });
    return (jsxRuntime.jsx(Component, { className: classes, id: id, children: children }));
}

exports.Heading = Heading;
exports.headingElementTypes = headingElementTypes;
exports.sizeOptions = sizeOptions;
