// @oracle/oraclejet-preact/cjs/menuTestsUtils-1c2805f7.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

function getIcon(menuItem, selector) {
    return menuItem.querySelector(selector);
}
function getAllMenuItemsByRole(content, role, hidden = true) {
    return content.getAllByRole(role, { hidden });
}
function getMenuItemByRole(content, role, hidden = true) {
    return content.getByRole(role, { hidden });
}
function getLabel(content, label) {
    return content.getByText(label);
}
function setObserverConfig() {
    if (!window.ResizeObserver) {
        // ResizeObserver isn't available in test environment
        const MockResizeObserver = jest.fn();
        MockResizeObserver.mockReturnValue({
            observe: () => null,
            unobserve: () => null,
            disconnect: () => null
        });
        Object.defineProperty(global, 'ResizeObserver', {
            writable: true,
            configurable: true,
            value: MockResizeObserver
        });
    }
}

exports.getAllMenuItemsByRole = getAllMenuItemsByRole;
exports.getIcon = getIcon;
exports.getLabel = getLabel;
exports.getMenuItemByRole = getMenuItemByRole;
exports.setObserverConfig = setObserverConfig;
