// @oracle/oraclejet-preact/cjs/PRIVATE_SelectCommon.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
require("identity-obj-proxy");
require('./UNSAFE_HighlightText.js');
require('./PRIVATE_List.js');
require('./UNSAFE_Flex.js');
require('./UNSAFE_Skeleton.js');
var utils_UNSAFE_logger = require('./utils/UNSAFE_logger.js');
var Flex = require('./Flex-2eb09dad.js');
var HighlightText = require('./HighlightText-445b5dcb.js');
var List = require('./List-467fca47.js');
var SkeletonContainer = require('./SkeletonContainer-dbb3ab5e.js');
var Skeleton = require('./Skeleton-506420e5.js');
require('./UNSAFE_Floating.js');
var UNSAFE_Layer = require('./UNSAFE_Layer.js');
var PRIVATE_SelectCommon_themes_dropdownStyles_css = require('./PRIVATE_SelectCommon/themes/dropdownStyles.css.js');
var utils_UNSAFE_arrayUtils = require('./utils/UNSAFE_arrayUtils.js');
var Floating = require('./Floating-3415cf36.js');
var hooks_UNSAFE_usePress = require('./hooks/UNSAFE_usePress.js');
var PRIVATE_ThemedIcons_DropdownArrowIcon = require('./PRIVATE_ThemedIcons/DropdownArrowIcon.js');
require('./utils/UNSAFE_classNames.js');
var classNames = require('./classNames-d619a292.js');
var hooks_PRIVATE_useCollectionGestureContext = require('./hooks/PRIVATE_useCollectionGestureContext.js');
var hooks_UNSAFE_useFormFieldContext = require('./hooks/UNSAFE_useFormFieldContext.js');
var UNSAFE_UserAssistance = require('./UNSAFE_UserAssistance.js');
var hooks_UNSAFE_useTranslationBundle = require('./hooks/UNSAFE_useTranslationBundle.js');
var UNSAFE_WindowOverlay = require('./UNSAFE_WindowOverlay.js');
var PRIVATE_SelectCommon_themes_mobileDropdownStyles_css = require('./PRIVATE_SelectCommon/themes/mobileDropdownStyles.css.js');
require('./UNSAFE_Divider.js');
var PRIVATE_SelectCommon_themes_selectMobileDropdownStyles_css = require('./PRIVATE_SelectCommon/themes/selectMobileDropdownStyles.css.js');
var Divider = require('./Divider-ef0eee24.js');
require('./utils/PRIVATE_clientHints.js');
var clientHints = require('./clientHints-b9e7be1e.js');
require('./utils/UNSAFE_keys.js');
require('./hooks/UNSAFE_useId.js');
require('./hooks/UNSAFE_useViewportIntersect.js');
require('./hooks/UNSAFE_useTabbableMode.js');
require('preact');
require('./hooks/UNSAFE_useComponentTheme.js');
require('./_curry1-6e9511bf.js');
require('./_curry2-53a902bf.js');
require('./hooks/UNSAFE_useInteractionStyle.js');
require('./utils/UNSAFE_mergeProps.js');
require('./hooks/UNSAFE_useHover.js');
require('./hooks/UNSAFE_useToggle.js');
require('./hooks/UNSAFE_useActive.js');
require('./utils/UNSAFE_mergeInterpolations.js');
require('./_curry3-ad2313fd.js');
require('./_has-aa203696.js');
require('./utils/UNSAFE_interpolations/flexitem.js');
require('./flexitem-939ec1ef.js');
require('./keys-65201aa4.js');
require('./PRIVATE_List/themes/ListStyles.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-8a4b93e4.js');
require('./UNSAFE_GroupedList/themes/redwood/GroupedListTheme.js');
require('./UNSAFE_GroupedList/themes/GroupedListStyles.css.js');

require('./UNSAFE_GroupedList/themes/redwood/GroupedListBaseTheme.css.js');

require('./PRIVATE_LoadMoreCollection.js');
require('./LoadMoreCollection-78b142a9.js');
require('./PRIVATE_Collection.js');
require('./Collection-f060a1e2.js');
require('./PRIVATE_VirtualizedCollection.js');
require('./VirtualizedCollection-fe629df3.js');
require('preact/compat');
require('./UNSAFE_FocusTrap.js');
require('./FocusTrap-2209a89f.js');
require('./utils/PRIVATE_tabbableUtils.js');
require('./_arity-353457ba.js');
require('./hooks/PRIVATE_useSelection.js');
require('./hooks/UNSAFE_useUser.js');
require('./index-c3ee789d.js');
require('./Common.js');
require('./Common/themes.js');
require('./Common/themes/redwood/theme.js');
require('./Common/themes/themeContract.css.js');
require('./utils/PRIVATE_collectionUtils.js');
require('./hooks/PRIVATE_useCurrentKey.js');
require('./hooks/PRIVATE_useCollectionFocusRing.js');
require('./hooks/PRIVATE_useTabbableModeSet.js');
require('./hooks/PRIVATE_useItemAction.js');
require('./UNSAFE_Selector.js');
require('./PRIVATE_ThemedIcons/CheckboxOffIcon.js');
require('./PRIVATE_Icons/CheckboxOff.js');
require('./UNSAFE_Icon.js');
require('./Icon-e47442ad.js');
require('./utils/UNSAFE_size.js');
require('./UNSAFE_Theme.js');
require('./utils-b9f57550.js');
require('./hooks/UNSAFE_useTooltip.js');
require('./hooks/UNSAFE_useFocus.js');
require('./hooks/UNSAFE_useTouch.js');
require('./hooks/UNSAFE_useAnimation.js');
require('./useAnimation-802f3cdb.js');
require('./hooks/UNSAFE_useThemeInterpolations.js');
require('./useThemeInterpolations-ee4e7934.js');
require('./hooks/UNSAFE_useColorScheme.js');
require('./hooks/UNSAFE_useScale.js');
require('./utils/UNSAFE_interpolations/theme.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./index-0a166c9e.js');
require('./utils/PRIVATE_floatingUtils.js');
require('./utils/PRIVATE_refUtils.js');
require('./hooks/UNSAFE_useOutsideClick.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-4bc33829.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');

require('./PRIVATE_ThemedIcons/CheckboxOnIcon.js');
require('./PRIVATE_Icons/CheckboxOn.js');
require('./PRIVATE_ThemedIcons/CheckboxMixedIcon.js');
require('./PRIVATE_Icons/CheckboxMixed.js');
require('./UNSAFE_Selector/themes/SelectorStyles.css.js');

require('./UNSAFE_Radio/themes/redwood/RadioIconTheme.js');
require('./UNSAFE_Radio/themes/RadioIconStyles.css.js');

require('./UNSAFE_Radio/themes/redwood/RadioIconBaseTheme.css.js');

require('./UNSAFE_Radio/themes/redwood/RadioIconVariants.css.js');

require('./utils/UNSAFE_interpolations/dimensions.js');
require('./utils/UNSAFE_interpolations/boxalignment.js');
require('./utils/UNSAFE_interpolations/flexbox.js');
require('./flexbox-8cd3009d.js');
require('./utils/UNSAFE_interpolations/borders.js');
require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');

require('./InlineHelpSource-ecc4f96f.js');
require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js');
require('./UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js');

require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js');

require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js');
require('./UNSAFE_ComponentMessage.js');
require('./ComponentMessage-12850cb4.js');
require('./PRIVATE_Message.js');
require('./utils/PRIVATE_timer.js');
require('./PRIVATE_Message/themes/MessageStyles.css.js');

require('./MessageCloseButton-64c284a2.js');
require('./UNSAFE_IconButton.js');
require('./IconButton-bc45151b.js');
require('./UNSAFE_BaseButton.js');
require('./BaseButton-934b844e.js');
require('./UNSAFE_ButtonLayout.js');
require('./ButtonLayout-648ecbd5.js');
require('./UNSAFE_Text.js');
require('./Text-4f1f698a.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./PRIVATE_ThemedIcons/CloseIcon.js');
require('./MessageDetail-dce2aadd.js');
require('./MessageFormattingUtils-3cf76724.js');
require('./utils/UNSAFE_getLocale.js');
require('./utils/UNSAFE_stringUtils.js');
require('./stringUtils-27007290.js');
require('./Message.types-20efdde7.js');
require('./MessageStartIcon-8a59b0b6.js');
require('./PRIVATE_ThemedIcons/MessageConfirmationIcon.js');
require('./PRIVATE_ThemedIcons/MessageErrorIcon.js');
require('./PRIVATE_ThemedIcons/MessageInfoIcon.js');
require('./PRIVATE_ThemedIcons/MessageWarningIcon.js');
require('./MessageSummary-5d0c793a.js');
require('./MessageTimestamp-eab05143.js');
require('./MessageUtils-ff85758f.js');
require('./utils/UNSAFE_soundUtils.js');
require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerTheme.js');
require('./UNSAFE_MessageBanner/themes/MessageBannerStyles.css.js');

require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.css.js');

require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js');

require('./MessagesManager-411c2e38.js');
require('./PRIVATE_TransitionGroup.js');
require('./hooks/UNSAFE_useMessagesContext.js');
require('./UNSAFE_HiddenAccessible.js');
require('./HiddenAccessible-c4ad96e7.js');
require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js');
require('./UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js');

require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js');

require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js');
require('./ComponentMessageContainer-93be9be1.js');
require('./hooks/UNSAFE_useFormContext.js');
require('./hooks/UNSAFE_useInputGroupContext.js');
require('./UNSAFE_Popup.js');
require('./hooks/PRIVATE_useAnimationStatus.js');
require('./UNSAFE_Popup/themes/PopupContract.css.js');
require('./UNSAFE_Popup/themes/redwood/PopupTheme.js');
require('./UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js');

require('./UNSAFE_Popup/themes/redwood/PopupVariants.css.js');

require('./CompactHelpSource-c5fd0d41.js');
require('./PRIVATE_ThemedIcons/HelpIcon.js');
require('./hooks/UNSAFE_useFocusWithin.js');
require('./useFocusWithin-3c76895a.js');
require('./UNSAFE_Label/themes/redwood/LabelTheme.js');
require('./UNSAFE_Label/themes/LabelStyles.css.js');

require('./UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js');

require('./UNSAFE_Label/themes/redwood/LabelVariants.css.js');
require('./UNSAFE_Divider/themes/DividerStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');



/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function renderItemText(item, itemText) {
    // if item-text is a string and the data has the specified field, return it;
    // otherwise log an error
    if (typeof itemText === 'string') {
        if (item.data?.hasOwnProperty(itemText)) {
            const retData = item.data[itemText];
            return String(retData);
        }
        utils_UNSAFE_logger.Logger.error('SelectMultiple: itemText specifies field that does not exist: ' + itemText);
        // return undefined if the field doesn't exist
        return undefined;
    }
    // if item-text is a function, call it to create a display label
    return itemText(item);
}

const itemStyles = "_1eswtdk";
const noOp = () => {};
function DefaultList({
  'aria-label': ariaLabel,
  currentItemVariant,
  currentKey,
  data,
  itemRenderer,
  itemText,
  onCurrentKeyChange,
  onItemAction,
  onLoadRange = noOp,
  onSelectionChange,
  searchText,
  selectedKeys,
  selectionMode
}) {
  const listItemRenderer = hooks.useCallback(listItemContext => {
    // specify the listItemContext props that we want to expose individually instead of
    // spreading the whole object so that we don't automatically include any new props that
    // may be added to it
    const itemRendererProps = {
      data: listItemContext.data,
      index: listItemContext.index,
      metadata: listItemContext.metadata,
      selector: listItemContext.selector,
      searchText,
      selectedKeys,
      onSelectionChange: ({
        target,
        value
      }) => {
        onSelectionChange?.({
          target,
          value: {
            all: false,
            keys: value
          }
        });
      }
    };
    const itemContext = {
      data: listItemContext.data,
      key: listItemContext.metadata.key,
      metadata: listItemContext.metadata
    };
    return itemRenderer ? itemRenderer(itemRendererProps) : jsxRuntime.jsxs(Flex.Flex, {
      children: [listItemContext.selector && listItemContext.selector(), jsxRuntime.jsx("div", {
        class: itemStyles,
        children: jsxRuntime.jsx(HighlightText.HighlightText, {
          matchText: searchText,
          children: renderItemText(itemContext, itemText)
        })
      })]
    });
  }, [itemRenderer, itemText, onSelectionChange, searchText, selectedKeys]);
  const keys = hooks.useMemo(() => {
    return {
      all: false,
      keys: selectedKeys ?? new Set()
    };
  }, [selectedKeys]);
  return jsxRuntime.jsx(List.List, {
    "aria-label": ariaLabel,
    allowTabbableMode: false,
    currentItemVariant: currentItemVariant,
    currentKey: currentKey,
    data: data ?? null,
    loadingIndicator: loadingIndicator,
    onCurrentKeyChange: onCurrentKeyChange,
    onItemAction: onItemAction,
    onLoadRange: onLoadRange,
    onSelectionChange: onSelectionChange,
    role: "listbox",
    selectedKeys: keys,
    selectionMode: selectionMode,
    children: context => listItemRenderer(context)
  });
}
/**
 * Custom 3 Skeleton loader for select multiple
 * TODO: to remove this once  is solved
 */
const loadingIndicator = jsxRuntime.jsx(SkeletonContainer.SkeletonContainer, {
  minimumCount: 3,
  children: () => {
    return jsxRuntime.jsx(Flex.Flex, {
      height: "12x",
      align: "center",
      children: jsxRuntime.jsx(Skeleton.Skeleton, {
        height: "4x"
      })
    });
  }
});

const KEYS = {
    TAB: 'Tab',
    ENTER: 'Enter',
    NUMPAD_ENTER: 'NumpadEnter',
    ESC: 'Escape',
    SPACE: 'Space',
    LEFT: 'ArrowLeft',
    UP: 'ArrowUp',
    RIGHT: 'ArrowRight',
    DOWN: 'ArrowDown',
    SHIFT_LEFT: 'ShiftLeft',
    SHIFT_RIGHT: 'ShiftRight',
    CTRL_LEFT: 'ControlLeft',
    CTRL_RIGHT: 'ControlRight',
    ALT_LEFT: 'AltLeft',
    ALT_RIGHT: 'AltRight',
    PAGE_UP: 'PageUp',
    PAGE_DOWN: 'PageDown',
    HOME: 'Home',
    END: 'End',
    BACKSPACE: 'Backspace',
    DELETE: 'Delete'
};
const isControlKey = (event) => {
    switch (event.code) {
        case KEYS.SHIFT_LEFT:
        case KEYS.SHIFT_RIGHT:
        case KEYS.CTRL_LEFT:
        case KEYS.CTRL_RIGHT:
        case KEYS.ALT_LEFT:
        case KEYS.ALT_RIGHT:
            return true;
        default:
            return event.metaKey || event.ctrlKey;
    }
};
const isFunctionKey = (event) => {
    return /\bF([1-9]|1[0-2])\b/i.test(event.code);
};
const isControlOrFunctionKey = (event) => {
    return isControlKey(event) || isFunctionKey(event);
};
const isSimpleClick = (event) => {
    return event.button === 0 && !event.ctrlKey;
};
const preventDefault = (event) => {
    event.preventDefault();
};
const preventDefaultForCurrentTarget = (event) => {
    if (event.currentTarget === event.target) {
        event.preventDefault();
    }
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
const isPointOutside = (event, ref) => {
    const target = event.target;
    if (target) {
        const doc = getDocument(target);
        if (!doc.contains(target))
            return false;
    }
    return ref.every((elRef) => !elRef.current?.contains(target));
};
const getDocument = (node) => node?.ownerDocument ?? document;
const useOutsideMousedown = ({ isDisabled: disabled = false, ref, handler }) => {
    hooks.useEffect(() => {
        const refs = utils_UNSAFE_arrayUtils.coerceArray(ref);
        if (!handler ||
            disabled ||
            (refs.length > 0 && refs.findIndex((elRef) => elRef.current === null) > -1))
            return;
        const onMouseDown = (event) => {
            if (isPointOutside(event, refs)) {
                handler?.(event);
            }
        };
        const firstElRef = refs[0];
        const doc = getDocument(firstElRef.current);
        if (doc) {
            doc.addEventListener('mousedown', onMouseDown, true);
            return () => {
                doc.removeEventListener('mousedown', onMouseDown, true);
            };
        }
        return undefined;
    }, [disabled, handler, ref]);
};

const DEFAULT_PLACEMENT = 'bottom-start';
function Dropdown({ anchorRef, children, dropdownRef, id, isOpen, onAutoDismiss, onPosition }) {
    const [placement, setPlacement] = hooks.useState(DEFAULT_PLACEMENT);
    hooks.useEffect(() => {
        // reset the placement to the default when the dropdown closes, not opens, so that it doesn't
        // initially open in the previous position and then move to the default position on the
        // rerender after the state is updated
        if (!isOpen) {
            setPlacement(DEFAULT_PLACEMENT);
        }
    }, [isOpen, setPlacement]);
    const handlePosition = hooks.useCallback((positionData) => {
        // update the placement so that the dropdown stays where it is until there is a collision,
        // i.e. it no longer fits, instead of letting it try to move back to the default position
        // on every render
        setPlacement(positionData.placement);
        // call the callback that was passed in through props
        onPosition?.(positionData);
    }, [onPosition, setPlacement]);
    const handleAutoDismiss = hooks.useCallback((event) => {
        onAutoDismiss?.(event);
    }, [onAutoDismiss]);
    const handleKeyDown = hooks.useCallback((event) => {
        if (event.defaultPrevented || isControlOrFunctionKey(event)) {
            return;
        }
        switch (event.code) {
            case KEYS.ESC:
            case KEYS.TAB:
                handleAutoDismiss(event);
                break;
        }
    }, [handleAutoDismiss]);
    // use the mousedown event to be consistent with legacy JET and so that we're hiding the dropdown
    // on the same event where focus will transfer
    useOutsideMousedown({
        isDisabled: !isOpen,
        ref: [anchorRef, dropdownRef],
        handler: handleAutoDismiss
    });
    // TODO: get dropdown y (mainAxis) offset from --oj-c-PRIVATE-DO-NOT-USE-private-core-global-dropdown-offset
    const offsetValue = { mainAxis: 4, crossAxis: 0 };
    const inlineStyle = isOpen
        ? {
            minWidth: `${anchorRef.current?.offsetWidth}px`
        }
        : {};
    return !isOpen ? null : (jsxRuntime.jsx(UNSAFE_Layer.Layer, { logicalParentRef: anchorRef, children: jsxRuntime.jsx(Floating.Floating, { anchorRef: anchorRef, sizeOptions: {
                isMaxHeightAdjusted: true,
                maxHeightCeiling: 400
            }, ref: dropdownRef, placement: placement, offsetValue: offsetValue, onPosition: handlePosition, children: jsxRuntime.jsx("div", { class: PRIVATE_SelectCommon_themes_dropdownStyles_css.dropdownStyles.base, id: id, style: inlineStyle, onKeyDown: handleKeyDown, onMouseDown: preventDefaultForCurrentTarget, children: children }) }) }));
}

// TODO: replace this with var(--oj-button-height) once it is available
const ojButtonHeight = '2.75rem';
// TODO: replace this with var(--oj-button-sm-height) once it is available
const ojButtonSmHeight = '2.25rem';
// TODO: replace this with var(--oj-button-border-radius) once it is available
const ojButtonBorderRadius = 'var(--oj-c-PRIVATE-DO-NOT-USE-core-border-radius-md)';
// TODO: replace this with var(--oj-button-borderless-chrome-text-color) once it is available
const ojButtonBorderlessChromeTextColor = 'var(--oj-c-PRIVATE-DO-NOT-USE-core-text-color-primary)';
// TODO: replace this with var (--oj-button-borderless-chrome-text-color-disabled) once it is available
const ojButtonBorderlessChromeTextColorDisabled = 'var(--oj-c-PRIVATE-DO-NOT-USE-core-text-color-disabled)';
// TODO: replace this with var(--oj-button-borderless-chrome-text-color-hover) once it is available
const ojButtonBorderlessChromeTextColorHover = 'var(--oj-c-PRIVATE-DO-NOT-USE-core-text-color-primary)';
// TODO: replace this with var(--oj-button-borderless-chrome-bg-color-hover) once it is available
const ojButtonBorderlessChromeBgColorHover = 'var(--oj-c-PRIVATE-DO-NOT-USE-core-bg-color-hover)';
// TODO: replace this with var(--oj-button-borderless-chrome-border-color-hover) once it is available
const ojButtonBorderlessChromeBorderColorHover = 'transparent';
// TODO: replace this with var(--oj-c-PRIVATE-DO-NOT-USE-button-icon-size) once it is available
const ojButtonIconSize = 'var(--oj-c-PRIVATE-DO-NOT-USE-core-icon-size-lg)';
const rootStyles$3 = {
  base: "_1hw79jq",
  sizeMd: "osnrmf",
  sizeSm: "_10cpwdl",
  enabled: "cb9pd7",
  disabled: "rmx6gs"
};
const noop = () => {};
function DropdownArrow({
  isDisabled = false,
  onClick,
  size = 'md'
}) {
  const preventDefault = hooks.useCallback(event => {
    event.preventDefault();
  }, []);
  const {
    pressProps
  } = hooks_UNSAFE_usePress.usePress(onClick ?? noop, {
    isDisabled
  });
  const styleClasses = classNames.classNames([rootStyles$3.base, size === 'sm' ? rootStyles$3.sizeSm : rootStyles$3.sizeMd, isDisabled ? rootStyles$3.disabled : rootStyles$3.enabled]);
  return jsxRuntime.jsx("span", {
    "aria-hidden": "true",
    class: styleClasses,
    tabIndex: -1,
    onMouseDown: preventDefault,
    ...pressProps,
    children: jsxRuntime.jsx(PRIVATE_ThemedIcons_DropdownArrowIcon.DropdownArrowIcon, {})
  });
}

const rootStyles$2 = {
  base: "_7ew2di",
  // TODO: These should really be applied directly to the child ListView so that the content
  // displays under the padding when scrolled, but we can't apply styles to the child collection.
  // TODO: Also need to figure out how to determine whether the collection is a ListView (in
  // the collectionRenderer case) in order to apply this styling.
  topGap: "_8a0yjj",
  bottomGap: "_1ik89q2",
  loading: "md5w5g"
};
function DropdownList({
  children,
  hasBottomGap = true,
  isLoading = false
}) {
  const styleClasses = classNames.classNames([rootStyles$2.base, rootStyles$2.topGap, hasBottomGap && rootStyles$2.bottomGap, isLoading && rootStyles$2.loading]);
  return jsxRuntime.jsx(hooks_PRIVATE_useCollectionGestureContext.CollectionGestureContext.Provider, {
    value: 'embedded',
    children: jsxRuntime.jsx("div", {
      class: styleClasses,
      onMouseDown: preventDefaultForCurrentTarget,
      children: children
    })
  });
}

const rootStyles$1 = "_14wyexj";
function DropdownUserAssistance(props) {
  // Prevent the focus from transferring when the user clicks on an empty/non-clickable area
  // of the user assistance
  const handleMouseDown = hooks.useCallback(event => {
    // Don't call preventDefault when the user clicks on the help source link, so that the
    // focus is allowed to transfer in that case.
    if (event.target?.tagName !== 'A') {
      event.preventDefault();
    }
  }, []);
  return jsxRuntime.jsx(hooks_UNSAFE_useFormFieldContext.FormFieldContext.Provider, {
    value: {
      isFocused: true
    },
    children: jsxRuntime.jsx("div", {
      class: rootStyles$1,
      onMouseDown: handleMouseDown,
      children: jsxRuntime.jsx(UNSAFE_UserAssistance.InlineUserAssistance, {
        ...props
      })
    })
  });
}

const rootStyles = "_29d6ea";
function EmptyResults() {
  const translations = hooks_UNSAFE_useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
  // TODO: figure out whether we need a separate resource string for this message;
  // (JET-53984 - Preact Select* components: empty results message)
  // call preventDefault for mousedown on this div so that focus does not transfer from
  // the input
  return jsxRuntime.jsx("div", {
    class: rootStyles,
    onMouseDown: preventDefaultForCurrentTarget,
    children: translations.select_noMatchesFound()
  });
}

function MobileDropdown({ children, id, isOpen }) {
    return !isOpen ? null : (jsxRuntime.jsx(UNSAFE_Layer.Layer, { children: jsxRuntime.jsx(UNSAFE_WindowOverlay.WindowOverlay, { children: jsxRuntime.jsx("div", { id: id, class: PRIVATE_SelectCommon_themes_mobileDropdownStyles_css.mobileDropdownStyles.base, role: "dialog", "aria-modal": "true", children: jsxRuntime.jsx(Flex.Flex, { width: "100%", height: "100%", maxHeight: "100%", direction: "column", children: children }) }) }) }));
}

function SelectMobileDropdown({ children, footer, header, id, isOpen }) {
    return (jsxRuntime.jsxs(MobileDropdown, { id: id, isOpen: isOpen, children: [header, jsxRuntime.jsx(Divider.Divider, {}), jsxRuntime.jsx("div", { class: PRIVATE_SelectCommon_themes_selectMobileDropdownStyles_css.selectMobileDropdownStyles.content, children: children }), jsxRuntime.jsx(Divider.Divider, {}), footer] }));
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Helper function to determine whether the current device is a mobile device
 * @returns true if running on a mobile device, false otherwise
 */
function isMobile() {
    const deviceType = clientHints.getClientHints().deviceType;
    return deviceType === 'phone' || deviceType === 'tablet';
}
/**
 * Helper function to determine whether the current device is a phone
 * @returns true if running on a phone, false otherwise
 */
function isPhone() {
    return clientHints.getClientHints().deviceType === 'phone';
}
/**
 * Helper function to determine whether the current device is a tablet
 * @returns true if running on a tablet, false otherwise
 */
function isTablet() {
    return clientHints.getClientHints().deviceType === 'tablet';
}
function isBeforeDataFetch(data) {
    return data.sizePrecision === 'atLeast' && data.totalSize === 0;
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Hook that manages state and behavior for a Select* component.  This hook creates state
 * variables and event listeners, returning properties to apply to components internally
 * rendered by a Select* component, as well as state information.
 *
 * @param data Specifies data for the dropdown list.
 * @param inputRef Ref to the input element.
 * @param isFocused Specifies whether the component has focus.
 * @param onFilter Callback function to trigger loading data for the dropdown list,
 * which may or may not be filtered by user entered text.
 *
 * @returns Component state and event listeners.
 */
function useSelectCommon({ data, inputRef, isFocused, onFilter }) {
    const mainFieldRef = hooks.useRef(null);
    const dropdownRef = hooks.useRef(null);
    const [isDropdownOpen, setDropdownOpen] = hooks.useState(false);
    const [isDropdownAbove, setDropdownAbove] = hooks.useState(false);
    const [currentRow, setCurrentRow] = hooks.useState({
        rowKey: undefined
    });
    const [selectedKeys, setSelectedKeys] = hooks.useState();
    const [isUserFiltering, setUserFiltering] = hooks.useState(false);
    const [userInput, setUserInput] = hooks.useState();
    const [searchText, setSearchText] = hooks.useState();
    const onMouseDown = hooks.useCallback((event) => {
        // const target = event.target as HTMLElement;
        if (event.defaultPrevented || !isSimpleClick(event)) {
            return;
        }
        if (!isDropdownOpen) {
            setDropdownOpen(true);
        }
        // this is needed to focus the input when clicking on the inside label when there is no
        // value;  otherwise the component looks focused but doesn't actually have physical focus
        setTimeout(function () {
            inputRef.current?.focus();
        }, 0);
    }, [inputRef, isDropdownOpen]);
    const handleDropdownArrowClick = hooks.useCallback(() => {
        setDropdownOpen(!isDropdownOpen);
        // focus the input so that if the user Tabs afterwards, it loses focus and the dropdown closes
        inputRef.current?.focus();
    }, [inputRef, isDropdownOpen]);
    const handleDropdownAutoDismiss = hooks.useCallback((event) => {
        if (isDropdownOpen) {
            setDropdownOpen(false);
            switch (event?.type) {
                case 'keydown':
                    const keyboardEvent = event;
                    switch (keyboardEvent.code) {
                        case KEYS.ESC:
                            // focus the input when the user cancels out of the dropdown
                            inputRef.current?.focus();
                            break;
                        default:
                            break;
                    }
                    break;
                case 'mousedown':
                    // if focus is in the dropdown, focus the input before the dropdown is hidden so that
                    // focus will transfer correctly when the target is not focusable;
                    // otherwise the text cursor appears in the input again
                    if (dropdownRef.current?.contains(document.activeElement)) {
                        inputRef.current?.focus();
                    }
                    break;
                default:
                    break;
            }
        }
    }, [inputRef, isDropdownOpen]);
    const handleDropdownPosition = hooks.useCallback((data) => {
        setDropdownAbove(data.placement.startsWith('top'));
    }, []);
    const handleUpDownArrowKeys = hooks.useCallback((event) => {
        if (!isDropdownOpen) {
            setDropdownOpen(true);
        }
        else {
            // transfer physical focus into the dropdown
            const firstFocusableElem = dropdownRef.current?.querySelector('[tabindex="0"]');
            firstFocusableElem?.focus();
            // if there is no current item set, make the first item current
            if (data !== undefined && currentRow.rowKey === undefined) {
                setCurrentRow({ rowKey: data?.data?.[0]?.metadata?.key });
            }
        }
        // prevent the page from scrolling
        event.preventDefault();
    }, [currentRow, data, isDropdownOpen]);
    const handleMainFieldKeyDown = hooks.useCallback((event) => {
        // const target = event.target as HTMLElement;
        // ignore control key and function key
        if (isControlOrFunctionKey(event)) {
            return;
        }
        switch (event.code) {
            case KEYS.PAGE_UP:
            case KEYS.PAGE_DOWN:
                // prevent the page from scrolling
                event.preventDefault();
                return;
            case KEYS.UP:
            case KEYS.DOWN:
                // prevent the page from scrolling
                event.preventDefault();
                break;
            case KEYS.ESC:
                if (isDropdownOpen) {
                    setDropdownOpen(false);
                    event.preventDefault();
                    // JET-53183 - core pack compat: stop propagation of handled events
                    // we don't want a parent collection component to stop editing the row when we close
                    // the dropdown
                    event.stopPropagation();
                }
                break;
            default:
                break;
        }
    }, [isDropdownOpen]);
    const handleMainFieldKeyUp = hooks.useCallback((event) => {
        // ignore control key and function key
        if (isControlOrFunctionKey(event)) {
            return;
        }
        switch (event.code) {
            case KEYS.UP:
            case KEYS.DOWN:
                handleUpDownArrowKeys(event);
                break;
            default:
                break;
        }
    }, [handleUpDownArrowKeys]);
    const handleInput = hooks.useCallback((detail) => {
        if (!isDropdownOpen) {
            setDropdownOpen(true);
        }
        setUserFiltering(true);
        setUserInput(detail.value);
        const str = detail.value?.trim();
        setSearchText(str);
        onFilter?.({ searchText: str });
    }, [isDropdownOpen, onFilter]);
    const onCurrentKeyChange = hooks.useCallback((detail) => {
        setCurrentRow({ rowKey: detail.value });
    }, []);
    const stopFiltering = hooks.useCallback(() => {
        setUserFiltering(false);
        setUserInput('');
        setSearchText('');
    }, []);
    // the user is not actively filtering if the dropdown is closed and the component doesn't
    // have focus
    hooks.useEffect(() => {
        if (!isDropdownOpen && !isFocused && isUserFiltering) {
            stopFiltering();
        }
    }, [isDropdownOpen, isFocused, isUserFiltering, stopFiltering]);
    // need to call onFilter with undefined searchText when the dropdown opens and the
    // user is NOT filtering
    hooks.useEffect(() => {
        if (isDropdownOpen && !isUserFiltering) {
            onFilter?.({ searchText: undefined });
        }
    }, [isDropdownOpen, isUserFiltering, onFilter]);
    // when the dropdown closes, reset the isDropdownAbove flag
    hooks.useEffect(() => {
        if (!isDropdownOpen) {
            setDropdownAbove(false);
        }
    }, [isDropdownOpen]);
    return {
        currentRow,
        dropdownRef,
        handleDropdownArrowClick,
        handleDropdownAutoDismiss,
        handleDropdownPosition,
        handleInput,
        handleMainFieldKeyDown,
        handleMainFieldKeyUp,
        handleUpDownArrowKeys,
        isDropdownAbove,
        isDropdownOpen,
        // the field should remain to look focused while the dropdown is open,
        // in case physical focus is in the dropdown
        isFocused: isFocused || isDropdownOpen,
        isUserFiltering,
        mainFieldRef,
        onCurrentKeyChange,
        onMouseDown,
        searchText,
        selectedKeys,
        setCurrentRow,
        setDropdownOpen,
        setSelectedKeys,
        setUserInput,
        stopFiltering,
        userInput
    };
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

exports.DefaultList = DefaultList;
exports.Dropdown = Dropdown;
exports.DropdownArrow = DropdownArrow;
exports.DropdownList = DropdownList;
exports.DropdownUserAssistance = DropdownUserAssistance;
exports.EmptyResults = EmptyResults;
exports.KEYS = KEYS;
exports.MobileDropdown = MobileDropdown;
exports.SelectMobileDropdown = SelectMobileDropdown;
exports.isBeforeDataFetch = isBeforeDataFetch;
exports.isControlKey = isControlKey;
exports.isControlOrFunctionKey = isControlOrFunctionKey;
exports.isFunctionKey = isFunctionKey;
exports.isMobile = isMobile;
exports.isPhone = isPhone;
exports.isSimpleClick = isSimpleClick;
exports.isTablet = isTablet;
exports.preventDefault = preventDefault;
exports.preventDefaultForCurrentTarget = preventDefaultForCurrentTarget;
exports.renderItemText = renderItemText;
exports.useSelectCommon = useSelectCommon;
