// @oracle/oraclejet-preact/cjs/SplitMenuButton-67c6f593.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
require('./UNSAFE_ButtonLayout.js');
var PRIVATE_ThemedIcons_NavDownIcon = require('./PRIVATE_ThemedIcons/NavDownIcon.js');
require('./UNSAFE_Menu.js');
require('./UNSAFE_BaseButton.js');
require('./utils/UNSAFE_classNames.js');
var hooks_UNSAFE_useTabbableMode = require('./hooks/UNSAFE_useTabbableMode.js');
var utils_UNSAFE_interpolations_dimensions = require('./utils/UNSAFE_interpolations/dimensions.js');
var utils_UNSAFE_mergeInterpolations = require('./utils/UNSAFE_mergeInterpolations.js');

var compat = require('preact/compat');
var hooks = require('preact/hooks');
var classNames = require('./classNames-d619a292.js');
var BaseButton = require('./BaseButton-934b844e.js');
var ButtonLayout = require('./ButtonLayout-648ecbd5.js');
var Menu = require('./Menu-66c84b6d.js');

const compStyles = {
  divider: "_1wi4nye",
  disabledDivider: "_1qljftu",
  icon: "knfga3",
  root: "_1jbmydi",
  childrenMargins: "_12wt61q"
};
const menuButtonStyling = ['embedded', 'min', 'noBorderRadiusStart'];
const menuButtonActiveStyling = [...menuButtonStyling, 'active'];
const actionbuttonStyling = ['embedded', 'fill', 'noBorderRadiusEnd'];
const actionButtonActiveStyling = [...actionbuttonStyling, 'active'];
const interpolations = [...Object.values(utils_UNSAFE_interpolations_dimensions.dimensionInterpolations)];
const styleInterpolations = utils_UNSAFE_mergeInterpolations.mergeInterpolations(interpolations);
/**
 * A SplitMenuButton is a combined action and menu button.
 */
const SplitMenuButton = compat.forwardRef(({
  children,
  label = '',
  variant = 'outlined',
  isDisabled = false,
  size = 'md',
  onAction,
  ...props
}, ref) => {
  const [isMenuOpen, setIsMenuOpen] = hooks.useState(false);
  const [isActive, setIsActive] = hooks.useState(false);
  const {
    class: cls,
    ...styles
  } = styleInterpolations(props);
  // SPACE or ENTER on root area invokes action, as does CLICK/TAP on LEFT SIDE ACTION AREA
  // DOWN ARROW invokes menu, as does CLICK/TAP on RIGHT SIDE MENU AREA
  const onKeyDown = hooks.useCallback(event => {
    if (event.key === ' ' || event.key === 'Enter') {
      setIsActive(true);
    } else if (event.key == 'ArrowDown') {
      !isDisabled && setIsMenuOpen(menuOpen => !menuOpen);
      // Prevent page scroll for down arrow, user definitely wanted to show menu in this case.
      // Don't prevent other keypress defaults, as user may want to hold SPACE down to scroll page.
      event.preventDefault();
    }
  }, [isDisabled, setIsMenuOpen, setIsActive]);
  const onKeyUp = hooks.useCallback(event => {
    if (event.key === ' ' || event.key === 'Enter') {
      !isDisabled && onAction && onAction();
      setIsActive(false);
    }
  }, [isDisabled, onAction, setIsActive]);
  const onBlur = hooks.useCallback(() => {
    setIsActive(false);
  }, [setIsActive]);
  const _ref = hooks.useRef(null);
  hooks.useImperativeHandle(ref, () => _ref.current, [_ref]);
  const menuHandler = () => {
    setIsMenuOpen(!isMenuOpen);
  };
  const handleClose = e => {
    //Focus has to be set on the launcher when menu is dismissed or
    //item is "selected"
    if (e.reason === 'dismissed' || e.reason === 'itemAction') {
      _ref.current?.focus();
    }
    setIsMenuOpen(false);
  };
  const dividerStyles = classNames.classNames([compStyles.divider, isDisabled && compStyles.disabledDivider]);
  const rootStyles = classNames.classNames([compStyles.root]);
  // Using Span for top button allows us to avoid :hover state rendering by
  // taking advantage of :enabled not on spans
  const rootButton = () => {
    return jsxRuntime.jsx(BaseButton.BaseButton, {
      elementDetails: {
        type: 'span'
      },
      ref: _ref,
      isDisabled: isDisabled,
      styling: ['container', 'min'],
      variant: variant,
      size: size,
      "aria-label": label,
      "aria-roledescription": "split menu button, press down arrow to invoke menu",
      "aria-expanded": isMenuOpen,
      children: jsxRuntime.jsxs(hooks_UNSAFE_useTabbableMode.TabbableModeContext.Provider, {
        value: {
          isTabbable: false
        },
        children: [jsxRuntime.jsx(BaseButton.BaseButton, {
          elementDetails: {
            type: 'span'
          },
          variant: variant,
          styling: getActionButtonStyling(isActive),
          size: size,
          isDisabled: isDisabled,
          "aria-hidden": true,
          onAction: onAction,
          children: label
        }), jsxRuntime.jsx("div", {
          class: `${dividerStyles}`
        }), jsxRuntime.jsx(BaseButton.BaseButton, {
          elementDetails: {
            type: 'span'
          },
          variant: variant,
          styling: getMenuButtonStyling(isMenuOpen),
          size: size,
          isDisabled: isDisabled,
          "aria-hidden": true,
          onAction: menuHandler,
          children: jsxRuntime.jsx(ButtonLayout.ButtonLayout, {
            display: "icons",
            startIcon: jsxRuntime.jsx(PRIVATE_ThemedIcons_NavDownIcon.NavDownIcon, {}),
            size: size,
            styling: "embedded"
          })
        })]
      })
    });
  };
  if (isDisabled) {
    return jsxRuntime.jsx("span", {
      role: "toolbar",
      class: `${rootStyles} ${cls}`,
      style: styles,
      children: rootButton()
    });
  } else {
    return jsxRuntime.jsxs("span", {
      role: "toolbar",
      class: `${rootStyles} ${cls}`,
      style: styles,
      onKeyDown: !isDisabled && onKeyDown,
      onKeyUp: onKeyUp,
      onBlur: onBlur,
      children: [rootButton(), jsxRuntime.jsx(Menu.Menu, {
        anchorRef: _ref,
        isOpen: isMenuOpen,
        onClose: handleClose,
        children: jsxRuntime.jsx("span", {
          className: compStyles.childrenMargins,
          children: children
        })
      })]
    });
  }
});
function getMenuButtonStyling(isActive) {
  return isActive ? menuButtonActiveStyling : menuButtonStyling;
}
function getActionButtonStyling(isActive) {
  return isActive ? actionButtonActiveStyling : actionbuttonStyling;
}

exports.SplitMenuButton = SplitMenuButton;
