// @oracle/oraclejet-preact/cjs/Subheading-2871737c.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var utils_UNSAFE_arrayUtils = require('./utils/UNSAFE_arrayUtils.js');
var UNSAFE_Subheading_themes_SubheadingStyles_css = require('./UNSAFE_Subheading/themes/SubheadingStyles.css.js');

const subheadingElementTypes = utils_UNSAFE_arrayUtils.stringLiteralArray([
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'div',
    'span'
]);
const headingsizeMap = {
    h1: '2xl',
    h2: 'xl',
    h3: 'lg',
    h4: 'md',
    h5: 'sm',
    h6: 'xs',
    div: 'md',
    span: 'md'
};
const sizeOptions = ['2xs', 'xs', 'sm', 'md', 'lg', 'xl', '2xl', 'inherit'];
const Subheading = ({ children, as, size, id, variant }) => {
    const Component = as;
    //@ts-ignore
    size = size ?? headingsizeMap[Component];
    const classes = UNSAFE_Subheading_themes_SubheadingStyles_css.multiVariantStyles({ size, variant });
    return (jsxRuntime.jsx(Component, { className: classes, id: id, children: children }));
};

exports.Subheading = Subheading;
exports.sizeOptions = sizeOptions;
exports.subheadingElementTypes = subheadingElementTypes;
