// @oracle/oraclejet-preact/cjs/UNSAFE_DrawerPopup.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
var UNSAFE_Layer = require('./UNSAFE_Layer.js');
require('./UNSAFE_Modal.js');
require('./UNSAFE_FocusTrap.js');
var utils_PRIVATE_tabbableUtils = require('./utils/PRIVATE_tabbableUtils.js');
var utils_PRIVATE_floatingUtils = require('./utils/PRIVATE_floatingUtils.js');
require('./utils/PRIVATE_clientHints.js');
var hooks_UNSAFE_useOutsideClick = require('./hooks/UNSAFE_useOutsideClick.js');
require('./hooks/UNSAFE_useAnimation.js');
var hooks_UNSAFE_useUser = require('./hooks/UNSAFE_useUser.js');
var hooks_UNSAFE_useComponentTheme = require('./hooks/UNSAFE_useComponentTheme.js');
var utils_UNSAFE_interpolations_colors = require('./utils/UNSAFE_interpolations/colors.js');
var utils_UNSAFE_mergeInterpolations = require('./utils/UNSAFE_mergeInterpolations.js');
var UNSAFE_DrawerPopup_themes_redwood_DrawerPopupTheme = require('./UNSAFE_DrawerPopup/themes/redwood/DrawerPopupTheme.js');
var useAnimation = require('./useAnimation-802f3cdb.js');
var clientHints = require('./clientHints-b9e7be1e.js');
var FocusTrap = require('./FocusTrap-2209a89f.js');
var Modal = require('./Modal-a5f87ce5.js');
require('preact/compat');
require('./index-c3ee789d.js');
require('preact');
require('./Common.js');
require('./Common/themes.js');
require('./Common/themes/redwood/theme.js');
require('./Common/themes/themeContract.css.js');
require('./useThemeInterpolations-ee4e7934.js');
require('./hooks/UNSAFE_useColorScheme.js');
require('./hooks/UNSAFE_useScale.js');
require('./utils/UNSAFE_interpolations/theme.js');
require('./UNSAFE_Theme.js');
require('./utils-b9f57550.js');
require('./utils/UNSAFE_classNames.js');
require('./classNames-d619a292.js');
require('./_curry1-6e9511bf.js');
require('./_curry3-ad2313fd.js');
require('./_curry2-53a902bf.js');
require('./_has-aa203696.js');
require('./useModal-2ac2ced5.js');
require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-8a4b93e4.js');

require('./_arity-353457ba.js');
require('./utils/UNSAFE_arrayUtils.js');
require('./utils/UNSAFE_logger.js');
require('./UNSAFE_DrawerPopup/themes/redwood/DrawerPopupBaseTheme.css.js');

require('./UNSAFE_DrawerPopup/themes/DrawerPopupStyles.css.js');

require('./UNSAFE_DrawerPopup/themes/redwood/DrawerPopupVariants.css.js');


// Constants
const displayModefullWidthThreshold = 600;
const zero = '0px';
const duration = 250;
const getElementHeight = (element) => {
    return Math.round(element.offsetHeight);
};
const getElementWidth = (element) => {
    return Math.round(element.getBoundingClientRect().width);
};
/**
 * A drawer popup adds a single slide-in side content alongside some primary content to an application window.
 *
 * This drawer always overlays the page and can be placed at the 'start', 'end' or 'bottom' edge.
 */
const DrawerPopup = ({ children, isOpen = false, autoFocusRef, placement = 'start', modality = 'modal', onClose, onTransitionEnd, 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy, ...props }) => {
    // Initialize states
    const [status, setStatus] = hooks.useState(isOpen ? 'initial' : 'unmounted');
    const [visibility, setVisibility] = hooks.useState('hidden');
    const [isOutsideClickDisabled, setIsOutsideClickDisabled] = hooks.useState(!isOpen);
    const [displayModeHorizontal, setDisplayModeHorizontal] = hooks.useState('overlay');
    const rootRef = hooks.useRef(null);
    const launcherRef = hooks.useRef(null);
    const isBottomPlacement = placement === 'bottom';
    // Animation config
    const animationStates = {
        opening: () => {
            if (isBottomPlacement) {
                return {
                    from: {
                        translateY: getTranslateY()
                    },
                    to: {
                        translateY: zero
                    },
                    options: {
                        duration: duration
                    }
                };
            }
            else {
                return {
                    from: {
                        translateX: getTranslateX()
                    },
                    to: {
                        translateX: zero
                    },
                    options: {
                        duration: duration
                    }
                };
            }
        },
        closing: () => {
            if (isBottomPlacement) {
                return {
                    from: {
                        translateY: zero
                    },
                    to: {
                        translateY: getTranslateY()
                    },
                    options: {
                        duration: duration
                    }
                };
            }
            else {
                return {
                    from: {
                        translateX: zero
                    },
                    to: {
                        translateX: getTranslateX()
                    },
                    options: {
                        duration: duration
                    }
                };
            }
        }
    };
    // Normalizing position
    const { direction } = hooks_UNSAFE_useUser.useUser();
    const getNormalizedPlacement = () => {
        return utils_PRIVATE_floatingUtils.normalizePosition(placement, direction);
    };
    const getTranslateX = () => {
        return `${getNormalizedPlacement() === 'right' ? '' : '-'}${getElementWidth(rootRef.current)}px`;
    };
    const getTranslateY = () => {
        return `${getElementHeight(rootRef.current)}px`;
    };
    // Animation
    const { nodeRef } = useAnimation.useAnimation(status, {
        animationStates,
        isAnimatedOnMount: true,
        onAnimationEnd: hooks.useCallback(() => {
            if (status === 'opening') {
                onTransitionEnd?.(true);
            }
            else if (!isOpen) {
                nodeRef(null);
                setVisibility('hidden');
                setStatus('unmounted');
                onTransitionEnd?.(false);
            }
        }, [status, isOpen])
    });
    hooks.useEffect(() => {
        // 1. Ignore status update on initial render
        if (status === 'unmounted' && !isOpen) {
            return;
        }
        // 2. Mount visually hidden component
        // Component gets mounted, but will be visually hidden to avoid flickering
        // Flickering happens because it gets rendered before animation hook repositions
        // it in next render cycle
        if (status === 'unmounted' && isOpen) {
            setStatus('initial');
            return;
        }
        // 3. Setup animation
        // Component has been mounted.
        // We will kick out animation as node is already present in DOM.
        // To avoid flickering, we still keep it visually hidden until the next render cycle.
        if (status === 'initial' && isOpen) {
            nodeRef(rootRef.current);
            setIsOutsideClickDisabled(false);
            setStatus('opening');
            return;
        }
        // 3. Unhide component. It becomes visible.
        if (status === 'opening' && isOpen) {
            setVisibility('visible');
            return;
        }
        // 4. Closing
        if (!isOpen) {
            setIsOutsideClickDisabled(true);
            setStatus('closing');
        }
    }, [isOpen, status]);
    // Launcher handler
    hooks.useEffect(() => {
        if (status === 'opening') {
            // Remember launcher
            const activeElement = utils_PRIVATE_tabbableUtils.getActiveElement();
            launcherRef.current = activeElement === document.body ? document.body : activeElement;
            // Register F6 key handler to enter the Popup
            launcherRef.current?.addEventListener('keydown', launcherKeyDownCallback);
        }
        if (status === 'closing') {
            return () => {
                // Deregister F6 key handler
                launcherRef.current?.removeEventListener('keydown', launcherKeyDownCallback);
            };
        }
        return;
    }, [status]);
    const launcherKeyDownCallback = (event) => {
        // Moves focus from launcher (must have focus) back to the Drawer
        if (launcherRef.current === utils_PRIVATE_tabbableUtils.getActiveElement() && event.code === 'F6') {
            // Prevent default F6 handlers.
            // F6 is a standard Chrome address bar shortcut on Windows.
            event.preventDefault();
            event.stopPropagation();
            // Focus on first tabbable in a container or container itself if there is none
            utils_PRIVATE_tabbableUtils.focusWithin(rootRef.current);
        }
    };
    const handleKeyDown = hooks.useCallback((event) => {
        const launcherEl = launcherRef.current;
        switch (event.code) {
            case 'Escape': {
                onClose?.({ reason: 'escapeKey' });
                break;
            }
            case 'F6':
                // Prevent default F6 handlers.
                // F6 is a standard Chrome address bar shortcut on Windows.
                event.preventDefault();
                event.stopPropagation();
                // Focus launcher
                if (launcherEl) {
                    utils_PRIVATE_tabbableUtils.focusOn(launcherEl);
                }
                break;
        }
    }, [onClose]);
    // OutsideClick handler
    const handleOutsideClick = hooks.useCallback(() => {
        // Trigger onClose event with outside click detail
        onClose?.({ reason: 'outsideClick' });
    }, [onClose]);
    hooks_UNSAFE_useOutsideClick.useOutsideClick({
        isDisabled: isOutsideClickDisabled,
        ref: rootRef,
        handler: handleOutsideClick
    });
    // Drawer placement
    const getDrawerPlacement = () => {
        if (isBottomPlacement) {
            return placement;
        }
        else {
            if (direction === 'rtl') {
                return placement === 'start' ? 'right' : 'left';
            }
            else {
                return placement === 'start' ? 'left' : 'right';
            }
        }
    };
    // Window resize handler
    // Handles display mode: Overlay <> Full overlay (full width side drawers)
    const handleResize = () => {
        const viewportWidth = getViewportWidth();
        if (displayModeHorizontal === 'overlay' && viewportWidth < displayModefullWidthThreshold) {
            setDisplayModeHorizontal('fullOverlay');
        }
        if (displayModeHorizontal === 'fullOverlay' && viewportWidth > displayModefullWidthThreshold) {
            setDisplayModeHorizontal('overlay');
        }
    };
    const clientHints$1 = clientHints.getClientHints();
    const getViewportWidth = () => {
        if (clientHints$1.platform === 'ios') {
            // On ios window.innerWidth is not recommended way of measuring the viewport
            return document.documentElement.clientWidth;
        }
        return window.innerWidth;
    };
    hooks.useEffect(() => {
        const root = rootRef.current;
        if (root) {
            const resizeObserver = new ResizeObserver(() => {
                handleResize();
            });
            resizeObserver.observe(document.body);
            return () => {
                resizeObserver.unobserve(document.body);
            };
        }
        return undefined;
    }, [handleResize]);
    // Styles
    const { classes } = hooks_UNSAFE_useComponentTheme.useComponentTheme(UNSAFE_DrawerPopup_themes_redwood_DrawerPopupTheme.DrawerPopupRedwoodTheme, {
        placement: getDrawerPlacement(),
        visibility,
        displayModeHorizontal
    });
    const interpolations = [...Object.values(utils_UNSAFE_interpolations_colors.colorInterpolations)];
    const styleInterpolations = utils_UNSAFE_mergeInterpolations.mergeInterpolations(interpolations);
    const { ...styles } = styleInterpolations(props);
    const renderDrawer = () => {
        return (jsxRuntime.jsx("div", { ref: rootRef, className: classes, style: styles, tabIndex: -1, role: "dialog", onKeyDown: handleKeyDown, "aria-label": ariaLabel, "aria-labeledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy, children: jsxRuntime.jsx(FocusTrap.FocusTrap, { autoFocusRef: autoFocusRef, children: children }) }));
    };
    const isMounted = status !== 'unmounted';
    if (isMounted) {
        // Modal vs. modelss upper wrapper
        if (modality === 'modal') {
            return jsxRuntime.jsx(Modal.Modal, { isOpen: isMounted, children: renderDrawer() });
        }
        else {
            return jsxRuntime.jsx(UNSAFE_Layer.Layer, { children: renderDrawer() });
        }
    }
    return null;
};

exports.DrawerPopup = DrawerPopup;
