// @oracle/oraclejet-preact/cjs/UNSAFE_PrefixSuffix.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var UNSAFE_PrefixSuffix_themes_PrefixSuffixStyles_css = require('./UNSAFE_PrefixSuffix/themes/PrefixSuffixStyles.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-8a4b93e4.js');

function PrefixSuffix({ hasEndContent, hasInsideLabel, hasStartContent, id, isDisabled, isFocused, text, variant }) {
    const variantClasses = UNSAFE_PrefixSuffix_themes_PrefixSuffixStyles_css.multiVariantStyles({
        variant,
        isFocused: isFocused ? 'yes' : 'no',
        isDisabled: isDisabled ? 'yes' : 'no',
        hasInsideLabel: hasInsideLabel ? 'yes' : 'no',
        hasStartContent: hasStartContent ? 'yes' : 'no',
        hasEndContent: hasEndContent ? 'yes' : 'no'
    });
    return (jsxRuntime.jsx("div", { class: variantClasses, id: id, children: jsxRuntime.jsx("span", { children: text }) }));
}

exports.PrefixSuffix = PrefixSuffix;
