// @oracle/oraclejet-preact/cjs/utils-b9f57550.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

var Common_themes_themeContract_css = require('./Common/themes/themeContract.css.js');

const rgb = (rgb) => `rgb(${rgb})`;
const rgba = (rgb, alpha) => `rgba(${rgb}, ${alpha})`;
// JET-59344 refactor of 'fallbackVar' from '@vanilla-extract/css'.
// directly importing the function was causing vanilla-extract to be bundled into
// the preact packages and causing issues with testing
const fallbackVar = (...values) => {
    const finalValue = values.reverse().reduce((acc, value) => {
        if (acc === '') {
            acc = String(value);
        }
        else {
            if (typeof value !== 'string' || !/^var\(--.*\)$/.test(value)) {
                throw new Error(`Invalid variable name: ${value}`);
            }
            acc = value.replace(/\)$/, `, ${acc})`);
        }
        return acc;
    }, '');
    return finalValue;
};
// MessageBanner is used by legacy oj-message-banner, so we need to fallback the size units
// var to the legacy CSS var
const xUnits = (cnt) => cnt === 1
    ? fallbackVar(Common_themes_themeContract_css.scaleVars.size.units, 'var(--oj-core-spacing-1x)')
    : `calc(${cnt} * ${fallbackVar(Common_themes_themeContract_css.scaleVars.size.units, 'var(--oj-core-spacing-1x)')})`;
const min = (val, minVal) => `min(${val},${minVal})`;
const max = (val, minVal) => `max(${val},${minVal})`;

exports.fallbackVar = fallbackVar;
exports.max = max;
exports.min = min;
exports.rgb = rgb;
exports.rgba = rgba;
exports.xUnits = xUnits;
