// @oracle/oraclejet-preact/cjs/hooks/PRIVATE_useItemAction.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var hooks = require('preact/hooks');
var utils_PRIVATE_collectionUtils = require('../utils/PRIVATE_collectionUtils.js');
require('../utils/UNSAFE_keys.js');
require('../utils/PRIVATE_clientHints.js');
require('../clientHints-b9e7be1e.js');

const useItemAction = (currentKey, dataState, onItemAction, itemSelector, allowSpacebar) => {
    if (!dataState || !onItemAction) {
        return {};
    }
    return {
        onClick: hooks.useCallback((event) => {
            const key = utils_PRIVATE_collectionUtils.keyExtractor(event.target, itemSelector);
            if (handleItemAction(key, dataState, onItemAction)) {
                event.stopPropagation();
            }
        }, [dataState, onItemAction]),
        onKeyDown: hooks.useCallback((event) => {
            if (event.key === 'Enter') {
                handleItemAction(currentKey, dataState, onItemAction);
            }
            else if (allowSpacebar && event.key === ' ') {
                handleItemAction(currentKey, dataState, onItemAction);
                event.preventDefault();
            }
        }, [dataState, onItemAction, currentKey])
    };
};
const handleItemAction = (key, dataState, onItemAction) => {
    if (key !== undefined) {
        const context = findItemContext(key, dataState);
        context && onItemAction({ context });
        return context !== undefined;
    }
    return false;
};
// returns a ListItemContext based on key
const findItemContext = (key, dataState) => {
    const index = dataState?.data?.findIndex((d) => d.metadata.key === key) ?? -1;
    if (index > -1) {
        const context = {
            index: index + (dataState?.offset || 0),
            data: dataState?.data[index].data,
            metadata: dataState?.data[index].metadata
        };
        return context;
    }
    return undefined;
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

exports.useItemAction = useItemAction;
