// @oracle/oraclejet-preact/cjs/hooks/PRIVATE_useDvtMeterEvents.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var hooks = require('preact/hooks');
var hooks_UNSAFE_useUser = require('./UNSAFE_useUser.js');
var hooks_UNSAFE_useTooltip = require('./UNSAFE_useTooltip.js');
require('../index-c3ee789d.js');
require('preact');
require('../Common.js');
require('../Common/themes.js');
require('../Common/themes/redwood/theme.js');
require('../Common/themes/themeContract.css.js');
require('preact/jsx-runtime');
require('preact/compat');
require('./UNSAFE_useId.js');
require('../UNSAFE_Floating.js');
require('../Floating-3415cf36.js');
require('../index-0a166c9e.js');
require('../utils/PRIVATE_floatingUtils.js');
require('../utils/PRIVATE_refUtils.js');
require('./UNSAFE_useOutsideClick.js');
require('../utils/UNSAFE_arrayUtils.js');
require('./UNSAFE_useComponentTheme.js');
require('../utils/UNSAFE_logger.js');
require('../_curry1-6e9511bf.js');
require('../_curry2-53a902bf.js');
require('../utils/UNSAFE_classNames.js');
require('../classNames-d619a292.js');
require('../utils-b9f57550.js');
require('../UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('../UNSAFE_Floating/themes/FloatingStyles.css.js');

require('../UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('../UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('../vanilla-extract-recipes-createRuntimeFn.esm-8a4b93e4.js');
require('../vanilla-extract-dynamic.esm-4bc33829.js');
require('../UNSAFE_Floating/themes/FloatingContract.css.js');
require('../UNSAFE_Layer.js');
require('../useThemeInterpolations-ee4e7934.js');
require('./UNSAFE_useColorScheme.js');
require('./UNSAFE_useScale.js');
require('../utils/UNSAFE_interpolations/theme.js');
require('../UNSAFE_Theme.js');
require('../utils/UNSAFE_mergeInterpolations.js');
require('../_curry3-ad2313fd.js');
require('../_has-aa203696.js');
require('./UNSAFE_useHover.js');
require('./UNSAFE_useToggle.js');
require('./UNSAFE_useFocus.js');
require('./UNSAFE_useTouch.js');
require('../utils/UNSAFE_mergeProps.js');
require('./UNSAFE_useAnimation.js');
require('../useAnimation-802f3cdb.js');
require('./UNSAFE_useThemeInterpolations.js');
require('./UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');

require('./UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');


const handleInput = (inputValue, currentInput, onInput) => {
    onInput?.({ value: inputValue });
    currentInput.current = inputValue;
};
const cancelEvent = (event) => {
    event.preventDefault();
    event.stopPropagation();
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function usePointerEvents(value, getValue, targetRef, currentInputRef, onCommit, onInput) {
    const isPointerDown = hooks.useRef(false);
    const pointerUpHandler = (event) => {
        isPointerDown.current = false;
        const val = getValue(event);
        if (val !== undefined) {
            onCommit?.({ value: val });
        }
    };
    const hoverHandler = (event) => {
        if (isPointerDown.current) {
            const val = getValue(event);
            if (val !== undefined && val != currentInputRef.current) {
                handleInput(val, currentInputRef, onInput);
            }
        }
    };
    const pointerDownHandler = (event) => {
        isPointerDown.current = true;
        targetRef.current?.setPointerCapture(event.pointerId);
        const val = getValue(event);
        if (val !== undefined && val != currentInputRef.current) {
            handleInput(val, currentInputRef, onInput);
        }
    };
    const blurHandler = () => {
        isPointerDown.current = false;
        onCommit?.({ value: value });
    };
    return {
        onPointerUp: pointerUpHandler,
        onBlur: blurHandler,
        onPointerMove: hoverHandler,
        onPointerDown: pointerDownHandler
    };
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function useKeyboardEvents(value, min, max, step, currentInputRef, onCommit, onInput) {
    const { direction } = hooks_UNSAFE_useUser.useUser();
    const isRtl = direction === 'rtl';
    const keyDownHandler = (event) => {
        const key = event.key;
        switch (key) {
            case 'Tab':
                return;
            case 'ArrowDown':
                handleInput(Math.max(min, value - step), currentInputRef, onInput);
                cancelEvent(event);
                break;
            case 'ArrowUp':
                handleInput(Math.min(max, value + step), currentInputRef, onInput);
                cancelEvent(event);
                break;
            case 'ArrowLeft': {
                const inputValue = isRtl ? Math.min(max, value + step) : Math.max(min, value - step);
                handleInput(inputValue, currentInputRef, onInput);
                cancelEvent(event);
                break;
            }
            case 'ArrowRight': {
                const inputValue = isRtl ? Math.max(min, value - step) : Math.min(max, value + step);
                handleInput(inputValue, currentInputRef, onInput);
                cancelEvent(event);
                break;
            }
        }
    };
    const keyUpHandler = (event) => {
        const key = event.key;
        switch (key) {
            case 'Enter':
                onCommit?.({ value: value });
                cancelEvent(event);
                break;
            case 'Tab':
                handleInput(value, currentInputRef, onInput);
                cancelEvent(event);
                break;
            case 'Home':
                handleInput(min, currentInputRef, onInput);
                cancelEvent(event);
                break;
            case 'End':
                handleInput(max, currentInputRef, onInput);
                cancelEvent(event);
                break;
        }
    };
    return {
        onKeyUp: keyUpHandler,
        onKeyDown: keyDownHandler
    };
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function useMeterDatatip(value, datatip, ariaDescribedBy) {
    const datatipText = datatip === null ? '' : datatip === undefined ? `${value}` : datatip;
    const [isDatatipOpen, setDatatipOpen] = hooks.useState(false);
    const { tooltipContent, tooltipProps } = hooks_UNSAFE_useTooltip.useTooltipControlled({
        text: datatipText,
        isOpen: isDatatipOpen,
        position: 'top',
        isDisabled: datatip === null,
        anchor: { x: 'pointer', y: 'pointer' },
        variant: 'datatip',
        offset: {
            mainAxis: 8,
            crossAxis: 0
        },
        onToggle: ({ value }) => setDatatipOpen(value)
    });
    tooltipProps['aria-describedby'] = [ariaDescribedBy, tooltipProps['aria-describedby']]
        .filter(Boolean)
        .join(' ');
    return { datatipContent: tooltipContent, datatipProps: tooltipProps };
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

exports.useKeyboardEvents = useKeyboardEvents;
exports.useMeterDatatip = useMeterDatatip;
exports.usePointerEvents = usePointerEvents;
