// @oracle/oraclejet-preact/cjs/hooks/UNSAFE_useTextField.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var hooks_PRIVATE_useMessageSeverity = require('./PRIVATE_useMessageSeverity.js');
require('../utils/UNSAFE_stringUtils.js');
require('./UNSAFE_useFocusWithin.js');
var hooks_UNSAFE_useFormFieldContextProps = require('./UNSAFE_useFormFieldContextProps.js');
var hooks_UNSAFE_useId = require('./UNSAFE_useId.js');
var useFocusWithin = require('../useFocusWithin-3c76895a.js');
var stringUtils = require('../stringUtils-27007290.js');
require('preact/hooks');

/**
 * A custom hook to determine the props for a component that renders
 * a text field
 */
function useTextField({ ariaDescribedBy, contentVariant, helpSourceLink, helpSourceText, isDisabled, isFocused, isLoading, isReadonly, isRequiredShown, labelEdge, messages, styleVariant, userAssistanceDensity, value }) {
    const baseId = hooks_UNSAFE_useId.useId();
    const labelId = labelEdge !== 'none' ? `${baseId}-label` : undefined;
    const inputId = !isReadonly ? `${baseId}-input` : undefined;
    const uaId = !(isDisabled || isReadonly) ? `${baseId}-ua` : undefined;
    const helpIconId = userAssistanceDensity === 'compact' ? `${baseId}-helpIcon` : undefined;
    const { focusProps, isFocused: isInputFocused } = useFocusWithin.useFocusWithin({ isDisabled });
    const hasValue = value !== undefined && (typeof value !== 'string' || value !== '');
    // Create the form field context
    const formFieldContext = hooks_UNSAFE_useFormFieldContextProps.useFormFieldContextProps({
        hasValue,
        isDisabled,
        isFocused,
        isInputFocused,
        isLoading,
        isReadonly,
        value
    });
    // Determine the highest severity from the messages
    const messageSeverity = hooks_PRIVATE_useMessageSeverity.useMessageSeverity(messages);
    // Determine the label variant
    const labelVariant = labelEdge === 'inside'
        ? messageSeverity === 'error'
            ? 'insideError'
            : messageSeverity === 'warning'
                ? 'insideWarning'
                : labelEdge
        : labelEdge !== 'none'
            ? labelEdge
            : undefined;
    return {
        baseId,
        formFieldContext,
        inputProps: {
            id: inputId,
            'aria-describedby': stringUtils.merge([helpIconId, uaId, ariaDescribedBy]),
            'aria-invalid': messageSeverity === 'error'
                ? 'true'
                : undefined,
            variant: styleVariant,
            onBlur: focusProps?.onfocusout,
            onFocus: focusProps?.onfocusin
        },
        labelProps: {
            forId: inputId,
            helpIconId,
            helpSourceLink,
            helpSourceText,
            id: labelId,
            isRequiredShown,
            userAssistanceDensity,
            variant: labelVariant
        },
        textFieldProps: {
            contentVariant,
            statusVariant: /error|warning/.test(messageSeverity ?? '')
                ? messageSeverity
                : undefined,
            styleVariant
        },
        userAssistanceProps: {
            id: uaId
        }
    };
}

exports.useTextField = useTextField;
