// @oracle/oraclejet-preact/es/hooks/PRIVATE_useItemAction.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { useCallback } from 'preact/hooks';
import { keyExtractor } from '../utils/PRIVATE_collectionUtils.js';
import '../utils/UNSAFE_keys.js';
import '../utils/PRIVATE_clientHints.js';
import '../clientHints-26cce6ec.js';

const useItemAction = (currentKey, dataState, onItemAction, itemSelector, allowSpacebar) => {
    if (!dataState || !onItemAction) {
        return {};
    }
    return {
        onClick: useCallback((event) => {
            const key = keyExtractor(event.target, itemSelector);
            if (handleItemAction(key, dataState, onItemAction)) {
                event.stopPropagation();
            }
        }, [dataState, onItemAction]),
        onKeyDown: useCallback((event) => {
            if (event.key === 'Enter') {
                handleItemAction(currentKey, dataState, onItemAction);
            }
            else if (allowSpacebar && event.key === ' ') {
                handleItemAction(currentKey, dataState, onItemAction);
                event.preventDefault();
            }
        }, [dataState, onItemAction, currentKey])
    };
};
const handleItemAction = (key, dataState, onItemAction) => {
    if (key !== undefined) {
        const context = findItemContext(key, dataState);
        context && onItemAction({ context });
        return context !== undefined;
    }
    return false;
};
// returns a ListItemContext based on key
const findItemContext = (key, dataState) => {
    const index = dataState?.data?.findIndex((d) => d.metadata.key === key) ?? -1;
    if (index > -1) {
        const context = {
            index: index + (dataState?.offset || 0),
            data: dataState?.data[index].data,
            metadata: dataState?.data[index].metadata
        };
        return context;
    }
    return undefined;
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

export { useItemAction };
