// @oracle/oraclejet-preact/es/hooks/UNSAFE_useTextField.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { useMessageSeverity } from './PRIVATE_useMessageSeverity.js';
import '../utils/UNSAFE_stringUtils.js';
import './UNSAFE_useFocusWithin.js';
import { useFormFieldContextProps } from './UNSAFE_useFormFieldContextProps.js';
import { useId } from './UNSAFE_useId.js';
import { u as useFocusWithin } from '../useFocusWithin-f96a205e.js';
import { m as merge } from '../stringUtils-5f2d2e94.js';
import 'preact/hooks';

/**
 * A custom hook to determine the props for a component that renders
 * a text field
 */
function useTextField({ ariaDescribedBy, contentVariant, helpSourceLink, helpSourceText, isDisabled, isFocused, isLoading, isReadonly, isRequiredShown, labelEdge, messages, styleVariant, userAssistanceDensity, value }) {
    const baseId = useId();
    const labelId = labelEdge !== 'none' ? `${baseId}-label` : undefined;
    const inputId = !isReadonly ? `${baseId}-input` : undefined;
    const uaId = !(isDisabled || isReadonly) ? `${baseId}-ua` : undefined;
    const helpIconId = userAssistanceDensity === 'compact' ? `${baseId}-helpIcon` : undefined;
    const { focusProps, isFocused: isInputFocused } = useFocusWithin({ isDisabled });
    const hasValue = value !== undefined && (typeof value !== 'string' || value !== '');
    // Create the form field context
    const formFieldContext = useFormFieldContextProps({
        hasValue,
        isDisabled,
        isFocused,
        isInputFocused,
        isLoading,
        isReadonly,
        value
    });
    // Determine the highest severity from the messages
    const messageSeverity = useMessageSeverity(messages);
    // Determine the label variant
    const labelVariant = labelEdge === 'inside'
        ? messageSeverity === 'error'
            ? 'insideError'
            : messageSeverity === 'warning'
                ? 'insideWarning'
                : labelEdge
        : labelEdge !== 'none'
            ? labelEdge
            : undefined;
    return {
        baseId,
        formFieldContext,
        inputProps: {
            id: inputId,
            'aria-describedby': merge([helpIconId, uaId, ariaDescribedBy]),
            'aria-invalid': messageSeverity === 'error'
                ? 'true'
                : undefined,
            variant: styleVariant,
            onBlur: focusProps?.onfocusout,
            onFocus: focusProps?.onfocusin
        },
        labelProps: {
            forId: inputId,
            helpIconId,
            helpSourceLink,
            helpSourceText,
            id: labelId,
            isRequiredShown,
            userAssistanceDensity,
            variant: labelVariant
        },
        textFieldProps: {
            contentVariant,
            statusVariant: /error|warning/.test(messageSeverity ?? '')
                ? messageSeverity
                : undefined,
            styleVariant
        },
        userAssistanceProps: {
            id: uaId
        }
    };
}

export { useTextField };
