// @oracle/oraclejet-preact/es/Common/themes/redwood/theme.typography.tokens.js
/* @oracle/oraclejet-preact: 15.1.2 */
import { x as xUnits, b as max } from '../../../utils-9a0cb8dc.js';
import '../themeContract.css.js';

/**
 * This has been changed to manually entered for now.  We need the fontSize to be based on the scale units.
 *
 * Do not edit directly
 * Generated on Tue, 27 Sep 2022 21:27:57 GMT
 */
const typography = {
    heading: {
        xs: {
            fontSize: xUnits(5),
            lineHeight: '1.4',
            fontWeight: '800'
        },
        sm: {
            fontSize: xUnits(6),
            lineHeight: '1.3333',
            fontWeight: '800'
        },
        md: {
            fontSize: xUnits(7),
            lineHeight: '1.2857',
            fontWeight: '800'
        },
        lg: {
            fontSize: xUnits(8),
            lineHeight: '1.25',
            fontWeight: '800'
        },
        xl: {
            fontSize: xUnits(9),
            lineHeight: '1.222',
            fontWeight: '800'
        },
        '2xl': {
            fontSize: xUnits(10),
            lineHeight: '1.3',
            fontWeight: '800'
        }
    },
    subheader: {
        '2xs': {
            fontSize: xUnits(3.5),
            lineHeight: '1.4286',
            fontWeight: '700'
        },
        xs: {
            fontSize: xUnits(4),
            lineHeight: '1.5',
            fontWeight: '700'
        },
        sm: {
            fontSize: xUnits(5),
            lineHeight: '1.4',
            fontWeight: '700'
        },
        md: {
            fontSize: xUnits(6),
            lineHeight: '1.3333',
            fontWeight: '700'
        },
        lg: {
            fontSize: xUnits(7),
            lineHeight: '1.2857',
            fontWeight: '700'
        },
        xl: {
            fontSize: xUnits(8),
            lineHeight: '1.25',
            fontWeight: '700'
        },
        '2xl': {
            fontSize: xUnits(9),
            lineHeight: '1.2222',
            fontWeight: '700'
        }
    },
    // min values are px units as they are absolute minimums and shouldn't get smaller (or bigger) if the
    // root font-size changes.
    body: {
        '2xs': {
            fontSize: max(xUnits(2.5), '10px'),
            lineHeight: '1.2',
            fontWeight: {
                regular: '400',
                semibold: '600',
                bold: '700'
            }
        },
        xs: {
            fontSize: max(xUnits(3), '12px'),
            lineHeight: '1.3333',
            fontWeight: {
                regular: '400',
                semibold: '600',
                bold: '700'
            }
        },
        sm: {
            fontSize: max(xUnits(3.4375), '12px'),
            lineHeight: '1.2',
            fontWeight: {
                regular: '400',
                semibold: '600',
                bold: '700'
            }
        },
        md: {
            fontSize: max(xUnits(4), '12px'),
            lineHeight: '1.25',
            fontWeight: {
                regular: '400',
                semibold: '600',
                bold: '700'
            }
        },
        lg: {
            fontSize: max(xUnits(4.5), '14px'),
            lineHeight: '1.3333',
            fontWeight: {
                regular: '400',
                semibold: '600',
                bold: '700'
            }
        },
        xl: {
            fontSize: max(xUnits(5), '16px'),
            lineHeight: '1.4',
            fontWeight: {
                regular: '400',
                semibold: '600',
                bold: '700'
            }
        }
    }
};

export { typography };
