// @oracle/oraclejet-preact/cjs/UNSAFE_Label/themes/LabelStyles.css.js
/* @oracle/oraclejet-preact: 15.1.2 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });



var baseStyle = 'LabelStyles_baseStyle__t9j3mr0';
var compoundVariants = [{variants:{inside:'isInside',readonly:'isReadonly',inEnabledForm:'isInEnabledForm'},style:'LabelStyles_insideReadonlyInEnabledForm__t9j3mr5'},{variants:{inside:'isInside',readonly:'isReadonly',inEnabledForm:'notInEnabledForm'},style:'LabelStyles_insideReadonlyNotInEnabledForm__t9j3mr6'},{variants:{inside:'isInside',animatedWhenInside:'notAnimated'},style:'LabelStyles_insideReadonlyNotInEnabledForm__t9j3mr6'},{variants:{inside:'isInside',readonly:'notReadonly',animatedWhenInside:'isAnimated'},style:'LabelStyles_insideNonReadonly__t9j3mr7'},{variants:{inside:'isInside',valueOrFocus:'hasValueOrFocus',animatedWhenInside:'isAnimated'},style:'LabelStyles_insideNonReadonlyHasValue__t9j3mr8'},{variants:{inside:'isInside',readonly:'notReadonly',focused:'isFocused'},style:'LabelStyles_insideFocused__t9j3mra'},{variants:{inside:'isInside',disabled:'isDisabled'},style:'LabelStyles_insideDisabled__t9j3mr9'},{variants:{position:'insideError',disabled:'nonDisabled',readonly:'notReadonly'},style:'LabelStyles_insideError__t9j3mrb'},{variants:{position:'insideWarning',disabled:'nonDisabled',readonly:'notReadonly'},style:'LabelStyles_insideWarning__t9j3mrc'},{variants:{inside:'notInside',labelWrapping:'truncate'},style:'LabelStyles_noWrap__t9j3mrd'},{variants:{inside:'isInside',userAssistanceDensity:'efficient'},style:'LabelStyles_insideBaseUADensityEfficientReflow__t9j3mr4'},{variants:{inside:'isInside',userAssistanceDensity:'reflow'},style:'LabelStyles_insideBaseUADensityEfficientReflow__t9j3mr4'},{variants:{userAssistanceDensity:'compact',position:'start'},style:'LabelStyles_uaDensityCompactStart__t9j3mrf'},{variants:{userAssistanceDensity:'compact',position:'top'},style:'LabelStyles_uaDensityCompactTop__t9j3mrg'}];
var insideBase = 'LabelStyles_insideBase__t9j3mr3';
var insideBaseUADensityEfficientReflow = 'LabelStyles_insideBaseUADensityEfficientReflow__t9j3mr4';
var insideDisabled = 'LabelStyles_insideDisabled__t9j3mr9';
var insideError = 'LabelStyles_insideError__t9j3mrb';
var insideFocused = 'LabelStyles_insideFocused__t9j3mra';
var insideNonReadonly = 'LabelStyles_insideNonReadonly__t9j3mr7';
var insideNonReadonlyHasValue = 'LabelStyles_insideNonReadonlyHasValue__t9j3mr8';
var insideReadonlyInEnabledForm = 'LabelStyles_insideReadonlyInEnabledForm__t9j3mr5';
var insideReadonlyNotInEnabledForm = 'LabelStyles_insideReadonlyNotInEnabledForm__t9j3mr6';
var insideWarning = 'LabelStyles_insideWarning__t9j3mrc';
var noWrap = 'LabelStyles_noWrap__t9j3mrd';
var startBase = 'LabelStyles_startBase__t9j3mr1';
var styles = {uaDensityCompactLabel:'LabelStyles_uaDensityCompactLabel__t9j3mrh',uaDensityCompactLabelStart:'LabelStyles_uaDensityCompactLabelStart__t9j3mri',uaDensityCompactLabelTopInside:'LabelStyles_uaDensityCompactLabelTopInside__t9j3mrj'};
var topBase = 'LabelStyles_topBase__t9j3mr2';
var uaDensityCompactLabel = 'LabelStyles_uaDensityCompactLabel__t9j3mrh';
var uaDensityCompactLabelStart = 'LabelStyles_uaDensityCompactLabelStart__t9j3mri';
var uaDensityCompactLabelTopInside = 'LabelStyles_uaDensityCompactLabelTopInside__t9j3mrj';
var uaDensityCompactStart = 'LabelStyles_uaDensityCompactStart__t9j3mrf';
var uaDensityCompactTop = 'LabelStyles_uaDensityCompactTop__t9j3mrg';
var variants = {position:{top:'LabelStyles_topBase__t9j3mr2',start:'LabelStyles_startBase__t9j3mr1',inside:'',insideError:'',insideWarning:''},inside:{isInside:'LabelStyles_insideBase__t9j3mr3',notInside:''},inEnabledForm:{isInEnabledForm:'',notInEnabledForm:''},labelWrapping:{wrap:'',truncate:''},readonly:{isReadonly:'',notReadonly:''},focused:{isFocused:'',nonFocused:''},disabled:{isDisabled:'',nonDisabled:''},formLayout:{isFormLayout:'',nonFormLayout:''},readonlyForm:{isReadonlyForm:'',notReadonlyForm:''},value:{hasValue:'',noValue:''},valueOrFocus:{hasValueOrFocus:'',noValueOrFocus:''},userAssistanceDensity:{compact:'LabelStyles_uaDensityCompact__t9j3mre',efficient:'',reflow:''},animatedWhenInside:{isAnimated:'',notAnimated:''}};

exports.baseStyle = baseStyle;
exports.compoundVariants = compoundVariants;
exports.insideBase = insideBase;
exports.insideBaseUADensityEfficientReflow = insideBaseUADensityEfficientReflow;
exports.insideDisabled = insideDisabled;
exports.insideError = insideError;
exports.insideFocused = insideFocused;
exports.insideNonReadonly = insideNonReadonly;
exports.insideNonReadonlyHasValue = insideNonReadonlyHasValue;
exports.insideReadonlyInEnabledForm = insideReadonlyInEnabledForm;
exports.insideReadonlyNotInEnabledForm = insideReadonlyNotInEnabledForm;
exports.insideWarning = insideWarning;
exports.noWrap = noWrap;
exports.startBase = startBase;
exports.styles = styles;
exports.topBase = topBase;
exports.uaDensityCompactLabel = uaDensityCompactLabel;
exports.uaDensityCompactLabelStart = uaDensityCompactLabelStart;
exports.uaDensityCompactLabelTopInside = uaDensityCompactLabelTopInside;
exports.uaDensityCompactStart = uaDensityCompactStart;
exports.uaDensityCompactTop = uaDensityCompactTop;
exports.variants = variants;
